---
title: "coverage_rep_RNAseq_test"
author: "Ruijuan Li"
date: "4/15/2017"
output: 
  html_document: 
    keep_md: yes
---

```{r}
# goal: 
# determine the most cost effecient coverage (1 VS 3/4 VS 1/2 VS 1/4) & rep (3 VS 2 VS 1) # for illumina RNAseq by checking the correlation of glmfit fitted values for reduced dataset with the full dataset (3 reps & full coverage)

# Trimmomatic trimming --> STAR mapping --> ReadsPerGene.out.tab    
```

# function 
```{r}
library(edgeR)
library(ggplot2)

##### with reps 
get.fitted.value <- function(reads.formated){

  # check test data type 
  gene_num_dim <- dim(reads.formated) 
  cat(cat(gene_num_dim[1]) , cat(" genes ") , cat("exsited in this dataset with "), cat(gene_num_dim[2]/2), cat(" reps")) 
  cat("\n") 
  
  # set up sample description 
  reads.formated.sample <-data.frame(file=colnames(reads.formated),
                               genotype=factor(gsub("([[:print:]]+)(_)([[:print:]]+)(_)(1|2|3)","\\1",colnames(reads.formated))),	
                               stage=factor(gsub("([[:print:]]+)(_)([[:print:]]+)(_)(1|2|3)","\\3",colnames(reads.formated))),	
                               group=factor(gsub("([[:print:]]+)(_)([[:print:]]+)(_)(1|2|3)","\\1\\3",colnames(reads.formated)))
  )
  
  # filter based on read count 
  cat("number of expressed genes if filtered based greater than 10 reads in at least 3 libs:")
  reads.formated.small <- reads.formated[rowSums(reads.formated > 10) >= 3,]
  cat(dim(reads.formated.small)[1])
  cat("\n")
  
  # normalize 
  dge.reads.formated <- DGEList(counts=reads.formated, group=reads.formated.sample$group)
  dge.reads.formated <- calcNormFactors(dge.reads.formated, method = "TMM") 
  
  # pairwise comparison using GLM model 
  # design matrix # w/o batch effect
  design.reads.formated <- model.matrix(~0+group, data = reads.formated.sample)
  dge.reads.formated <- estimateGLMCommonDisp(dge.reads.formated, design.reads.formated,verbose = F) # Disp = 0.13675 , BCV = 0.3698
  dge.reads.formated <- estimateGLMTrendedDisp(dge.reads.formated,design.reads.formated)
  dge.reads.formated <- estimateGLMTagwiseDisp(dge.reads.formated,design.reads.formated)
  
  ## fit model & look for DEGs
  fit.reads.formated <- glmFit(dge.reads.formated, design.reads.formated)
  
  # colnames(fit.flowering.halfhalf$coefficients)
  lrt.reads.formated <- glmLRT(fit.reads.formated, contrast = c(1, -1))
  number.reads.formated <- summary(de.reads.formated <- decideTestsDGE(lrt.reads.formated, p=0.05))
  DEgene.reads.formated <- topTags(lrt.reads.formated,n = Inf)$table[topTags(lrt.reads.formated,n = Inf)$table$FDR<0.05,]
  
  cat("number of differentially expressed genes between genotypes:")
  cat(dim(DEgene.reads.formated)[1])
  cat("\n")
  
  return(fit.reads.formated)
} 

##### w/o rep 
get.fitted.value.no.rep <- function(reads.formated){

  # check test data type 
  gene_num_dim <- dim(reads.formated) 
  cat(cat(gene_num_dim[1]) , cat(" genes ") , cat("exsited in this dataset with "), cat(gene_num_dim[2]/2), cat(" reps")) 
  cat("\n") 
  
  # set up sample description 
  reads.formated.sample <-data.frame(file=colnames(reads.formated),
                               genotype=factor(gsub("([[:print:]]+)(_)([[:print:]]+)(_)(1|2|3)","\\1",colnames(reads.formated))),	
                               stage=factor(gsub("([[:print:]]+)(_)([[:print:]]+)(_)(1|2|3)","\\3",colnames(reads.formated))),	
                               group=factor(gsub("([[:print:]]+)(_)([[:print:]]+)(_)(1|2|3)","\\1\\3",colnames(reads.formated)))
  )
  
  # filter based on read count 
  cat("number of expressed genes if filtered based greater than 10 reads in 2 libs:")
  reads.formated.small <- reads.formated[rowSums(reads.formated > 10) >= 2,]
  cat(dim(reads.formated.small)[1])
  cat("\n")
  
  # normalize 
  dge.reads.formated <- DGEList(counts=reads.formated, group=reads.formated.sample$group)
  dge.reads.formated <- calcNormFactors(dge.reads.formated, method = "TMM") 
  
  # pairwise comparison using GLM model 
  # design matrix # w/o batch effect
  design.reads.formated <- model.matrix(~0+group, data = reads.formated.sample)
  # pick a reasonable dispersion value for data w/o rep to account for biological varaibility ... 
  bcv <- 0.4 # pick 0.4 because this is close to the value I get from 3 reps 
  dge.reads.formated <- estimateGLMCommonDisp(dge.reads.formated, design.reads.formated,verbose = F) # Disp = 0.13675 , BCV = 0.3698
  cat("set common dispersion to 0.4\n")  
  dge.reads.formated$common.dispersion <- bcv ^ 2
  # dge.reads.formated <- estimateGLMTrendedDisp(dge.reads.formated,design.reads.formated)
  # dge.reads.formated <- estimateGLMTagwiseDisp(dge.reads.formated,design.reads.formated)

  ## fit model & look for DEGs
  fit.reads.formated <- glmFit(dge.reads.formated, design.reads.formated)
  
  # colnames(fit.flowering.halfhalf$coefficients)
  lrt.reads.formated <- glmLRT(fit.reads.formated, contrast = c(1, -1))
  number.reads.formated <- summary(de.reads.formated <- decideTestsDGE(lrt.reads.formated, p=0.05))
  DEgene.reads.formated <- topTags(lrt.reads.formated,n = Inf)$table[topTags(lrt.reads.formated,n = Inf)$table$FDR<0.05,]
  
  cat("number of differentially expressed genes between genotypes:")
  cat(dim(DEgene.reads.formated)[1])
  cat("\n")
  
  return(fit.reads.formated) 
}  

``` 

## coverage test (all VS 3/4 VS 1/2 VS 1/4)
```{r} 
######### all trimmed data 
if(!Sys.getenv("USER")=="jmaloof")  setwd("~/Desktop/Brassica_project/KIAT_RNA_seq/analysis/")
flowering.read.count.all <- read.table("read_count/read.count.flowering.all.tsv", header = T, check.names = F)
rownames(flowering.read.count.all) <- flowering.read.count.all[,1]
flowering.read.count.all <- flowering.read.count.all[,-1]

# format data 
head(flowering.read.count.all)
dim(flowering.read.count.all) # 101040     6 
colnames(flowering.read.count.all)

# replace sample ID 
colnames(flowering.read.count.all) <- c("Da-Ol-1_flowering_3","Da-Ae_flowering_3","Da-Ae_flowering_1","Da-Ae_flowering_2","Da-Ol-1_flowering_1","Da-Ol-1_flowering_2")
head(flowering.read.count.all)

fitted.all.3 <- get.fitted.value(flowering.read.count.all)

##### half of the reads 
flowering.read.count.half <- read.table("read_count/read.count.flowering.half.tsv", header = T, check.names = F)
rownames(flowering.read.count.half) <- flowering.read.count.half[,1]
flowering.read.count.half <- flowering.read.count.half[,-1]

# format data 
head(flowering.read.count.half)
dim(flowering.read.count.half) # 101040     6 
colnames(flowering.read.count.half)

# replace sample ID 
colnames(flowering.read.count.half) <- c("Da-Ol-1_flowering_3","Da-Ae_flowering_3","Da-Ae_flowering_1","Da-Ae_flowering_2","Da-Ol-1_flowering_1","Da-Ol-1_flowering_2")
head(flowering.read.count.half)

fitted.half.3 <- get.fitted.value(flowering.read.count.half)

##### 1/4 of the reads 
flowering.read.count.halfhalf <- read.table("read_count/read.count.flowering.halfhalf.tsv", header = T, check.names = F)
rownames(flowering.read.count.halfhalf) <- flowering.read.count.halfhalf[,1]
flowering.read.count.halfhalf <- flowering.read.count.halfhalf[,-1]

# format data 
head(flowering.read.count.halfhalf)
dim(flowering.read.count.halfhalf) # 101040     6 
colnames(flowering.read.count.halfhalf)

# replace sample ID 
colnames(flowering.read.count.halfhalf) <- c("Da-Ol-1_flowering_3","Da-Ae_flowering_3","Da-Ae_flowering_1","Da-Ae_flowering_2","Da-Ol-1_flowering_1","Da-Ol-1_flowering_2") 
head(flowering.read.count.halfhalf)

fitted.quarter.3 <- get.fitted.value(flowering.read.count.halfhalf)

### 3/4 
flowering.read.count.3_4 <- read.table("read_count/read.count.flowering.3_4.tsv", header = T, check.names = F)
rownames(flowering.read.count.3_4) <- flowering.read.count.3_4[,1]
flowering.read.count.3_4 <- flowering.read.count.3_4[,-1]

# format data 
head(flowering.read.count.3_4)
dim(flowering.read.count.3_4) # 101040     6 
colnames(flowering.read.count.3_4)

# replace sample ID 
colnames(flowering.read.count.3_4) <- c("Da-Ol-1_flowering_3","Da-Ae_flowering_3","Da-Ae_flowering_1","Da-Ae_flowering_2","Da-Ol-1_flowering_1","Da-Ol-1_flowering_2")
head(flowering.read.count.3_4)

fitted.three.quarter.3 <- get.fitted.value(flowering.read.count.3_4) 
```

# rep test: 2 reps & no rep
```{r}
##### two reps 
flowering.read.count.two.rep <- read.table("read_count/read.count.flowering.two.rep.tsv", header = T, check.names = F)
rownames(flowering.read.count.two.rep) <- flowering.read.count.two.rep[,1]
flowering.read.count.two.rep <- flowering.read.count.two.rep[,-1]

# format data 
head(flowering.read.count.two.rep)
dim(flowering.read.count.two.rep) # 101040     6 
colnames(flowering.read.count.two.rep)

# replace sample ID 
colnames(flowering.read.count.two.rep) <- c("Da-Ol-1_flowering_1","Da-Ae_flowering_1","Da-Ae_flowering_2","Da-Ol-1_flowering_2")
head(flowering.read.count.two.rep)

fitted.all.two.rep <- get.fitted.value(flowering.read.count.two.rep)

##### no rep (1)  
flowering.read.count.no.rep <- read.table("read_count/read.count.flowering.no.rep.tsv", header = T, check.names = F)
rownames(flowering.read.count.no.rep) <- flowering.read.count.no.rep[,1]
flowering.read.count.no.rep <- flowering.read.count.no.rep[,-1]

# format data 
head(flowering.read.count.no.rep)
dim(flowering.read.count.no.rep) # 101040     6 
colnames(flowering.read.count.no.rep)

# replace sample ID 
colnames(flowering.read.count.no.rep) <- c("Da-Ol-1_flowering_1","Da-Ae_flowering_1")
head(flowering.read.count.no.rep)

fitted.all.no.rep <- get.fitted.value.no.rep(flowering.read.count.no.rep) 

# (2)
flowering.read.count.no.rep.2 <- read.table("read_count/read.count.flowering.all.tsv", header = T, check.names = F)
head(flowering.read.count.no.rep.2)
rownames(flowering.read.count.no.rep.2) <- flowering.read.count.no.rep.2[,1]
flowering.read.count.no.rep.2 <- flowering.read.count.no.rep.2[,-c(1, 2, 3, 5, 7)]

# format data 
head(flowering.read.count.no.rep.2)
dim(flowering.read.count.no.rep.2) # 101040     6 
colnames(flowering.read.count.no.rep.2)

# replace sample ID 
colnames(flowering.read.count.no.rep.2) <- c("Da-Ae_flowering_2","Da-Ol-1_flowering_2")
head(flowering.read.count.no.rep.2)

fitted.all.no.rep.2 <- get.fitted.value.no.rep(flowering.read.count.no.rep.2)

# (3)
flowering.read.count.no.rep.3 <- read.table("read_count/read.count.flowering.all.tsv", header = T, check.names = F)
head(flowering.read.count.no.rep.3)
rownames(flowering.read.count.no.rep.3) <- flowering.read.count.no.rep.3[,1]
flowering.read.count.no.rep.3 <- flowering.read.count.no.rep.3[,-c(1, 2, 3, 4, 6)]

# format data 
head(flowering.read.count.no.rep.3)
dim(flowering.read.count.no.rep.3) # 101040     6 
colnames(flowering.read.count.no.rep.3)

# replace sample ID 
colnames(flowering.read.count.no.rep.3) <- c("Da-Ae_flowering_3","Da-Ol-1_flowering_3")
head(flowering.read.count.no.rep.3)

fitted.all.no.rep.3 <- get.fitted.value.no.rep(flowering.read.count.no.rep.3)
``` 

# correlation test for the fitted coefficient  
```{r} 
hist(fitted.all.3$coefficients)

filter <- apply(fitted.all.3$coefficients,1,max) > -18 #get rid of extremely low expressed genes, so that this is more normal

# fitted coefficient 
ref.vs.rep2 <- cor(fitted.all.3$coefficients[filter,], fitted.all.two.rep$coefficients[filter,])
ref.vs.rep1 <- cor(fitted.all.3$coefficients[filter,], fitted.all.no.rep$coefficients[filter,])
ref.vs.rep1.2 <- cor(fitted.all.3$coefficients[filter,], fitted.all.no.rep.2$coefficients[filter,]) # rep2
ref.vs.rep1.3 <- cor(fitted.all.3$coefficients[filter,], fitted.all.no.rep.3$coefficients[filter,]) # rep3
ref.vs.three.quarter <- cor(fitted.all.3$coefficients[filter,], fitted.three.quarter.3$coefficients[filter,])
ref.vs.one.quarter <- cor(fitted.all.3$coefficients[filter,], fitted.quarter.3$coefficients[filter,])
ref.vs.half <- cor(fitted.all.3$coefficients[filter,], fitted.half.3$coefficients[filter,])

ref.vs.rep2 <- ref.vs.rep2[1,]
ref.vs.rep1 <- ref.vs.rep1[1,]
ref.vs.rep1.2 <- ref.vs.rep1.2[1,]
ref.vs.rep1.3 <- ref.vs.rep1.3[1,]
ref.vs.three.quarter <- ref.vs.three.quarter[1,]
ref.vs.one.quarter <- ref.vs.one.quarter[1,]
ref.vs.half <- ref.vs.half[1,]

coefficients.fit <- as.data.frame(rbind(ref.vs.three.quarter, ref.vs.rep2, ref.vs.half, ref.vs.rep1, ref.vs.rep1.2, ref.vs.rep1.3, ref.vs.one.quarter))
coefficients.fit$value <- as.numeric(rowMeans(coefficients.fit))
coefficients.fit

coefficients.fit$type <- factor(rownames(coefficients.fit), levels = c("ref.vs.three.quarter", "ref.vs.rep2", "ref.vs.half", "ref.vs.one.quarter", "ref.vs.rep1", "ref.vs.rep1.2", "ref.vs.rep1.3"))
coefficients.fit$type <- factor(rownames(coefficients.fit), levels = c("ref.vs.three.quarter", "ref.vs.half", "ref.vs.one.quarter", "ref.vs.rep2","ref.vs.rep1", "ref.vs.rep1.2", "ref.vs.rep1.3"),labels=c("ref.vs.three.quarter", "ref.vs.half", "ref.vs.one.quarter","ref.vs.two.reps","ref.vs.one.rep", "ref.vs.one.rep.2", "ref.vs.one.rep.3"))

# plot 
p.coefficient <- ggplot(data = coefficients.fit)
p.coefficient <- p.coefficient + geom_count(mapping = aes(x = type, y = value^2), stat = "identity")
p.coefficient <- p.coefficient + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0)) 
p.coefficient <- p.coefficient + labs(y = "squared coefficient correlation", x = "")

p.coefficient  
#ggsave(p.coefficient, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/output/p.coefficients.png", width = 4, height = 4)    
```









