---
title: "KIAT_RNA_seq"
author: "Ruijuan Li"
date: "August 1, 2016"
output: html_document
---

# download data, QC, and mapping (using kallisto)
```{r}
# 1) download file from ftp using ftp_file_transfer.sh
# 1.1) unzip files using gunzip *  
# 2) get first 1000 sequences for each file extract_1000_reads.sh, actually worked with all data instead of the 1st 1000 seqs
# 3) QC file using fastQC: fastqc *.fq -o /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/fastqc_result
# 4) download Brassica_napus.annotation_v5.cds.fa to whitney 
# 5) collapse fastqc summary result into one file & get overrepresented sequences using ./make_summary_report.sh 

# 6) reformat summary.all.txt file 
setwd("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/")
fastQC_summary <- read.delim("./fastQC_result/summary_all.txt", header = F)
dim(fastQC_summary)
head(fastQC_summary)
tail(fastQC_summary)

fastQC_summary_wide <- reshape(fastQC_summary, timevar="V2",idvar="V3",direction="wide")
colnames(fastQC_summary_wide) <- gsub("(V1)(.)([[:print:]]+)","\\3",colnames(fastQC_summary_wide))
dim(fastQC_summary_wide)

write.csv(fastQC_summary_wide, file = "fastQC_summary_wide.csv")

# 7) checked the source of overrepresented sequences using web nucleotide blast tool, found that most of them are from 
# cholorplast, mitochondra, rRNA, tansposon, etc. Some are illumina or TruSeq adapters 

# blast sequences against Trimmomatic adapters suggest that TruSeq3-PE-2.fa should be used to trimm the data. (Q: how to deal with the others? will they affect the final result? search... )

# Qs on plstids seqs contamination: 1) why are there these sequences? check the lib making step 2) will they affect downstream analysis if kept? CDS mapping VS genome mapping 3) If yes, how to remove them? (I need to figure these Qs out.)

# 8) using kallisto to build the reference cds index 
# kallisto index -k 19 -i Brassica_napus.annotation_v5.cds.19.kai /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/Brassica_napus.annotation_v5.cds.fa

# 9) map to reference cds index and get read count file using bunchrun_kallisto.sh (using est_counts from Kallisto)
```

# using STAR 
```{r}
# 1) get B.napus reference genome: Brassica_napus_v4.1.chromosomes.fa 
# make STAR reference index: STAR --runMode genomeGenerate --genomeDir star_nome/ --genomeFastaFiles Brassica_napus_v4.1.chromosomes.fa --runThreadN 2 
# mapping, only map the 2nd end using run_star.sh 

# play with SAM/BAM file 
# 1) convert SAM/BAM file 
# samtools view -h file.bam > file.sam
# samtools view -b -S file.sam > file.bam  

# 2) sort BAM file 
# samtools sort -m 1000000000 test.bam -T test_sorted (running)

# --> to do: use cufflinks for 







```

# SNP calling from all individuals & check sample seperation 
```{r}




```

# Expression analysis 1) formatting data   
```{r}
parent.read.count <- read.table("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/read_count/read.count.tsv", header = T, check.names = F)
rownames(parent.read.count) <- parent.read.count[,1]
parent.read.count <- parent.read.count[,-1]

# format data 
head(parent.read.count)
dim(parent.read.count) # 101040     54 
colnames(parent.read.count)
even_indexes<-seq(2,length(colnames(parent.read.count)),2)
parent.read.count.one <- parent.read.count[,even_indexes]
dim(parent.read.count.one) # 101040     27 
colnames(parent.read.count.one)
colnames(parent.read.count.one) <- sub("_2.fq","",colnames(parent.read.count.one),fixed = TRUE) # remove _2.fq 

# sample description  
sample_des <- read.csv("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/parent_summary_corrected.csv")
dim(sample_des) # 27 22
sorted_sample_des <- sample_des[order(sample_des$SampleID),]

sorted_sample_des[,4:7]
new_sample_ID <- paste(sorted_sample_des$Cultivar, sorted_sample_des$Stage, sorted_sample_des$rep, sep = "_")
new_sample_ID

# replace sample ID 
colnames(parent.read.count.one) <- new_sample_ID
head(parent.read.count.one)
save(parent.read.count.one,file="/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.Rdata")
```

# fatty acid biosynthesis pathways genes (GO:0006633)
```{r}
fatty_acid_gene_index <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/fatty_acid_biosynthesis_genes_ID.txt", header = F)
head(fatty_acid_gene_index)
nrow(fatty_acid_gene_index)
rownames(fatty_acid_gene_index) <- fatty_acid_gene_index$V1

# extract only fatty acid genes from dataset 
fatty.acid.genes <- (merge(parent.read.count.one.small, fatty_acid_gene_index, by="row.names"))[,1:28]
head(fatty.acid.genes)
rownames(fatty.acid.genes) <- fatty.acid.genes$Row.names
fatty.acid.genes <- fatty.acid.genes[,-1]
dim(fatty.acid.genes) # 207 27 
fatty.acid.genes$total <- rowSums(fatty.acid.genes)
dim(fatty.acid.genes)

# # get only the top 5 highly expressed fatty acid genes 
# five.fatty.acid.genes <- head(tmp[order(tmp$total, decreasing = T), ], 5)
# five.fatty.acid.genes
# write.csv(t(five.fatty.acid.genes), file =  "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/five.fatty.acid.gene.csv")
# 
# # reformat for ggplot 
# five.fatty.acid.genes <- five.fatty.acid.genes[,-28]
# # colnames(five.fatty.acid.genes) <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1_\\3",colnames(five.fatty.acid.genes))
# five.fatty.acid.genes$geneID <- rownames(five.fatty.acid.genes)
# five.fatty.acid.genes.melt <- melt(five.fatty.acid.genes)
# five.fatty.acid.genes.melt$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",five.fatty.acid.genes.melt$variable)
# five.fatty.acid.genes.melt$tissue <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",five.fatty.acid.genes.melt$variable)
# five.fatty.acid.genes.melt$rep <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\5",five.fatty.acid.genes.melt$variable)
# five.fatty.acid.genes.melt$group <- paste(five.fatty.acid.genes.melt$gt, five.fatty.acid.genes.melt$tissue, five.fatty.acid.genes.melt$geneID, sep = "_")
# head(five.fatty.acid.genes.melt)
# tail(five.fatty.acid.genes.melt)
# 
# # to do: reshape, calculate & make ggplot ... 
# five.fatty.acid.genes.melt.reshape <- reshape(five.fatty.acid.genes.melt[,c("value", "rep", "group")], idvar = "group", direction = "wide", timevar = "rep")
# dim(five.fatty.acid.genes.melt.reshape)
# five.fatty.acid.genes.melt.reshape
# 
# five.fatty.acid.genes.melt.reshape$mean <- apply(five.fatty.acid.genes.melt.reshape[,c(2:4)], 1, mean, na.rm=TRUE)
# five.fatty.acid.genes.melt.reshape$min <- apply(five.fatty.acid.genes.melt.reshape[,c(2:4)], 1, max, na.rm=T)
# five.fatty.acid.genes.melt.reshape$max <- apply(five.fatty.acid.genes.melt.reshape[,c(2:4)], 1, min, na.rm=T)
# five.fatty.acid.genes.melt.reshape
# five.fatty.acid.genes.melt.reshape$gt <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\1",five.fatty.acid.genes.melt.reshape$group)
# five.fatty.acid.genes.melt.reshape$tissue <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\3",five.fatty.acid.genes.melt.reshape$group)
# five.fatty.acid.genes.melt.reshape$geneID <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\5",five.fatty.acid.genes.melt.reshape$group)
# five.fatty.acid.genes.melt.reshape 
# 
# p.fatty.acid <- ggplot(data = five.fatty.acid.genes.melt.reshape)
# p.fatty.acid <- p.fatty.acid + geom_line(aes(x = factor(tissue), y = mean,group=gt, color=gt))
# p.fatty.acid <- p.fatty.acid + facet_grid(~geneID~gt)
# p.fatty.acid <- p.fatty.acid + geom_errorbar(mapping=aes(x=tissue,ymin=min,ymax=max, width=0.25))
# p.fatty.acid <- p.fatty.acid + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
# p.fatty.acid <- p.fatty.acid + labs(y = "mean expression value", x="tissue")
# 
# p.fatty.acid   
# 
# ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/F1/fatty.acid.png", width = 8, height = 11)
```

# data summary statisitics read count number & expressed genes 
```{r}
parent.read.count.normalized <- read.table("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/read.count.normalized.tsv", header = T, check.names = F)
rownames(parent.read.count.normalized) <- parent.read.count.normalized[,1]
parent.read.count.normalized <- parent.read.count.normalized[,-1]

# format data 
head(parent.read.count.normalized)
dim(parent.read.count.normalized) # 101040     54 
colnames(parent.read.count.normalized)
even_indexes<-seq(2,length(colnames(parent.read.count.normalized)),2)
parent.read.count.one.normalized <- parent.read.count.normalized[,even_indexes]
dim(parent.read.count.one.normalized) # 101040     27 
colnames(parent.read.count.one.normalized)
colnames(parent.read.count.one.normalized) <- sub("_2.fq","",colnames(parent.read.count.one.normalized),fixed = TRUE) # remove _2.fq 

# sample description  
new_sample_ID

# replace sample ID 
colnames(parent.read.count.one.normalized) <- new_sample_ID
head(parent.read.count.one.normalized)
save(parent.read.count.one.normalized,file="/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.normalized.Rdata")

# check number of genes with reads number > 10 in each lib # make summary table 
expressed_gene_number <- 
sapply(colnames(parent.read.count.one.normalized), function(RNAlib){
  sum(parent.read.count.one.normalized[,RNAlib]>3)
}
)
expressed_gene_number <- as.data.frame(expressed_gene_number)
expressed_gene_number

# top 10 highly expressed genes in each lib 
most_highly_expressed_genes <- 
sapply(colnames((parent.read.count.one.normalized)), function(RNAlib){
  rownames(head(parent.read.count.one.normalized[order(parent.read.count.one.normalized[,RNAlib],decreasing = TRUE),], 10))
}
)
write.csv(t(most_highly_expressed_genes), file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/most_highly_expressed_genes.csv")

temp <- sample_des[,c(4,5,6,8)]
rownames(temp) <- paste(sample_des[,4], sample_des[,5], sample_des[,6], sep = "_")
sample_statistics <- merge(temp, expressed_gene_number, by=0) 
colnames(sample_statistics) <- c("sample_ID", "Cultivar", "Tissue", "rep", "total_reads", "expressed_gene_number")
head(sample_statistics)
colnames(sample_statistics)
write.csv(sample_statistics, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/sample_statisitcs.csv")

```


# expression analysis 2) set up sample description 
```{r}
# load necessary libs & functions 
library(edgeR)
library(ggplot2)

# filter based on read count, assign group, normalize, design matrix, calculate dispersion   
# set up group 
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.Rdata")
parent.read.count.one <- parent.read.count.one[,colSums(parent.read.count.one) > 1000000]  
dim(parent.read.count.one) # 101040     27 
parent.read.count.one.sample<-data.frame(file=colnames(parent.read.count.one),
                             batch=factor(gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\5",colnames(parent.read.count.one))),  
                             genotype=factor(gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",colnames(parent.read.count.one))),	
                             stage=factor(gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",colnames(parent.read.count.one))),	
                             group=factor(gsub("(Da-Ae|Da-Ol-1)(_)([[:print:]]+)(_)(1|2|3)","\\1\\3",colnames(parent.read.count.one)))
)

parent.read.count.one.sample
ftable(parent.read.count.one.sample,row.vars="stage",col.vars=c("batch","genotype"))

# filter based on read count 
parent.read.count.one.small <- parent.read.count.one[rowSums(parent.read.count.one > 10) >= 3,]
dim(parent.read.count.one.small) # 60975    27 
save(parent.read.count.one.small, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.small.Rdata")
```

# expression analysis 3) voom transformation 
```{r}
# voom transformation 
# source("https://bioconductor.org/biocLite.R")
# biocLite("DESeq2")
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.small.Rdata")
library("DESeq2")

dds.parent <- DESeqDataSetFromMatrix(countData = round(parent.read.count.one.small), colData = parent.read.count.one.sample, design = ~ batch + genotype*stage)

vsd.parent <- varianceStabilizingTransformation(dds.parent)
vstMat.parent <- assay(vsd.parent)
colnames(vstMat.parent) <- colnames(parent.read.count.one)
save(vstMat.parent, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/vstMat.parent.Rdata")
```

# expression analysis 4) MDS, clustering, and expression analysis 
```{r}
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/vstMat.parent.Rdata")
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.small.Rdata")

# normalize 
dge.data.parent <- DGEList(counts=parent.read.count.one.small, group=parent.read.count.one.sample$group)
dge.data.parent <- calcNormFactors(dge.data.parent, method = "TMM") 
dge.data.parent$sample
hist(dge.data.parent$samples$norm.factors)

# MDS to check sample seperation, also using dengrogram.  
mds <- plotMDS(dge.data.parent, method = "bcv",labels = dge.data.parent$samples$group)

x <- as.data.frame(mds$x)
y <- as.data.frame(mds$y)
distance_matrix <- merge(x, y, by="row.names")
distance_matrix$group <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1\\3",distance_matrix$Row.names)
distance_matrix$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",distance_matrix$Row.names)
distance_matrix$tissue <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",distance_matrix$Row.names)

colnames(distance_matrix) <- c("lib","x","y","group","gt","tissue")
head(distance_matrix)

# making color MDS figure 
p.mds <- ggplot(data = distance_matrix)
p.mds <- p.mds + geom_point(aes(x, y, color=factor(tissue)), size=3) 
p.mds <- p.mds + labs(y = "BCV distance 2", x="BCV distance 1")
p.mds <- p.mds + facet_grid(~gt)
p.mds

ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/MDS.png", width = 11, height = 8)

# clustering to check sample seperation 
library(ggdendro)

hc.parent <- hclust(dist(t(vstMat.parent)))

ggdata.parent <- dendro_data(as.dendrogram(hc.parent))
ggdata.parent$labels$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",ggdata.parent$labels$label) 
ggdata.parent$labels$tissue <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",ggdata.parent$labels$label)
ggdata.parent$labels$group <- paste(ggdata.parent$labels$gt, ggdata.parent$labels$tissue, sep = "_")
ggdata.parent$labels

# start ggplot here 
p.dengro <- ggplot(data = segment(ggdata.parent))
p.dendro <- p.dengro + geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) + theme_dendro()
p.dendro <- p.dendro + geom_text(data = label(ggdata.parent), aes(x = x, y = y, label = label, hjust=0, color=group)) + coord_flip() + scale_y_reverse(expand=c(0.2, 0))
p.dendro
ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/clustering_new.png", width = 15, height = 8) 

# # check correlation among samples 
# cor.test(parent.read.count.one.small$`Da-ae_Young_1`, parent.read.count.one.small$`Da-ae_Young_2`)
# cor.test(parent.read.count.one.small$`Da-al-1_Young_2`, parent.read.count.one.small$`Da-al-1_Young_1`)
# cor.test(parent.read.count.one.small$`Da-al-1_Young_1`, parent.read.count.one.small$`Da-ae_early-silique_3`)

# pairwise comparison using GLM model 
# design matrix
# design.parent.batch <- model.matrix(~batch+group, data = parent.read.count.one.sample) 
design.parent <- model.matrix(~0+group, data = parent.read.count.one.sample)

# # estimate dispersion
# # w/ batch effect
# dge.data.parent.batch <- estimateGLMCommonDisp(dge.data.parent.batch, design.parent.batch,verbose = TRUE) # Disp = 0.22238 , BCV = 0.4716
# dge.data.parent.batch <- estimateGLMTrendedDisp(dge.data.parent.batch,design.parent.batch)
# dge.data.parent.batch <- estimateGLMTagwiseDisp(dge.data.parent.batch,design.parent.batch)
# plotBCV(dge.data.parent.batch)

# w/o batch effect
dge.data.parent <- estimateGLMCommonDisp(dge.data.parent, design.parent,verbose = TRUE) # Disp = 0.25572 , BCV = 0.5057
dge.data.parent <- estimateGLMTrendedDisp(dge.data.parent,design.parent)
dge.data.parent <- estimateGLMTagwiseDisp(dge.data.parent,design.parent)
plotBCV(dge.data.parent)

# fit GLM model & test for differential expression
# w/ batch effect
# fit.parent.batch <- glmFit(dge.data.parent.batch, design.parent.batch)
# lrt.young.batch <- glmLRT(fit.parent.batch, contrast = c(0,0,0,0,0,0,1,0,0,0,0,-1))
# topTags(lrt.young.batch)
# DEgene.young.batch <- topTags(lrt.young.batch,n = Inf)$table[topTags(lrt.young.batch,n = Inf)$table$FDR<0.05,]
# summary(de.young.batch <- decideTestsDGE(lrt.young.batch, p=0.05))
# # summary(de.young.batch <- decideTestsDGE(lrt.young.batch, p=0.05, lfc = 2))
# # -1  5072
# # 0  51131
# # 1   4772
# 
# detags.young <- rownames(lrt.young.batch)[as.logical(de.young.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.young.png")
# plotSmear(lrt.young.batch, de.tags=detags.young, main="young")
# abline(h = c(-2, 2), col = "blue")
# dev.off()
# 
# lrt.flowering.batch <- glmLRT(fit.parent.batch, contrast = c(0,0,0,0,1,0,0,0,0,-1,0,0))
# topTags(lrt.flowering.batch)
# summary(de.flowering.batch <- decideTestsDGE(lrt.flowering.batch, p=0.05))
# DEgene.flowering.batch <- topTags(lrt.flowering.batch,n = Inf)$table[topTags(lrt.flowering.batch,n = Inf)$table$FDR<0.05,]
# # -1  3150
# # 0  54342
# # 1   3483
# detags.flowering <- rownames(lrt.flowering.batch)[as.logical(de.flowering.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.flowering.png")
# plotSmear(lrt.flowering.batch, de.tags=detags.flowering, main="flowering")
# abline(h = c(-2, 2), col = "blue")
# dev.off()
# 
# lrt.early.silique.batch <- glmLRT(fit.parent.batch, contrast = c(0,0,0,1,0,0,0,0,-1,0,0,0))
# topTags(lrt.early.silique.batch)
# summary(de.early.silique.batch <- decideTestsDGE(lrt.early.silique.batch, p=0.05))
# DEgene.early.silique.batch <- topTags(lrt.early.silique.batch,n = Inf)$table[topTags(lrt.early.silique.batch,n = Inf)$table$FDR<0.05,]
# # -1  3542
# # 0  53647
# # 1   3786
# detags.early.silique <- rownames(lrt.early.silique.batch)[as.logical(de.early.silique.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.early.silique.png")
# plotSmear(lrt.early.silique.batch, de.tags=detags.early.silique, main="early silique")
# abline(h = c(-2, 2), col = "blue")
# dev.off()
# 
# 
# lrt.late.silique.batch <- glmLRT(fit.parent.batch, contrast = c(0,0,0,0,0,1,0,0,0,0,-1,0))
# topTags(lrt.late.silique.batch)
# summary(de.late.silique.batch <- decideTestsDGE(lrt.late.silique.batch, p=0.05))
# # -1  4064
# # 0  52880
# # 1   4031
# detags.late.silique <- rownames(lrt.late.silique.batch)[as.logical(de.late.silique.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.late.silique.png")
# plotSmear(lrt.late.silique.batch, de.tags=detags.late.silique, main="late silique")
# abline(h = c(-2, 2), col = "blue")
# dev.off()
# 
# lrt.bolting.batch <- glmLRT(fit.parent.batch, coef = "groupDa-al-1bolting")
# topTags(lrt.bolting.batch)
# summary(de.bolting.batch <- decideTestsDGE(lrt.bolting.batch, p=0.05, lfc = 1))
# # -1  1949
# # 0  55971
# # 1   3055
# detags.bolting <- rownames(lrt.bolting.batch)[as.logical(de.bolting.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.bolting.png")
# plotSmear(lrt.bolting.batch, de.tags=detags.bolting, main="bolting")
# abline(h = c(-2, 2), col = "blue")
# dev.off()

## w/o batch effect
fit.parent <- glmFit(dge.data.parent, design.parent)

lrt.young <- glmLRT(fit.parent, contrast = c(0,0,0,0,1,0,0,0,0,-1))
topTags(lrt.young)
summary(de.young <- decideTestsDGE(lrt.young, p=0.05))
DEgene.young <- topTags(lrt.young,n = Inf)$table[topTags(lrt.young,n = Inf)$table$FDR<0.05,]
# -1  4553
# 0  51999
# 1   4423

lrt.flowering <- glmLRT(fit.parent, contrast = c(0,0,1,0,0,0,0,-1,0,0))
topTags(lrt.flowering)
summary(de.flowering <- decideTestsDGE(lrt.flowering, p=0.05))
DEgene.flowering <- topTags(lrt.flowering,n = Inf)$table[topTags(lrt.flowering,n = Inf)$table$FDR<0.05,]
# -1  2877
# 0  54894
# 1   3204

lrt.early.silique <- glmLRT(fit.parent, contrast = c(0,1,0,0,0,0,-1,0,0,0))
topTags(lrt.early.silique)
summary(de.early.silique <- decideTestsDGE(lrt.early.silique, p=0.05))
DEgene.early.silique <- topTags(lrt.early.silique,n = Inf)$table[topTags(lrt.early.silique,n = Inf)$table$FDR<0.05,]
# -1  3200
# 0  54280
# 1   3495

lrt.late.silique <- glmLRT(fit.parent, contrast = c(0,0,0,1,0,0,0,0,-1,0))
topTags(lrt.late.silique)
summary(de.late.silique <- decideTestsDGE(lrt.late.silique, p=0.05))
DEgene.late.silique <- topTags(lrt.late.silique,n = Inf)$table[topTags(lrt.late.silique,n = Inf)$table$FDR<0.05,]
# -1  3745
# 0  53562
# 1   3668

lrt.bolting <- glmLRT(fit.parent, contrast = c(1,0,0,0,0,-1,0,0,0,0))
topTags(lrt.bolting)
summary(de.bolting <- decideTestsDGE(lrt.bolting, p=0.05))
DEgene.bolting <- topTags(lrt.bolting,n = Inf)$table[topTags(lrt.bolting,n = Inf)$table$FDR<0.05,]
# -1  2794
# 0  56327
# 1   1854

### barplot for DEGs number 
### reformat for ggplot barplot 
DEGs_number_between_gt.parent <- data.frame(young = as.data.frame(summary(de.young <- decideTestsDGE(lrt.young, p=0.05)))$Freq,
                                            bolting = as.data.frame(summary(de.bolting <- decideTestsDGE(lrt.bolting, p=0.05)))$Freq,
                                     flowering = as.data.frame(summary(de.flowering <- decideTestsDGE(lrt.flowering, p=0.05)))$Freq, 
                                     early.silique = as.data.frame(summary(de.early.silique <- decideTestsDGE(lrt.early.silique, p=0.05)))$Freq,
                                     late.silique = as.data.frame(summary(de.late.silique <- decideTestsDGE(lrt.late.silique, p=0.05)))$Freq)

rownames(DEGs_number_between_gt.parent) <- c("down", "no", "up")
DEGs_number_between_gt.parent <- DEGs_number_between_gt.parent[c("down", "up"),]
DEGs_number_between_gt.parent

library(reshape2)
DEGs_number_between_gt.melt.parent <- melt(DEGs_number_between_gt.parent)
DEGs_number_between_gt.melt.parent$DE <- rep(c("down", "up"), 5)
colnames(DEGs_number_between_gt.melt.parent) <- c("tissue", "number", "DE")
DEGs_number_between_gt.melt.parent

# reorder: up 1st down 2nd 
DEGs_number_between_gt.melt.parent$DE <- factor(DEGs_number_between_gt.melt.parent$DE, levels = c("up", "down"))

DEGs_number_between_gt.melt.parent <- DEGs_number_between_gt.melt.parent[order(DEGs_number_between_gt.melt.parent$DE),]
DEGs_number_between_gt.melt.parent

### making ggplot for DEGs 
p.DEGs_number_between_gt.parent <- ggplot(data = DEGs_number_between_gt.melt.parent)
p.DEGs_number_between_gt.parent <- p.DEGs_number_between_gt.parent + geom_bar(mapping = aes(fill=DE, x = factor(DE), y = number), stat = "identity")
p.DEGs_number_between_gt.parent <- p.DEGs_number_between_gt.parent + facet_grid(~tissue)
p.DEGs_number_between_gt.parent <- p.DEGs_number_between_gt.parent + labs(y = "number of differentially expressed genes", x = "")

p.DEGs_number_between_gt.parent
ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/DEGs_number_between_gt.parent.png", width = 11, height = 8)

## between developmental stages 
# Da-Ae 
lrt.young.vs.bolting.ae <- glmLRT(fit.parent, contrast = c(1,0,0,0,-1,0,0,0,0,0))
topTags(lrt.young.vs.bolting.ae)
summary(de.young.vs.bolting.ae <- decideTestsDGE(lrt.young.vs.bolting.ae, p=0.05))
DEgene.young.vs.bolting.ae <- topTags(lrt.young.vs.bolting.ae,n = Inf)$table[topTags(lrt.young.vs.bolting.ae,n = Inf)$table$FDR<0.05,]
# -1   442
# 0  59872
# 1    661

lrt.bolting.vs.flowering.ae <- glmLRT(fit.parent, contrast = c(-1,0,1,0,0,0,0,0,0,0))
topTags(lrt.bolting.vs.flowering.ae)
summary(de.bolting.vs.flowering.ae <- decideTestsDGE(lrt.bolting.vs.flowering.ae, p=0.05))
DEgene.bolting.vs.flowering.ae <- topTags(lrt.bolting.vs.flowering.ae,n = Inf)$table[topTags(lrt.bolting.vs.flowering.ae,n = Inf)$table$FDR<0.05,]
# -1   736
# 0  57389
# 1   2850

lrt.flowering.vs.early.silique.ae <- glmLRT(fit.parent, contrast = c(0,1,-1,0,0,0,0,0,0,0))
topTags(lrt.flowering.vs.early.silique.ae)
summary(de.flowering.vs.early.silique.ae <- decideTestsDGE(lrt.flowering.vs.early.silique.ae, p=0.05))
DEgene.flowering.vs.early.silique.ae <- topTags(lrt.flowering.vs.early.silique.ae,n = Inf)$table[topTags(lrt.flowering.vs.early.silique.ae,n = Inf)$table$FDR<0.05,]
# -1  3811
# 0  56054
# 1   1110

lrt.early.vs.late.silique.ae <- glmLRT(fit.parent, contrast = c(0,-1,0,1,0,0,0,0,0,0))
topTags(lrt.early.vs.late.silique.ae)
summary(de.early.vs.late.silique.ae <- decideTestsDGE(lrt.early.vs.late.silique.ae, p=0.05))
DEgene.early.vs.late.silique.ae <- topTags(lrt.early.vs.late.silique.ae,n = Inf)$table[topTags(lrt.early.vs.late.silique.ae,n = Inf)$table$FDR<0.05,]
# -1   285
# 0  60430
# 1    260

### ggplot 
### reformat for ggplot barplot 
DEGs_number_between_tissue.ae <- data.frame("bolting-vs-young" = as.data.frame(summary(de.young.vs.bolting.ae <- decideTestsDGE(lrt.young.vs.bolting.ae, p=0.05)))$Freq, 
                                            "flowering-vs-bolting" = as.data.frame(summary(de.bolting.vs.flowering.ae <- decideTestsDGE(lrt.bolting.vs.flowering.ae, p=0.05)))$Freq,
                                     "early.slique-vs-flowering"= as.data.frame(summary(de.flowering.vs.early.silique.ae <- decideTestsDGE(lrt.flowering.vs.early.silique.ae, p=0.05)))$Freq, 
                                     "late.silique-vs-early.silique" = as.data.frame(summary(de.early.vs.late.silique.ae <- decideTestsDGE(lrt.early.vs.late.silique.ae, p=0.05)))$Freq)

rownames(DEGs_number_between_tissue.ae) <- c("down", "no", "up")
DEGs_number_between_tissue.ae <- DEGs_number_between_tissue.ae[c("down", "up"),]
DEGs_number_between_tissue.ae

DEGs_number_between_tissue.ae.melt.parent <- melt(DEGs_number_between_tissue.ae)
DEGs_number_between_tissue.ae.melt.parent$DE <- rep(c("down", "up"), 4)
colnames(DEGs_number_between_tissue.ae.melt.parent) <- c("tissue", "number", "DE")
DEGs_number_between_tissue.ae.melt.parent

# reorder: up 1st down 2nd 
DEGs_number_between_tissue.ae.melt.parent$DE <- factor(DEGs_number_between_tissue.ae.melt.parent$DE, levels = c("up", "down"))

DEGs_number_between_tissue.ae.melt.parent <- DEGs_number_between_tissue.ae.melt.parent[order(DEGs_number_between_tissue.ae.melt.parent$DE),]
DEGs_number_between_tissue.ae.melt.parent

### making ggplot for DEGs 
p.DEGs_number_between_tissue.ae.parent <- ggplot(data = DEGs_number_between_tissue.ae.melt.parent)
p.DEGs_number_between_tissue.ae.parent <- p.DEGs_number_between_tissue.ae.parent + geom_bar(mapping = aes(fill=DE, x = factor(DE), y = number), stat = "identity")
p.DEGs_number_between_tissue.ae.parent <- p.DEGs_number_between_tissue.ae.parent + facet_grid(~tissue)
p.DEGs_number_between_tissue.ae.parent <- p.DEGs_number_between_tissue.ae.parent + labs(y = "number of differentially expressed genes", x = "")

p.DEGs_number_between_tissue.ae.parent
ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/DEGs_number_between_tissue.ae.parent.png", width = 11, height = 8)

### Da-Ol-1 
lrt.young.vs.bolting.ol <- glmLRT(fit.parent, contrast = c(0,0,0,0,0,1,0,0,0,-1))
topTags(lrt.young.vs.bolting.ol)
summary(de.young.vs.bolting.ol <- decideTestsDGE(lrt.young.vs.bolting.ol, p=0.05))
DEgene.young.vs.bolting.ol <- topTags(lrt.young.vs.bolting.ol,n = Inf)$table[topTags(lrt.young.vs.bolting.ol,n = Inf)$table$FDR<0.05,]
# -1  2683
# 0  54969
# 1   3323

lrt.bolting.vs.flowering.ol <- glmLRT(fit.parent, contrast = c(0,0,0,0,0,-1,0,1,0,0))
topTags(lrt.bolting.vs.flowering.ol)
summary(de.bolting.vs.flowering.ol <- decideTestsDGE(lrt.bolting.vs.flowering.ol, p=0.05))
DEgene.bolting.vs.flowering.ol <- topTags(lrt.bolting.vs.flowering.ol,n = Inf)$table[topTags(lrt.bolting.vs.flowering.ol,n = Inf)$table$FDR<0.05,]
# -1   914
# 0  58313
# 1   1748

lrt.flowering.vs.early.silique.ol <- glmLRT(fit.parent, contrast = c(0,0,0,0,0,0,1,-1,0,0))
topTags(lrt.flowering.vs.early.silique.ol)
summary(de.flowering.vs.early.silique.ol <- decideTestsDGE(lrt.flowering.vs.early.silique.ol, p=0.05))
DEgene.flowering.vs.early.silique.ol <- topTags(lrt.flowering.vs.early.silique.ol,n = Inf)$table[topTags(lrt.flowering.vs.early.silique.ol,n = Inf)$table$FDR<0.05,]
# -1  4845
# 0  54087
# 1   2043

lrt.early.vs.late.silique.ol <- glmLRT(fit.parent, contrast = c(0,0,0,0,0,0,-1,0,1,0))
topTags(lrt.early.vs.late.silique.ol)
summary(de.early.vs.late.silique.ol <- decideTestsDGE(lrt.early.vs.late.silique.ol, p=0.05))
DEgene.early.vs.late.silique.ol <- topTags(lrt.early.vs.late.silique.ol,n = Inf)$table[topTags(lrt.early.vs.late.silique.ol,n = Inf)$table$FDR<0.05,]
# -1  1042
# 0  58199
# 1   1734

### ggplot 
### reformat for ggplot barplot 
DEGs_number_between_tissue.ol <- data.frame("bolting-vs-young" = as.data.frame(summary(de.young.vs.bolting.ol <- decideTestsDGE(lrt.young.vs.bolting.ol, p=0.05)))$Freq, 
                                            "flowering-vs-bolting" = as.data.frame(summary(de.bolting.vs.flowering.ol <- decideTestsDGE(lrt.bolting.vs.flowering.ol, p=0.05)))$Freq,
                                     "early.slique-vs-flowering"= as.data.frame(summary(de.flowering.vs.early.silique.ol <- decideTestsDGE(lrt.flowering.vs.early.silique.ol, p=0.05)))$Freq, 
                                     "late.silique-vs-early.silique" = as.data.frame(summary(de.early.vs.late.silique.ol <- decideTestsDGE(lrt.early.vs.late.silique.ol, p=0.05)))$Freq)

rownames(DEGs_number_between_tissue.ol) <- c("down", "no", "up")
DEGs_number_between_tissue.ol <- DEGs_number_between_tissue.ol[c("down", "up"),]
DEGs_number_between_tissue.ol

DEGs_number_between_tissue.ol.melt.parent <- melt(DEGs_number_between_tissue.ol)
DEGs_number_between_tissue.ol.melt.parent$DE <- rep(c("down", "up"), 4)
colnames(DEGs_number_between_tissue.ol.melt.parent) <- c("tissue", "number", "DE")
DEGs_number_between_tissue.ol.melt.parent

# reorder: up 1st down 2nd 
DEGs_number_between_tissue.ol.melt.parent$DE <- factor(DEGs_number_between_tissue.ol.melt.parent$DE, levels = c("up", "down"))

DEGs_number_between_tissue.ol.melt.parent <- DEGs_number_between_tissue.ol.melt.parent[order(DEGs_number_between_tissue.ol.melt.parent$DE),]
DEGs_number_between_tissue.ol.melt.parent

### making ggplot for DEGs 
p.DEGs_number_between_tissue.ol.parent <- ggplot(data = DEGs_number_between_tissue.ol.melt.parent)
p.DEGs_number_between_tissue.ol.parent <- p.DEGs_number_between_tissue.ol.parent + geom_bar(mapping = aes(fill=DE, x = factor(DE), y = number), stat = "identity")
p.DEGs_number_between_tissue.ol.parent <- p.DEGs_number_between_tissue.ol.parent + facet_grid(~tissue)
p.DEGs_number_between_tissue.ol.parent <- p.DEGs_number_between_tissue.ol.parent + labs(y = "number of differentially expressed genes", x = "")

p.DEGs_number_between_tissue.ol.parent
ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/DEGs_number_between_tissue.ol.parent.png", width = 11, height = 8) 


### plot smear 
# young
detags.young.1 <- rownames(lrt.young)[as.logical(de.young)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.young.1.png")
plotSmear(lrt.young, de.tags=detags.young.1, main="young")
abline(h = c(-2, 2), col = "blue")
dev.off()

# bolting 
detags.bolting.1 <- rownames(lrt.bolting)[as.logical(de.bolting)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.bolting.1.png")
plotSmear(lrt.bolting, de.tags=detags.bolting.1, main="bolting")
abline(h = c(-2, 2), col = "blue")
dev.off()

# early silique 
detags.early.silique.1 <- rownames(lrt.early.silique)[as.logical(de.early.silique)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.early.silique.1.png")
plotSmear(lrt.early.silique, de.tags=detags.early.silique.1, main="early silique")
abline(h = c(-2, 2), col = "blue")
dev.off()

# late silique 
detags.late.silique.1 <- rownames(lrt.late.silique)[as.logical(de.late.silique)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.late.silique.1.png")
plotSmear(lrt.late.silique, de.tags=detags.late.silique.1, main="late silique")
abline(h = c(-2, 2), col = "blue")
dev.off()

# flowering 
detags.flowering.1 <- rownames(lrt.flowering)[as.logical(de.flowering)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.flowering.1.png")
plotSmear(lrt.flowering, de.tags=detags.flowering.1, main="flowering")
abline(h = c(-2, 2), col = "blue")
dev.off()

# plot histogram 
### 

```

# explore the silique problem, much fewer genes were differentially expressed between early & late silique in Da-Ae than in Da-Ol-1, why? (actually De-Ae has fewer DEGs between all stages, need to find out why later) 
```{r}
# calcuate overlaps & unique genes 
dim(DEgene.young.vs.bolting.ol) # 6006    5 
dim(DEgene.young.vs.bolting.ae) # 1103    5
summary(rownames(DEgene.young.vs.bolting.ol) %in% rownames(DEgene.young.vs.bolting.ae)) 
# Mode   FALSE    TRUE    NA's 
# logical    5556     450       0 
dim(DEgene.bolting.vs.flowering.ol) # 2662    5 
dim(DEgene.bolting.vs.flowering.ae) # 3586    5 
summary(rownames(DEgene.bolting.vs.flowering.ol) %in% rownames(DEgene.bolting.vs.flowering.ae)) 
#    Mode   FALSE    TRUE    NA's 
# logical    1528    1134       0
dim(DEgene.flowering.vs.early.silique.ol) # 6888    5 
dim(DEgene.flowering.vs.early.silique.ae) # 4921    5 
summary(rownames(DEgene.flowering.vs.early.silique.ol) %in% rownames(DEgene.flowering.vs.early.silique.ae)) 
#   Mode   FALSE    TRUE    NA's 
# logical    3200    3688       0 
dim(DEgene.early.vs.late.silique.ol) # 2776 5 
dim(DEgene.early.vs.late.silique.ae) # 545 5 
summary(rownames(DEgene.early.vs.late.silique.ol) %in% rownames(DEgene.early.vs.late.silique.ae)) 
#    Mode   FALSE    TRUE    NA's 
# logical    2436     340       0 

# draw venn diagram 
# install.packages("VennDiagram")
require(VennDiagram)

# young vs bolting
grid.newpage()
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/young.vs.bolting.png")
venn.plot <- draw.pairwise.venn(6006, 1103, 450, # 1st number for group A, 2nd for group B, 3rd for overlaps 
                                c("Da-Ol-1", "Da-Ae"),
                                fill = c("lightblue","pink"),
                                lty          = "blank",
                                cex = 2,
                                cat.cex = 1.5,
                                cat.pos = 9)
grid.draw(venn.plot)
dev.off()

# bolting vs flowering 
grid.newpage()
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/bolting.vs.flowering.png")
venn.plot <- draw.pairwise.venn(2662, 3586, 1134, # 1st number for group A, 2nd for group B, 3rd for overlaps 
                                c("Da-Ol-1", "Da-Ae"),
                                fill = c("lightblue","pink"),
                                lty          = "blank",
                                cex = 2,
                                cat.cex = 1.5,
                                cat.pos = 9)
grid.draw(venn.plot)
dev.off()

# flowering vs early silique 
grid.newpage()
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/flowering.vs.early.silique.png")
venn.plot <- draw.pairwise.venn(6888, 4921, 3688, # 1st number for group A, 2nd for group B, 3rd for overlaps 
                                c("Da-Ol-1", "Da-Ae"),
                                fill = c("lightblue","pink"),
                                lty          = "blank",
                                cex = 2,
                                cat.cex = 1.5,
                                cat.pos = 9)
grid.draw(venn.plot)
dev.off()

# early vs late silique 
grid.newpage()
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/early.vs.late.silique.png")
venn.plot <- draw.pairwise.venn(2776, 545, 340, # 1st number for group A, 2nd for group B, 3rd for overlaps 
                                c("Da-Ol-1", "Da-Ae"),
                                fill = c("lightblue","pink"),
                                lty          = "blank",
                                cex = 2,
                                cat.cex = 1.5,
                                cat.pos = 9)
grid.draw(venn.plot)
dev.off()

# explore the uniquely differentially expressed genes in Da-Ol-1, what are their expression patterns in Da-Ae? 
# 1) find genes that are only differentially expressed in Da-Ol-1 
index <- which(!(rownames(DEgene.early.vs.late.silique.ol) %in% rownames(DEgene.early.vs.late.silique.ae)))

# 2) extract the expression values of these genes from early, late silique of Da-Ae & Da-Ol-1 
uniq.gene.ol <- parent.read.count.one.small[index,]
uniq.gene.ol <- uniq.gene.ol[,c("Da-Ae_early-silique_3", "Da-Ol-1_early-silique_3", "Da-Ol-1_late-silique_3", "Da-Ae_late-silique_3", "Da-Ae_early-silique_1", "Da-Ae_early-silique_2", "Da-Ae_late-silique_1", "Da-Ae_late-silique_2", "Da-Ol-1_early-silique_1", "Da-Ol-1_early-silique_2", "Da-Ol-1_late-silique_1", "Da-Ol-1_late-silique_2")]

dim(uniq.gene.ol) # 2436   12 

# 3) draw heatmap of their expression values... I don't think this can explain the result 
# library(gplots)
# hr <- hclust(dist(uniq.gene.ol))
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/heatmap.png")
# heatmap.2(as.matrix(uniq.gene.ol), Rowv = as.dendrogram(hr), scale = "row", density.info="none", trace="none", keysize = 0.5)
# dev.off()
# 
# # 4) make figure of their mean expression values, may not be useful 
# DEgene.early.vs.late.silique.ae.all <- topTags(lrt.early.vs.late.silique.ae,n = Inf)$table[topTags(lrt.early.vs.late.silique.ae,n = Inf)$table$FDR<=1,]
# 
# uniq.gene.ol.ae <- (cbind((DEgene.early.vs.late.silique.ae.all[index,]), (DEgene.early.vs.late.silique.ol[index,])))[,c(1,5,6,10)]
# colnames(uniq.gene.ol.ae) <- c("logFC_ae","FDR_ae","logFC_ol","FDR_ol")
# colnames(uniq.gene.ol.ae)
# 
# hist(uniq.gene.ol.ae$logFC_ae)
# hist(uniq.gene.ol.ae$logFC_ol)
# hist(uniq.gene.ol.ae$FDR_ae)
# hist(uniq.gene.ol.ae$FDR_ol)
# 
# # how many are with the same direction and how many with diff directions? 
# sum(uniq.gene.ol.ae$logFC_ae>0 & uniq.gene.ol.ae$logFC_ol>0) 
# sum(uniq.gene.ol.ae$logFC_ae<0 & uniq.gene.ol.ae$logFC_ol<0)
# dim(uniq.gene.ol.ae)

# 687+506 are with same directions while 2436-687-506 are with diff directions. 
```

# new design to see how many fatty acid genes' expression are differentially expressed between genotypes and during developmental stages  
```{r}
# assign group 
ftable(parent.read.count.one.sample,row.vars=c("stage"),col.vars=c("batch","genotype"))

# normalize 
dge.new <- DGEList(counts=parent.read.count.one.small, group=parent.read.count.one.sample$group)
dim(dge.new) # [1] 30876    28 
dge.new <- calcNormFactors(dge.new, method = "TMM") 
dge.new$sample
# design matrix 
parent.read.count.one.sample$genotype <- as.factor(parent.read.count.one.sample$genotype)
parent.read.count.one.sample$tissue <- as.factor(parent.read.count.one.sample$stage)
parent.read.count.one.sample$genotype <- relevel(parent.read.count.one.sample$genotype,ref="Da-Ol-1")
parent.read.count.one.sample$tissue <- relevel(parent.read.count.one.sample$tissue,ref="Young")

design.new <- model.matrix(~tissue*genotype,data = parent.read.count.one.sample) 
colnames(design.new)

# calculate dispersion
dge.new <- estimateGLMCommonDisp(dge.new, design.new,verbose = TRUE) # Disp = 0.25646 , BCV = 0.5064 
dge.new <- estimateGLMTrendedDisp(dge.new,design.new)
dge.new <- estimateGLMTagwiseDisp(dge.new,design.new)
# plotBCV(dge.new)

# use interaction as coefficient and see how many fatty acid genes are important for interaction term... 
fit.new <- glmFit(dge.new, design.new)
lrt.new.interaction <- glmLRT(fit.new,coef = c("tissuebolting:genotypeDa-Ae", "tissueearly-silique:genotypeDa-Ae", "tissueflowering:genotypeDa-Ae", "tissuelate-silique:genotypeDa-Ae"))
topTags(lrt.new.interaction)
DEgene.new.interaction <- topTags(lrt.new.interaction,n = Inf)$table[topTags(lrt.new.interaction,n = Inf)$table$FDR<0.05,]
nrow(DEgene.new.interaction) # number of genes with interaction effect in any tissue types 

dim(fatty.acid.genes) # 207 fatty acid genes expressed in my data 

index_2 <- which(rownames(fatty.acid.genes) %in% rownames(DEgene.new.interaction)) 
rownames(fatty.acid.genes[index_2,])

# get expression plot for these several fatty acid genes 
# fatty.acid.genes.interaction <- as.data.frame(vstMat.parent[c("BnaA03g41660D","BnaA05g06830D","BnaC07g32730D"),])

fatty.acid.genes.interaction <- as.data.frame(vstMat.parent[c(rownames(fatty.acid.genes[index_2,])),])
fatty.acid.genes.interaction
dim(fatty.acid.genes.interaction)

## add gene annotation to ID 
### annotation was obtained by blast sequences against Arabidopsis transcript database and the top one was used for annotation. 

fatty.acid.genes.annotation <- read.csv("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/fatty_acid_gene_name.csv", header = F)

fatty.acid.genes.annotation
rownames(fatty.acid.genes.annotation) <- fatty.acid.genes.annotation$V1
fatty.acid.genes.interaction <- merge(fatty.acid.genes.interaction, fatty.acid.genes.annotation, by="row.names")
dim(fatty.acid.genes.interaction)
fatty.acid.genes.interaction

write.csv(fatty.acid.genes.interaction.ID, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/fatty.acid.genes.interaction.ID.txt")

# reformat for ggplot 
rownames(fatty.acid.genes.interaction) <- paste(fatty.acid.genes.interaction$Row.names, fatty.acid.genes.interaction$V2, sep = "-")
head(fatty.acid.genes.interaction)
fatty.acid.genes.interaction <- fatty.acid.genes.interaction[,-c(1,29:31)]
dim(fatty.acid.genes.interaction)
head(fatty.acid.genes.interaction)

fatty.acid.genes.interaction$geneID <- rownames(fatty.acid.genes.interaction)
fatty.acid.genes.interaction.melt <- melt(fatty.acid.genes.interaction)
fatty.acid.genes.interaction.melt$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",fatty.acid.genes.interaction.melt$variable)
fatty.acid.genes.interaction.melt$tissue <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",fatty.acid.genes.interaction.melt$variable)
fatty.acid.genes.interaction.melt$rep <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\5",fatty.acid.genes.interaction.melt$variable)
fatty.acid.genes.interaction.melt$group <- paste(fatty.acid.genes.interaction.melt$gt, fatty.acid.genes.interaction.melt$tissue, fatty.acid.genes.interaction.melt$geneID, sep = "_")
head(fatty.acid.genes.interaction.melt)
tail(fatty.acid.genes.interaction.melt)
dim(fatty.acid.genes.interaction.melt)

# to do: reshape, calculate & make ggplot ... 
fatty.acid.genes.interaction.melt.reshape <- reshape(fatty.acid.genes.interaction.melt[,c("value", "rep", "group")], idvar = "group", direction = "wide", timevar = "rep")
dim(fatty.acid.genes.interaction.melt.reshape)
fatty.acid.genes.interaction.melt.reshape

fatty.acid.genes.interaction.melt.reshape$mean <- apply(fatty.acid.genes.interaction.melt.reshape[,c(2:4)], 1, mean, na.rm=TRUE)
fatty.acid.genes.interaction.melt.reshape$min <- apply(fatty.acid.genes.interaction.melt.reshape[,c(2:4)], 1, max, na.rm=T)
fatty.acid.genes.interaction.melt.reshape$max <- apply(fatty.acid.genes.interaction.melt.reshape[,c(2:4)], 1, min, na.rm=T)
fatty.acid.genes.interaction.melt.reshape
fatty.acid.genes.interaction.melt.reshape$gt <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\1",fatty.acid.genes.interaction.melt.reshape$group)
fatty.acid.genes.interaction.melt.reshape$tissue <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\3",fatty.acid.genes.interaction.melt.reshape$group)
fatty.acid.genes.interaction.melt.reshape$geneID <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\5",fatty.acid.genes.interaction.melt.reshape$group)
fatty.acid.genes.interaction.melt.reshape 

# order data to specific orders: young, bolting, flowering, early-silique, and late-silique
fatty.acid.genes.interaction.melt.reshape$tissue <- factor(fatty.acid.genes.interaction.melt.reshape$tissue, levels = c("Young","bolting","flowering","early-silique","late-silique"))

fatty.acid.genes.interaction.melt.reshape <- fatty.acid.genes.interaction.melt.reshape[order(fatty.acid.genes.interaction.melt.reshape$tissue),]
fatty.acid.genes.interaction.melt.reshape

p.fatty.acid.interaction <- ggplot(data = fatty.acid.genes.interaction.melt.reshape)
p.fatty.acid.interaction <- p.fatty.acid.interaction + geom_line(aes(x = factor(tissue), y = mean,group=gt, color=gt))
p.fatty.acid.interaction <- p.fatty.acid.interaction + facet_grid(~geneID~gt)
p.fatty.acid.interaction <- p.fatty.acid.interaction + geom_errorbar(mapping=aes(x=tissue,ymin=min,ymax=max, width=0.25))
p.fatty.acid.interaction <- p.fatty.acid.interaction + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p.fatty.acid.interaction <- p.fatty.acid.interaction + labs(y = "mean expression value", x="tissue")

p.fatty.acid.interaction

# based on the figure, three genes were identified to have huge interaction effect, extract these three genes and make figure only for them --> modify the code  

ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/fatty.acid.interaction.png", width = 8, height = 20)

```

# get only 3 genes 
```{r}

fatty.acid.genes.three <- as.data.frame(vstMat.parent[c("BnaA05g06830D","BnaA10g09300D",  "BnaC09g50630D"),])
fatty.acid.genes.three
dim(fatty.acid.genes.three)

## add gene annotation to ID 
### annotation was obtained by blast sequences against Arabidopsis transcript database and the top one was used for annotation. 

fatty.acid.genes.annotation <- read.csv("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/fatty_acid_gene_name.csv", header = F)

fatty.acid.genes.annotation
rownames(fatty.acid.genes.annotation) <- fatty.acid.genes.annotation$V1
fatty.acid.genes.three <- merge(fatty.acid.genes.three, fatty.acid.genes.annotation, by="row.names")
dim(fatty.acid.genes.three)
fatty.acid.genes.three

write.csv(fatty.acid.genes.three, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/fatty.acid.genes.three.ID.txt")

# reformat for ggplot 
rownames(fatty.acid.genes.three) <- paste(fatty.acid.genes.three$Row.names, fatty.acid.genes.three$V2, sep = "-")
head(fatty.acid.genes.three)
fatty.acid.genes.three <- fatty.acid.genes.three[,-c(1,29:31)]
dim(fatty.acid.genes.three)
head(fatty.acid.genes.three)

fatty.acid.genes.three$geneID <- rownames(fatty.acid.genes.three)
fatty.acid.genes.three.melt <- melt(fatty.acid.genes.three)
fatty.acid.genes.three.melt$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",fatty.acid.genes.three.melt$variable)
fatty.acid.genes.three.melt$tissue <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",fatty.acid.genes.three.melt$variable)
fatty.acid.genes.three.melt$rep <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\5",fatty.acid.genes.three.melt$variable)
fatty.acid.genes.three.melt$group <- paste(fatty.acid.genes.three.melt$gt, fatty.acid.genes.three.melt$tissue, fatty.acid.genes.three.melt$geneID, sep = "_")
head(fatty.acid.genes.three.melt)
tail(fatty.acid.genes.three.melt)
dim(fatty.acid.genes.three.melt)

# to do: reshape, calculate & make ggplot ... 
fatty.acid.genes.three.melt.reshape <- reshape(fatty.acid.genes.three.melt[,c("value", "rep", "group")], idvar = "group", direction = "wide", timevar = "rep")
dim(fatty.acid.genes.three.melt.reshape)
fatty.acid.genes.three.melt.reshape

fatty.acid.genes.three.melt.reshape$mean <- apply(fatty.acid.genes.three.melt.reshape[,c(2:4)], 1, mean, na.rm=TRUE)
fatty.acid.genes.three.melt.reshape$min <- apply(fatty.acid.genes.three.melt.reshape[,c(2:4)], 1, max, na.rm=T)
fatty.acid.genes.three.melt.reshape$max <- apply(fatty.acid.genes.three.melt.reshape[,c(2:4)], 1, min, na.rm=T)
fatty.acid.genes.three.melt.reshape
fatty.acid.genes.three.melt.reshape$gt <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\1",fatty.acid.genes.three.melt.reshape$group)
fatty.acid.genes.three.melt.reshape$tissue <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\3",fatty.acid.genes.three.melt.reshape$group)
fatty.acid.genes.three.melt.reshape$geneID <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\5",fatty.acid.genes.three.melt.reshape$group)
fatty.acid.genes.three.melt.reshape 

# order data to specific orders: young, bolting, flowering, early-silique, and late-silique
fatty.acid.genes.three.melt.reshape$tissue <- factor(fatty.acid.genes.three.melt.reshape$tissue, levels = c("Young","bolting","flowering","early-silique","late-silique"))

fatty.acid.genes.three.melt.reshape <- fatty.acid.genes.three.melt.reshape[order(fatty.acid.genes.three.melt.reshape$tissue),]
fatty.acid.genes.three.melt.reshape

p.fatty.acid.three <- ggplot(data = fatty.acid.genes.three.melt.reshape)
p.fatty.acid.three <- p.fatty.acid.three + geom_line(aes(x = factor(tissue), y = mean,group=gt, color=gt))
p.fatty.acid.three <- p.fatty.acid.three + facet_grid(~geneID~gt)
p.fatty.acid.three <- p.fatty.acid.three + geom_errorbar(mapping=aes(x=tissue,ymin=min,ymax=max, width=0.25))
p.fatty.acid.three <- p.fatty.acid.three + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p.fatty.acid.three <- p.fatty.acid.three + labs(y = "mean expression value", x="tissue")

p.fatty.acid.three

# based on the figure, three genes were identified to have huge interaction effect, extract these three genes and make figure only for them --> modify the code  

ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/fatty.acid.three.png", width = 11, height = 8)
```


# GO enrichment analysis 
```{r}
# load all necessary functions 
source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")

# young 
DEgene.young.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.young))
DEgene.young.GO.ORA
write.csv(DEgene.young.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.young.GO.ORA.csv")

# bolting 
DEgene.bolting.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.bolting))
DEgene.bolting.GO.ORA
write.csv(DEgene.bolting.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.bolting.GO.ORA.csv")

# flowering 
DEgene.flowering.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.flowering))
DEgene.flowering.GO.ORA
write.csv(DEgene.flowering.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.flowering.GO.ORA.csv")

# early silique 
DEgene.early.silique.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.early.silique))
DEgene.early.silique.GO.ORA
write.csv(DEgene.early.silique.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.early.silique.GO.ORA.csv")

DEgene.late.silique.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.late.silique))
DEgene.late.silique.GO.ORA
write.csv(DEgene.late.silique.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.late.silique.GO.ORA.csv")

DEgene.silique.ae.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.early.vs.late.silique.ae))
DEgene.silique.ae.GO.ORA
write.csv(DEgene.late.silique.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.silique.ae.GO.ORA.csv")

DEgene.silique.ol.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.early.vs.late.silique.ol))
DEgene.silique.ol.GO.ORA
write.csv(DEgene.silique.ol.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.silique.ol.GO.ORA.csv")

DEgene.young.vs.bolting.ol.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.young.vs.bolting.ol))
DEgene.young.vs.bolting.ol.GO.ORA

DEgene.young.vs.bolting.ae.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.young.vs.bolting.ae))
DEgene.young.vs.bolting.ae.GO.ORA

DEgene.bolting.vs.flowering.ol.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.bolting.vs.flowering.ol))
DEgene.bolting.vs.flowering.ol.GO.ORA

DEgene.bolting.vs.flowering.ae.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.bolting.vs.flowering.ae))
DEgene.bolting.vs.flowering.ae.GO.ORA

DEgene.flowering.vs.early.silique.ol.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.flowering.vs.early.silique.ol))
DEgene.bolting.vs.flowering.ol.GO.ORA

DEgene.flowering.vs.early.silique.ae.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.flowering.vs.early.silique.ae))
DEgene.bolting.vs.flowering.ae.GO.ORA
```

### other important genes 
# Acetyl-CoA-ACP-transacetylase: GO:0004313 none 
# B-Ketoacyl-ACP-synthase: IPR014030 --> A 
# Malonyl-CoA-ACP-transacetylase: IPR016036 --> B  
# B-ketoacyl-ACP-reductase: GO:0004316 --> C 
# Enoyl-ACP-hydrase: IPR018376 --> D
# Enoyl-ACP-reductase: GO:0016631 & GO:0004319 none 
# Palmitoyl-thioesterase: IPR002472 --> E 
```{r}
# B-Ketoacyl-ACP-synthase 
A.ID <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/B-Ketoacyl-ACP-synthase", header = F)
A.ID

A <- as.data.frame(vstMat.parent[c(A.ID$V1),])
A
dim(A)

# reformat for ggplot 
head(A)

A$geneID <- rownames(A)
A.melt <- melt(A)
A.melt$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",A.melt$variable)
A.melt$tissue <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",A.melt$variable)
A.melt$rep <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\5",A.melt$variable)
A.melt$group <- paste(A.melt$gt, A.melt$tissue, A.melt$geneID, sep = "_")
head(A.melt)
tail(A.melt)
dim(A.melt)

# to do: reshape, calculate & make ggplot ... 
A.melt.reshape <- reshape(A.melt[,c("value", "rep", "group")], idvar = "group", direction = "wide", timevar = "rep")
dim(A.melt.reshape)
A.melt.reshape

A.melt.reshape$mean <- apply(A.melt.reshape[,c(2:4)], 1, mean, na.rm=TRUE)
A.melt.reshape$min <- apply(A.melt.reshape[,c(2:4)], 1, max, na.rm=T)
A.melt.reshape$max <- apply(A.melt.reshape[,c(2:4)], 1, min, na.rm=T)
A.melt.reshape
A.melt.reshape$gt <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\1",A.melt.reshape$group)
A.melt.reshape$tissue <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\3",A.melt.reshape$group)
A.melt.reshape$geneID <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\5",A.melt.reshape$group)
A.melt.reshape 

# order data to specific orders: young, bolting, flowering, early-silique, and late-silique
A.melt.reshape$tissue <- factor(A.melt.reshape$tissue, levels = c("Young","bolting","flowering","early-silique","late-silique"))

A.melt.reshape <- A.melt.reshape[order(A.melt.reshape$tissue),]
A.melt.reshape

p.A <- ggplot(data = A.melt.reshape)
p.A <- p.A + geom_line(aes(x = factor(tissue), y = mean,group=gt, color=gt))
p.A <- p.A + facet_grid(~geneID~gt)
p.A <- p.A + geom_errorbar(mapping=aes(x=tissue,ymin=min,ymax=max, width=0.25))
p.A <- p.A + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p.A <- p.A + labs(y = "mean expression value", x="tissue", title="B-Ketoacyl-ACP synthase")

p.A
ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/A.png", width = 8, height = 20)

# Malonyl-CoA-ACP-transacetylase  
B.ID <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/Malonyl-CoA-ACP-transacetylase", header = F)
B.ID

B <- as.data.frame(vstMat.parent[c(B.ID$V1),])
B
dim(B)

# reformat for ggplot 
head(B)

B$geneID <- rownames(B)
B.melt <- melt(B)
B.melt$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",B.melt$variable)
B.melt$tissue <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",B.melt$variable)
B.melt$rep <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\5",B.melt$variable)
B.melt$group <- paste(B.melt$gt, B.melt$tissue, B.melt$geneID, sep = "_")
head(B.melt)
tail(B.melt)
dim(B.melt)

# to do: reshape, calculate & make ggplot ... 
B.melt.reshape <- reshape(B.melt[,c("value", "rep", "group")], idvar = "group", direction = "wide", timevar = "rep")
dim(B.melt.reshape)
B.melt.reshape

B.melt.reshape$mean <- apply(B.melt.reshape[,c(2:4)], 1, mean, na.rm=TRUE)
B.melt.reshape$min <- apply(B.melt.reshape[,c(2:4)], 1, max, na.rm=T)
B.melt.reshape$max <- apply(B.melt.reshape[,c(2:4)], 1, min, na.rm=T)
B.melt.reshape
B.melt.reshape$gt <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\1",B.melt.reshape$group)
B.melt.reshape$tissue <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\3",B.melt.reshape$group)
B.melt.reshape$geneID <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\5",B.melt.reshape$group)
B.melt.reshape 

# order data to specific orders: young, bolting, flowering, early-silique, and late-silique
B.melt.reshape$tissue <- factor(B.melt.reshape$tissue, levels = c("Young","bolting","flowering","early-silique","late-silique"))

B.melt.reshape <- B.melt.reshape[order(B.melt.reshape$tissue),]
B.melt.reshape

p.B <- ggplot(data = B.melt.reshape)
p.B <- p.B + geom_line(aes(x = factor(tissue), y = mean,group=gt, color=gt))
p.B <- p.B + facet_grid(~geneID~gt)
p.B <- p.B + geom_errorbar(mapping=aes(x=tissue,ymin=min,ymax=max, width=0.25))
p.B <- p.B + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p.B <- p.B + labs(y = "mean expression value", x="tissue", title="Malonyl-CoA-ACP transacetylase")

p.B

ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/B.png", width = 8, height = 10)

# B-ketoacyl-ACP-reductase: GO:0004316 --> C 
C.ID <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/B-ketoacyl-ACP-reductase", header = F)
C.ID

C <- as.data.frame(vstMat.parent[c(C.ID$V1),])
C
dim(C)

# reformat for ggplot 
head(C)

C$geneID <- rownames(C)
C.melt <- melt(C)
C.melt$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",C.melt$variable)
C.melt$tissue <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",C.melt$variable)
C.melt$rep <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\5",C.melt$variable)
C.melt$group <- paste(C.melt$gt, C.melt$tissue, C.melt$geneID, sep = "_")
head(C.melt)
tail(C.melt)
dim(C.melt)

# to do: reshape, calculate & make ggplot ... 
C.melt.reshape <- reshape(C.melt[,c("value", "rep", "group")], idvar = "group", direction = "wide", timevar = "rep")
dim(C.melt.reshape)
C.melt.reshape

C.melt.reshape$mean <- apply(C.melt.reshape[,c(2:4)], 1, mean, na.rm=TRUE)
C.melt.reshape$min <- apply(C.melt.reshape[,c(2:4)], 1, max, na.rm=T)
C.melt.reshape$max <- apply(C.melt.reshape[,c(2:4)], 1, min, na.rm=T)
C.melt.reshape
C.melt.reshape$gt <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\1",C.melt.reshape$group)
C.melt.reshape$tissue <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\3",C.melt.reshape$group)
C.melt.reshape$geneID <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\5",C.melt.reshape$group)
C.melt.reshape 

# order data to specific orders: young, bolting, flowering, early-silique, and late-silique
C.melt.reshape$tissue <- factor(C.melt.reshape$tissue, levels = c("Young","bolting","flowering","early-silique","late-silique"))

C.melt.reshape <- C.melt.reshape[order(C.melt.reshape$tissue),]
C.melt.reshape

p.C <- ggplot(data = C.melt.reshape)
p.C <- p.C + geom_line(aes(x = factor(tissue), y = mean,group=gt, color=gt))
p.C <- p.C + facet_grid(~geneID~gt)
p.C <- p.C + geom_errorbar(mapping=aes(x=tissue,ymin=min,ymax=max, width=0.25))
p.C <- p.C + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p.C <- p.C + labs(y = "mean expression value", x="tissue", title="B-ketoacyl-ACP reductase")

p.C

ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/C.png", width = 8, height = 8)

# Enoyl-ACP-hydrase: IPR018376 --> D
D.ID <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/Enoyl-ACP-hydrase", header = F)
D.ID

D <- as.data.frame(vstMat.parent[c(D.ID$V1),])
D
dim(D)

# reformat for ggplot 
head(D)

D$geneID <- rownames(D)
D.melt <- melt(D)
D.melt$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",D.melt$variable)
D.melt$tissue <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",D.melt$variable)
D.melt$rep <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\5",D.melt$variable)
D.melt$group <- paste(D.melt$gt, D.melt$tissue, D.melt$geneID, sep = "_")
head(D.melt)
tail(D.melt)
dim(D.melt)

# to do: reshape, calculate & make ggplot ... 
D.melt.reshape <- reshape(D.melt[,c("value", "rep", "group")], idvar = "group", direction = "wide", timevar = "rep")
dim(D.melt.reshape)
D.melt.reshape

D.melt.reshape$mean <- apply(D.melt.reshape[,c(2:4)], 1, mean, na.rm=TRUE)
D.melt.reshape$min <- apply(D.melt.reshape[,c(2:4)], 1, max, na.rm=T)
D.melt.reshape$max <- apply(D.melt.reshape[,c(2:4)], 1, min, na.rm=T)
D.melt.reshape
D.melt.reshape$gt <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\1",D.melt.reshape$group)
D.melt.reshape$tissue <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\3",D.melt.reshape$group)
D.melt.reshape$geneID <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\5",D.melt.reshape$group)
D.melt.reshape 

# order data to specific orders: young, bolting, flowering, early-silique, and late-silique
D.melt.reshape$tissue <- factor(D.melt.reshape$tissue, levels = c("Young","bolting","flowering","early-silique","late-silique"))

D.melt.reshape <- D.melt.reshape[order(D.melt.reshape$tissue),]
D.melt.reshape

p.D <- ggplot(data = D.melt.reshape)
p.D <- p.D + geom_line(aes(x = factor(tissue), y = mean,group=gt, color=gt))
p.D <- p.D + facet_grid(~geneID~gt)
p.D <- p.D + geom_errorbar(mapping=aes(x=tissue,ymin=min,ymax=max, width=0.25))
p.D <- p.D + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p.D <- p.D + labs(y = "mean expression value", x="tissue", title="Enoyl-ACP hydrase")

p.D

ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/D.png", width = 8, height = 20)

# Palmitoyl-thioesterase: IPR002472 --> E 
E.ID <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/Palmitoyl-thioesterase", header = F)
E.ID

E <- as.data.frame(vstMat.parent[c(E.ID$V1),])
E
dim(E)

# reformat for ggplot 
head(E)

E$geneID <- rownames(E)
E.melt <- melt(E)
E.melt$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",E.melt$variable)
E.melt$tissue <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",E.melt$variable)
E.melt$rep <-  gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\5",E.melt$variable)
E.melt$group <- paste(E.melt$gt, E.melt$tissue, E.melt$geneID, sep = "_")
head(E.melt)
tail(E.melt)
dim(E.melt)

# to do: reshape, calculate & make ggplot ... 
E.melt.reshape <- reshape(E.melt[,c("value", "rep", "group")], idvar = "group", direction = "wide", timevar = "rep")
dim(E.melt.reshape)
E.melt.reshape

E.melt.reshape$mean <- apply(E.melt.reshape[,c(2:4)], 1, mean, na.rm=TRUE)
E.melt.reshape$min <- apply(E.melt.reshape[,c(2:4)], 1, max, na.rm=T)
E.melt.reshape$max <- apply(E.melt.reshape[,c(2:4)], 1, min, na.rm=T)
E.melt.reshape
E.melt.reshape$gt <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\1",E.melt.reshape$group)
E.melt.reshape$tissue <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\3",E.melt.reshape$group)
E.melt.reshape$geneID <- gsub("([[:print:]]+)(_)([[:print:]]+)(_)([[:print:]]+)","\\5",E.melt.reshape$group)
E.melt.reshape 

# order data to specific orders: young, bolting, flowering, early-silique, and late-silique
E.melt.reshape$tissue <- factor(E.melt.reshape$tissue, levels = c("Young","bolting","flowering","early-silique","late-silique"))

E.melt.reshape <- E.melt.reshape[order(E.melt.reshape$tissue),]
E.melt.reshape

p.E <- ggplot(data = E.melt.reshape)
p.E <- p.E + geom_line(aes(x = factor(tissue), y = mean,group=gt, color=gt))
p.E <- p.E + facet_grid(~geneID~gt)
p.E <- p.E + geom_errorbar(mapping=aes(x=tissue,ymin=min,ymax=max, width=0.25))
p.E <- p.E + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p.E <- p.E + labs(y = "mean expression value", x="tissue", title="Palmitoyl thioesterase")

p.E

ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/E.png", width = 8, height = 20) 

```





