---
title: "KIAT_SNP"
author: "Ruijuan Li"
date: "11/2/2016"
output: 
  html_document: 
    keep_md: yes
---
# using freebayes to call SNP 
```{r}
# call as diploid 
vcf.Ae.Ol.no.bolting.unique.diploid <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/data/freebayes/as_diploid/Ae_Ol_unique_diploid_modified.vcf",as.is=T,na.strings = ".") 

dim(vcf.Ae.Ol.no.bolting.unique.diploid) # 2102031      11  

vcf.header.Ae.Ol.unique.diploid <- system("grep '#C' ~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/data/freebayes/as_diploid/Ae_Ol_unique_diploid_modified.vcf",intern = TRUE) 
vcf.header.Ae.Ol.unique.diploid
vcf.header.Ae.Ol.unique.diploid <- sub("#","",vcf.header.Ae.Ol.unique.diploid) #get rid of the pound sign
vcf.header.Ae.Ol.unique.diploid

vcf.header.Ae.Ol.unique.diploid <- unlist(strsplit(vcf.header.Ae.Ol.unique.diploid,split="\t"))
vcf.header.Ae.Ol.unique.diploid
colnames(vcf.Ae.Ol.no.bolting.unique.diploid) <- vcf.header.Ae.Ol.unique.diploid
head(vcf.Ae.Ol.no.bolting.unique.diploid) # why no print out??? 

system("grep '##INFO' ~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/data/freebayes/as_diploid/Ae_Ol_unique_diploid_modified.vcf") 
system("grep '##FORMAT' ~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/data/freebayes/as_diploid/Ae_Ol_unique_diploid_modified.vcf") 

# Before splitting add NAs to blank cells
vcf.Ae.Ol.no.bolting.unique.diploid$Ae

vcf.Ae.Ol.no.bolting.unique.diploid$Ae[is.na(vcf.Ae.Ol.no.bolting.unique.diploid$Ae)] <- "NA:NA:NA:NA:NA:NA:NA"

Ae.tmp.unique <- matrix(
  unlist(strsplit(vcf.Ae.Ol.no.bolting.unique.diploid$Ae,split = ":")),
  nrow=nrow(vcf.Ae.Ol.no.bolting.unique.diploid),  
  byrow=TRUE
  )

head(Ae.tmp.unique)

colnames(Ae.tmp.unique) <- paste("Ae",c("gt","tot.depth","ref.depth","ref.qual","alt.depth","alt.qual","gen.lik"),sep="_")

vcf.Ae.Ol.no.bolting.unique.diploid$Ol[is.na(vcf.Ae.Ol.no.bolting.unique.diploid$Ol)] <- "NA:NA:NA:NA:NA:NA:NA"

Ol.tmp.unique <- matrix(
  unlist(strsplit(vcf.Ae.Ol.no.bolting.unique.diploid$Ol,split = ":")),
  nrow=nrow(vcf.Ae.Ol.no.bolting.unique.diploid),
  byrow=TRUE
  )
head(Ol.tmp.unique)
colnames(Ol.tmp.unique) <- paste("Ol",c("gt","tot.depth","ref.depth","ref.qual","alt.depth","alt.qual","gen.lik"),sep="_")

vcf.Ae.Ol.no.bolting.unique.diploid <- cbind(vcf.Ae.Ol.no.bolting.unique.diploid,Ae.tmp.unique,Ol.tmp.unique,stringsAsFactors=FALSE)
summary(vcf.Ae.Ol.no.bolting.unique.diploid)
head(vcf.Ae.Ol.no.bolting.unique.diploid)

vcf.Ae.Ol.no.bolting.unique.diploid[,c("Ae_tot.depth","Ae_ref.depth","Ae_ref.qual","Ae_alt.depth","Ae_alt.qual","Ol_tot.depth","Ol_ref.depth","Ol_ref.qual","Ol_alt.depth","Ol_alt.qual")] <- 
  apply(vcf.Ae.Ol.no.bolting.unique.diploid[,c("Ae_tot.depth","Ae_ref.depth","Ae_ref.qual","Ae_alt.depth","Ae_alt.qual","Ol_tot.depth","Ol_ref.depth","Ol_ref.qual","Ol_alt.depth","Ol_alt.qual")],
        2,
        as.numeric
        )
summary(vcf.Ae.Ol.no.bolting.unique.diploid)
head(vcf.Ae.Ol.no.bolting.unique.diploid)
dim(vcf.Ae.Ol.no.bolting.unique.diploid) # 2102031      25   

######### 
### 1) filter based on QUAl score 

# make a histogram of QUAL scores 
hist(vcf.Ae.Ol.no.bolting.unique.diploid$QUAL)

length(vcf.Ae.Ol.no.bolting.unique.diploid$QUAL) # a total of 21002031 SNPs 
sum(vcf.Ae.Ol.no.bolting.unique.diploid$QUAL>20) / length(vcf.Ae.Ol.no.bolting.unique.diploid$QUAL) # 81% are above QUAL score of 20, which means less than 0.01 probability that it isn't polymorphic 

hist(vcf.Ae.Ol.no.bolting.unique.diploid[vcf.Ae.Ol.no.bolting.unique.diploid$QUAL<20,]$QUAL)
hist(vcf.Ae.Ol.no.bolting.unique.diploid[vcf.Ae.Ol.no.bolting.unique.diploid$QUAL<1,]$QUAL)

# subset the data to keep positions where the quality score is 40 or higher 
vcf.Ae.Ol.no.bolting.unique.HQ <- vcf.Ae.Ol.no.bolting.unique.diploid[vcf.Ae.Ol.no.bolting.unique.diploid$QUAL>40,]
dim(vcf.Ae.Ol.no.bolting.unique.HQ) # 1502350      25  
sum(vcf.Ae.Ol.no.bolting.unique.diploid$QUAL>40) / length(vcf.Ae.Ol.no.bolting.unique.diploid$QUAL) # 71% of SNPs were retained with QUAL > 40

# count the number 
table(vcf.Ae.Ol.no.bolting.unique.HQ$Ae_gt)
table(vcf.Ae.Ol.no.bolting.unique.HQ$Ol_gt)

head(vcf.Ae.Ol.no.bolting.unique.HQ[which(vcf.Ae.Ol.no.bolting.unique.HQ$Ae_gt=="0/1"),], 20)

tmp.unique <- ftable(vcf.Ae.Ol.no.bolting.unique.HQ[,c("Ae_gt","Ol_gt")])
head(tmp.unique)
write.table(vcf.Ae.Ol.no.bolting.unique.HQ, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/SNP_calling/output/gt_Ae_Ol.vcf") 
```

# check why there are so many heterozygous and why 3, 4, 5... 
```{r}
# 1) Julin suggested plot by mapping score, total depth, and QUAL score... 
# 2) also visualize SNPs in igv to spot the realignment problem, SNP calls from end of reads & intron-exon insertion is bad, don’t believe… 
# 3) 3, 4, and 5 visualize them in IGV…  

vcf.Ae.Ol.no.bolting.unique.HQ$Ae_gt=="1/1" 
vcf.Ae.Ol.no.bolting.unique.HQ$Ae_gt=="0/1"
vcf.Ae.Ol.no.bolting.unique.HQ$Ae_gt=="0/1" 

head(vcf.Ae.Ol.no.bolting.unique.HQ)

# plot QUAL by Ae_gt 
library(ggplot2)
p.QUAL.Ae <- ggplot(data = vcf.Ae.Ol.no.bolting.unique.HQ) 
p.QUAL.Ae <- p.QUAL.Ae + geom_histogram(aes(QUAL, fill=Ae_gt)) + scale_x_log10() 
p.QUAL.Ae <- p.QUAL.Ae + facet_wrap(~Ae_gt, ncol = 1)
p.QUAL.Ae

# plot QUAL by Ol_gt 
p.QUAL.Ol <- ggplot(data = vcf.Ae.Ol.no.bolting.unique.HQ) 
p.QUAL.Ol <- p.QUAL.Ol + geom_histogram(aes(QUAL, fill=Ol_gt)) + scale_x_log10() 
p.QUAL.Ol <- p.QUAL.Ol + facet_wrap(~Ol_gt, ncol = 1)
p.QUAL.Ol

# plot total depth by Ae_gt 
p.tot.depth.Ae <- ggplot(data = vcf.Ae.Ol.no.bolting.unique.HQ) 
p.tot.depth.Ae <- p.tot.depth.Ae + geom_histogram(aes(Ae_tot.depth+Ol_tot.depth, fill=Ae_gt)) + scale_x_log10() 
p.tot.depth.Ae <- p.tot.depth.Ae + facet_wrap(~Ae_gt, ncol = 1)
p.tot.depth.Ae

# plot total depth by Ol_gt 
p.tot.depth.Ol <- ggplot(data = vcf.Ae.Ol.no.bolting.unique.HQ) 
p.tot.depth.Ol <- p.tot.depth.Ol + geom_histogram(aes(Ae_tot.depth+Ol_tot.depth, fill=Ol_gt)) + scale_x_log10() 
p.tot.depth.Ol <- p.tot.depth.Ol + facet_wrap(~Ol_gt, ncol = 1)
p.tot.depth.Ol

# no difference in mapping score, because we used only mapping score of 255 for SNP calling. mapping score of 255 means unique mapping, check stats didn't tell me anything... 

#### export vcf file to igv for visulization 
dim(vcf.Ae.Ol.no.bolting.unique.HQ) # 1502350      25 
vcf.Ae.Ol.no.bolting.unique.HQ.export <- vcf.Ae.Ol.no.bolting.unique.HQ[,c(1:11)]

write.table(vcf.Ae.Ol.no.bolting.unique.HQ.export, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/output/vcf.Ae.Ol.no.bolting.unique.HQ.vcf", row.names = F)

# conclusion: 2,3,4,5 are caused by re-alignment problem
```

# continue explore why many 0/1 in my inbred lines (arround 30%), already filtered by QUAL 
```{r}
# to further filter the SNPs, I only retain SNPs with read depth > 3 & < 501 in both Da-Ae & Da-Ol-1 
vcf.Ae.Ol.no.bolting.unique.HQ.filtered <- vcf.Ae.Ol.no.bolting.unique.HQ[which(vcf.Ae.Ol.no.bolting.unique.HQ$Ae_tot.depth > 3 & vcf.Ae.Ol.no.bolting.unique.HQ$Ae_tot.depth < 501 & vcf.Ae.Ol.no.bolting.unique.HQ$Ol_tot.depth > 3 & vcf.Ae.Ol.no.bolting.unique.HQ$Ol_tot.depth < 501),] 

dim(vcf.Ae.Ol.no.bolting.unique.HQ.filtered) # 834016     25 

table(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$Ae_gt) 
table(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$Ol_gt)

# after filtered by read depth, high het problem still exist 

######## below, work to figure the cause... 

# 1) check the ref/alt allele ratio for the het calls, distribution plot for Ae & Ol 
# get het & homo call for Ae & Ol 
vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ae.het <- vcf.Ae.Ol.no.bolting.unique.HQ.filtered[(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$Ae_gt=="0/1"),]
vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ae.ref <- vcf.Ae.Ol.no.bolting.unique.HQ.filtered[(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$Ae_gt=="0/0"),]
vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ae.alt <- vcf.Ae.Ol.no.bolting.unique.HQ.filtered[(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$Ae_gt=="1/1"),]

table(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$Ae_gt) 
dim(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ae.het); dim(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ae.ref); dim(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ae.alt)

vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ol.het <- vcf.Ae.Ol.no.bolting.unique.HQ.filtered[(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$Ol_gt=="0/1"),]
vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ol.ref <- vcf.Ae.Ol.no.bolting.unique.HQ.filtered[(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$Ol_gt=="0/0"),]
vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ol.alt <- vcf.Ae.Ol.no.bolting.unique.HQ.filtered[(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$Ol_gt=="1/1"),]

table(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$Ol_gt) 
dim(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ol.het); dim(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ol.ref); dim(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ol.alt) 

# function to draw plot
pl.allele.ratio.Ae <- function(vcf.subset){
  pl <- ggplot(data = vcf.subset) 
  pl <- pl + geom_histogram(aes(Ae_alt.depth/Ae_ref.depth)) + scale_x_log10() 
  pl <- pl +labs(list(x="log10(alt.depth/ref.depth)")) 
  return(pl)  
} 
pl.allele.ratio.Ol <- function(vcf.subset){
  pl <- ggplot(data = vcf.subset) 
  pl <- pl + geom_histogram(aes(Ol_alt.depth/Ol_ref.depth)) + scale_x_log10() 
  pl <- pl +labs(list(x="log10(alt.depth/ref.depth)")) 
  return(pl)  
} 

pl.allele.ratio.Ae(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ae.het) 
pl.allele.ratio.Ol(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ol.het) 

## non-biallelic SNPs...oh, those are indels...(CIGAR score for them...)  
# check CIGAR score 

# 1) seperate CIGAR scores 
head(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.Ae.het)
info <- matrix(
  unlist(strsplit(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$INFO,split = ";")),
  nrow=nrow(vcf.Ae.Ol.no.bolting.unique.HQ.filtered),  
  byrow=TRUE
  )

head(info)

colnames(info) <- c("AB","ABP","AC","AF","AN","AO","CIGAR","DP","DPB","DPRA","EPP","EPPR","GTI","LEN","MEANALT",
                    "MQM","MQMR","NS","NUMALT","ODDS","PAIRED","PAIREDR","PAO","PQA","PQR","PRO","QA","QR","RO",
                    "RPL","RPP","RPPR","RPR","RUN","SAF","SAP","SAR","SRF","SRP","SRR","TYPE")

head(info)
info <- as.data.frame(info)
colnames(info)

for (i in 1:ncol(info)){
  info[,i] <- gsub("([[:print:]]+)(=)([[:print:]])","\\3",info[,i])
}

head(info)

vcf.Ae.Ol.no.bolting.unique.HQ.filtered <- cbind(vcf.Ae.Ol.no.bolting.unique.HQ.filtered,info,stringsAsFactors=FALSE)
head(vcf.Ae.Ol.no.bolting.unique.HQ.filtered)
unique(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$TYPE)

# 0/1 suggested by Julin, might be caused re-alignment (this can be checked based on CIGAR score)
# need to work on this...  
vcf.Ae.Ol.no.bolting.unique.HQ.filtered.SNP <- 
  vcf.Ae.Ol.no.bolting.unique.HQ.filtered[vcf.Ae.Ol.no.bolting.unique.HQ.filtered$TYPE=="snp",]

table(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.SNP$Ae_gt)
table(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.SNP$Ol_gt)

vcf.Ae.Ol.no.bolting.unique.HQ.filtered.other <- 
  vcf.Ae.Ol.no.bolting.unique.HQ.filtered[vcf.Ae.Ol.no.bolting.unique.HQ.filtered$TYPE!="snp",]

table(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.other$Ae_gt)
table(vcf.Ae.Ol.no.bolting.unique.HQ.filtered.other$Ol_gt)

## 
unique(vcf.Ae.Ol.no.bolting.unique.HQ.filtered$CIGAR)

unique(vcf.Ae.Ol.no.bolting.unique.HQ.filtered[vcf.Ae.Ol.no.bolting.unique.HQ.filtered$Ae_gt=="0/1",]$CIGAR)




```


```{r}
# which SNPs would be most useful for a downstream QTL analysis between Ae & Ol? 

# SNPs between Ae & ref
sum(vcf.Ae.Ol.no.bolting.unique.HQ$Ae_gt=="1/1") # 595704   
Ae_ref.unique <- vcf.Ae.Ol.no.bolting.unique.HQ[vcf.Ae.Ol.no.bolting.unique.HQ$Ae_gt=="1/1",]
# read depth > 3 and read depth < 501  
Ae_ref_after_filter.unique <- Ae_ref.unique[Ae_ref.unique$Ae_tot.depth > 3 &Ae_ref.unique$Ae_tot.depth <  501,] 
dim(Ae_ref_after_filter.unique) # 394737     25   
head(Ae_ref_after_filter.unique)

# SNPs between Ol & ref 
sum(vcf.Ae.Ol.no.bolting.unique.HQ$Ol_gt=="1/1") # 421125  
Ol_ref.unique <- vcf.Ae.Ol.no.bolting.unique.HQ[vcf.Ae.Ol.no.bolting.unique.HQ$Ol_gt=="1/1",]
# read depth > 3 and read depth < 501  
Ol_ref_after_filter.unique <- Ol_ref.unique[Ol_ref.unique$Ol_tot.depth > 3 &Ol_ref.unique$Ol_tot.depth <  501,] 
dim(Ol_ref_after_filter.unique) # 391944     25   

# SNPs between Ae & Ol 
sum(vcf.Ae.Ol.no.bolting.unique.HQ$Ae_gt=="1/1" & vcf.Ae.Ol.no.bolting.unique.HQ$Ol_gt=="0/0") +  sum(vcf.Ae.Ol.no.bolting.unique.HQ$Ae_gt=="0/0" & vcf.Ae.Ol.no.bolting.unique.HQ$Ol_gt=="1/1") # 314046 SNPs between Ae & Ol 

314046/length(vcf.Ae.Ol.no.bolting.unique.HQ$CHROM) # 21% of the total SNPs 
# subset SNPs between Ae & Ol 
vcf.Ae.Ol.no.bolting.unique.HQ.new <- vcf.Ae.Ol.no.bolting.unique.HQ[((vcf.Ae.Ol.no.bolting.unique.HQ$Ae_gt=="1/1" & vcf.Ae.Ol.no.bolting.unique.HQ$Ol_gt=="0/0") | (vcf.Ae.Ol.no.bolting.unique.HQ$Ae_gt=="0/0" & vcf.Ae.Ol.no.bolting.unique.HQ$Ol_gt=="1/1")),] 

dim(vcf.Ae.Ol.no.bolting.unique.HQ.new) # 314046     25  
# read depth > 3 and read depth < 501  

Ae_Ol_after_filter.unique  <- 
vcf.Ae.Ol.no.bolting.unique.HQ.new[(vcf.Ae.Ol.no.bolting.unique.HQ.new$Ae_tot.depth > 3 & vcf.Ae.Ol.no.bolting.unique.HQ.new$Ae_tot.depth < 501 & vcf.Ae.Ol.no.bolting.unique.HQ.new$Ol_tot.depth > 3 & vcf.Ae.Ol.no.bolting.unique.HQ.new$Ol_tot.depth < 501),]  
dim(Ae_Ol_after_filter.unique)  # 232477     25    
  
######### using IGV to check SNP 
  

#########
# number of SNPs on per chromosome # subgenome group/chr ID/main or random chromosome 
chr_ID <- unique(Ae_Ol_after_filter.unique$CHROM)
number.unique <- list()

for (chr in chr_ID) {
  number.unique[chr] <- sum(sum(Ae_Ol_after_filter.unique$CHROM==chr))
  print(number.unique[chr])
}

SNP_chr.unique <- as.data.frame(t(as.data.frame(number.unique)))
SNP_chr.unique$ID <- rownames(SNP_chr.unique)
head(SNP_chr.unique)
# split to main & random chromosomes 
# main 
SNP_chr_main.unique <- SNP_chr.unique[!grepl("random", SNP_chr.unique$ID),]  
SNP_chr_main.unique$subgenome <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", SNP_chr_main.unique$ID)   
SNP_chr_main.unique$chr_ID <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", SNP_chr_main.unique$ID) 

# random 
SNP_chr_random.unique <- SNP_chr.unique[grep("random", SNP_chr.unique$ID),]  
SNP_chr_random.unique$subgenome <- gsub("(chr)(A|C|U)(01|02|03|04|05|06|07|08|09|10|nn)(_)(random)", "\\2", SNP_chr_random.unique$ID)  
SNP_chr_random.unique$chr_ID <- gsub("(chr)(A|C|U)(01|02|03|04|05|06|07|08|09|10|nn)(_)(random)", "\\3", SNP_chr_random.unique$ID) 

# do a little calculation: how many SNPs on main & random chromosomes? 
sum(SNP_chr_main.unique$V1) # 200317    
sum(SNP_chr_random.unique$V1) # 32160    
sum(SNP_chr_main.unique$V1)/sum(sum(SNP_chr_main.unique$V1),sum(SNP_chr_random.unique$V1)) # 86.2% 

# make a plot to see how many SNPs on each chromosome 
library(ggplot2)
pl.SNP.main.1.unique <- ggplot(data = SNP_chr_main.unique) 
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + geom_bar(aes(x=factor(chr_ID), y = V1, fill=subgenome), stat = "identity")
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + facet_wrap(~subgenome, ncol = 1)
# pl.SNP.main.1 <- pl.SNP.main.1 + geom_text(aes(x=factor(chr_ID), y = number, label=factor(number), size=1)) 
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + theme(legend.position = "none")
pl.SNP.main.1.unique 

ggsave(pl.SNP.main.1.unique, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/SNP_calling/output/SNP.main.1.unique.01_26.png", height = 6, width = 8)

########### plot the position along the chromosome of each SNP, read depth 
head(Ae_Ol_after_filter.unique) 
# write.table(Ae_Ol_after_filter.unique, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/SNP_calling/output/Ae_Ol_after_filter.unique.vcf")
# get main chromosome and random chromsome seperately 
Ae_Ol_after_filter.main.unique <- Ae_Ol_after_filter.unique[grep("random", Ae_Ol_after_filter.unique$CHROM, invert = T),]
dim(Ae_Ol_after_filter.main.unique) # 200317     25   

Ae_Ol_after_filter.main.unique$subgenome <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", Ae_Ol_after_filter.main.unique$CHROM)
Ae_Ol_after_filter.main.unique$chr_ID <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", Ae_Ol_after_filter.main.unique$CHROM)

# number of SNPs on A & C subgenome
sum(Ae_Ol_after_filter.main.unique$subgenome=="A")/nrow(Ae_Ol_after_filter.main.unique) # 69.6%
sum(Ae_Ol_after_filter.main.unique$subgenome=="C") # 61464

## make plot for main & random chromosome seperately (distribution of SNPs per Mb)
library(ggplot2)
pl.SNP.main.2.unique <- ggplot(data = Ae_Ol_after_filter.main.unique)
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + geom_histogram(aes(x=POS, fill=subgenome), binwidth = 1000000) 
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + facet_grid(chr_ID ~subgenome)
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + theme(legend.position = "none")
pl.SNP.main.2.unique

ggsave("~/Desktop/Brassica_project/KIAT_RNA_seq/SNP_calling/output/SNP.main.2.unique.01.26.2017.png", width = 23, height = 13)

Ae_Ol_after_filter.random.unique <- Ae_Ol_after_filter.unique[grep("random", Ae_Ol_after_filter.unique$CHROM),] 
dim(Ae_Ol_after_filter.random.unique) # 32160    25   

### check Da-Ae VS ref, do I get the same pattern? 

chr_ID_Ae_ref.unique <- unique(Ae_ref_after_filter.unique$CHROM)
number_Ae_ref.unique <- c()

for (chr in chr_ID_Ae_ref.unique) {
  number_Ae_ref.unique[chr] <- sum(sum(Ae_ref_after_filter.unique$CHROM==chr))
  print(number_Ae_ref.unique[chr])
}

SNP_chr_Ae_ref.unique <- as.data.frame(number_Ae_ref.unique)
SNP_chr_Ae_ref.unique
SNP_chr_Ae_ref.unique$ID <- rownames(SNP_chr_Ae_ref.unique)
# split to main & random chromosomes
# main
SNP_chr_main_Ae_ref.unique <- SNP_chr_Ae_ref.unique[!grepl("random", SNP_chr_Ae_ref.unique$ID),]
SNP_chr_main_Ae_ref.unique$subgenome <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", SNP_chr_main_Ae_ref.unique$ID)
SNP_chr_main_Ae_ref.unique$subgenome
SNP_chr_main_Ae_ref.unique$chr_ID <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", SNP_chr_main_Ae_ref.unique$ID)
SNP_chr_main_Ae_ref.unique$chr_ID
SNP_chr_main_Ae_ref.unique$number_Ae_ref
# random
SNP_chr_random_Ae_ref.unique <- SNP_chr_Ae_ref.unique[grep("random", SNP_chr_Ae_ref.unique$ID),]
SNP_chr_random_Ae_ref.unique$subgenome <- gsub("(chr)(A|C|U)(01|02|03|04|05|06|07|08|09|10|nn)(_)(random)", "\\2", SNP_chr_random_Ae_ref.unique$ID)
SNP_chr_random_Ae_ref.unique$chr_ID <- gsub("(chr)(A|C|U)(01|02|03|04|05|06|07|08|09|10|nn)(_)(random)", "\\3", SNP_chr_random_Ae_ref.unique$ID)

# do a little calculation: how many SNPs on main & random chromosomes?
sum(SNP_chr_main_Ae_ref.unique$number) # 325903  
sum(SNP_chr_random_Ae_ref.unique$number) # 68834  
sum(SNP_chr_main_Ae_ref.unique$number)/sum(sum(SNP_chr_main_Ae_ref.unique$number),sum(SNP_chr_random_Ae_ref.unique$number)) # 82.6% 

sum(SNP_chr_main_Ae_ref.unique[SNP_chr_main_Ae_ref.unique$subgenome=="A",]$number_Ae_ref) # 203780  
sum(SNP_chr_main_Ae_ref.unique[SNP_chr_main_Ae_ref.unique$subgenome=="C",]$number_Ae_ref) # 122123    

SNP_chr_main_Ae_ref.unique$chr_ID_2 <- paste(SNP_chr_main_Ae_ref.unique$subgenome, SNP_chr_main_Ae_ref.unique$chr_ID, sep = "")
SNP_chr_main_Ae_ref.unique$chr_ID_2
 
# make a plot to see how many SNPs on each chromosome 
pl.SNP.main.Ae_ref.unique <- ggplot(data = SNP_chr_main_Ae_ref.unique) 
pl.SNP.main.Ae_ref.unique <- pl.SNP.main.Ae_ref.unique + geom_bar(aes(x=factor(chr_ID_2), y = number_Ae_ref.unique, fill=subgenome), stat = "identity")
# pl.SNP.main.Ae_ref <- pl.SNP.main.Ae_ref + facet_wrap(~subgenome, ncol = 1)
# pl.SNP.main.1 <- pl.SNP.main.1 + geom_text(aes(x=factor(chr_ID), y = number, label=factor(number), size=1))
pl.SNP.main.Ae_ref.unique <- pl.SNP.main.Ae_ref.unique + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
# pl.SNP.main.Ae_ref <- pl.SNP.main.Ae_ref + theme(legend.position = "none")
pl.SNP.main.Ae_ref.unique

# ggsave(pl.SNP.main.Ae_ref, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/SNP_calling/output/SNP.main.Ae_ref.png", height = 6, width = 10) 

chr_ID_Ol_ref.unique <- unique(Ol_ref_after_filter.unique$CHROM)
number_Ol_ref.unique <- c()  

for (chr in chr_ID_Ol_ref.unique) {
  number_Ol_ref.unique[chr] <- sum(sum(Ol_ref_after_filter.unique$CHROM==chr))
  print(number_Ol_ref.unique[chr])s
}

SNP_chr_Ol_ref.unique <- as.data.frame(number_Ol_ref.unique)
SNP_chr_Ol_ref.unique
SNP_chr_Ol_ref.unique$ID <- rownames(SNP_chr_Ol_ref.unique)
# split to main & random chromosomes
# main
SNP_chr_main_Ol_ref.unique <- SNP_chr_Ol_ref.unique[!grepl("random", SNP_chr_Ol_ref.unique$ID),]
SNP_chr_main_Ol_ref.unique$subgenome <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", SNP_chr_main_Ol_ref.unique$ID)
SNP_chr_main_Ol_ref.unique$subgenome
SNP_chr_main_Ol_ref.unique$chr_ID <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", SNP_chr_main_Ol_ref.unique$ID)
SNP_chr_main_Ol_ref.unique$chr_ID
SNP_chr_main_Ol_ref.unique$number_Ol_ref
# random
SNP_chr_random_Ol_ref.unique <- SNP_chr_Ol_ref.unique[grep("random", SNP_chr_Ol_ref.unique$ID),]
SNP_chr_random_Ol_ref.unique$subgenome <- gsub("(chr)(A|C|U)(01|02|03|04|05|06|07|08|09|10|nn)(_)(random)", "\\2", SNP_chr_random_Ol_ref.unique$ID)
SNP_chr_random_Ol_ref.unique$chr_ID <- gsub("(chr)(A|C|U)(01|02|03|04|05|06|07|08|09|10|nn)(_)(random)", "\\3", SNP_chr_random_Ol_ref.unique$ID)

# do a little calculation: how many SNPs on main & random chromosomes?
sum(SNP_chr_main_Ol_ref.unique$number) #322119  
sum(SNP_chr_random_Ol_ref.unique$number) # 69825   

sum(SNP_chr_main_Ol_ref.unique[SNP_chr_main_Ol_ref.unique$subgenome=="A",]$number_Ol_ref) # 189154   
sum(SNP_chr_main_Ol_ref.unique[SNP_chr_main_Ol_ref.unique$subgenome=="C",]$number_Ol_ref) # 132965  

SNP_chr_main_Ol_ref.unique$chr_ID_2 <- paste(SNP_chr_main_Ol_ref.unique$subgenome, SNP_chr_main_Ol_ref.unique$chr_ID, sep = "")

# make a plot to see how many SNPs on each chromosome
pl.SNP.main.Ol_ref.unique <- ggplot(data = SNP_chr_main_Ol_ref.unique)
pl.SNP.main.Ol_ref.unique <- pl.SNP.main.Ol_ref.unique + geom_bar(aes(x=factor(chr_ID_2), y = number_Ol_ref.unique, fill=subgenome), stat = "identity")
# pl.SNP.main.Ol_ref <- pl.SNP.main.Ol_ref + facet_wrap(~subgenome, ncol = 1)
# pl.SNP.main.1 <- pl.SNP.main.1 + geom_text(aes(x=factor(chr_ID), y = number, label=factor(number), size=1))
pl.SNP.main.Ol_ref.unique <- pl.SNP.main.Ol_ref.unique + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
# pl.SNP.main.Ol_ref <- pl.SNP.main.Ol_ref + theme(legend.position = "none")
pl.SNP.main.Ol_ref.unique

ggsave(pl.SNP.main.Ol_ref.unique, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/SNP_calling/output/SNP.main.Ol_ref.01_26_2017.png", height = 6, width = 10)

# combine 2 graphs
library(cowplot)
plot.w.ref.unique<-plot_grid(
  pl.SNP.main.Ae_ref.unique+labs(title="Da-Ae VS. reference genome"),
  pl.SNP.main.Ol_ref.unique+labs(title="Da-Ol-1 VS. reference genome"),
  ncol=1, nrow = 2,labels=c("","","",""))

plot.w.ref.unique 
save_plot("~/Desktop/Brassica_project/KIAT_RNA_seq/SNP_calling/output/Ae_Ol_ref", plot.w.ref, ncol = 1, nrow = 2,base_aspect_ratio = 0.8)  

```


# using GATK to call SNP 
```{r}
# 0) get uniquely mapped reads --> from sam to bam --> sort bam file (Do we need to do this step? yes, this step should be done for polyploid, based on discussion with Julin...)
# 0.1) get uniquely mapped reads 
# samtools view ../alignment_freebayes/Ae_paired.star.trim.dir/Aligned.sortedByCoord.out.bam | awk '$5 == "255"' > Ae_unique_mapped.bam (screen -r 35901.ttys009.coloma)
# samtools view ../alignment_freebayes/Ol_paired.star.trim.dir/Aligned.sortedByCoord.out.bam | awk '$5 == "255"' > Ol_unique_mapped.bam (screen -r 35910.ttys009.coloma)
# 0.2) from sam to bam for sam w/o headline 
# samtools view -bT /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa Ae_unique_mapped.sam > Ae_test.bam
# samtools view -bT /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa Ol_unique_mapped.sam > Ol_test.bam
# 0.3) sort bam file 
# samtools sort Ae_test.bam -o Ae_unique_sorted
# samtools sort Ol_test.bam -o Ol_unique_sorted 

# 1.1) mark PCR duplicate (for the 1st time, stay with non unique reads and start from this step *** need more understanding of this step)
# picard MarkDuplicates I=../alignment_freebayes/Ae_paired.star.trim.dir/Aligned.sortedByCoord.out.bam O=Ae_marked_duplicates.bam M=marked_dup_metrics.txt 
# picard MarkDuplicates I=../alignment_freebayes/Ol_paired.star.trim.dir/Aligned.sortedByCoord.out.bam O=Ol_marked_duplicates.bam M=marked_dup_metrics.txt 

# 1.2) index bam file.. (don't need this? )
# samtools index Ae_marked_duplicates.bam 
# samtools index Ol_marked_duplicates.bam

# 1.3) add readgroup 
# picard AddOrReplaceReadGroups I=Ae_marked_duplicates.bam O=Ae_marked_duplicates_addrg.bam RGID=Ae RGLB=lib1 RGPL=illumina RGPU=unit1 RGSM=Ae
# picard AddOrReplaceReadGroups I=Ol_marked_duplicates.bam O=Ol_marked_duplicates_addrg.bam RGID=Ol RGLB=lib1 RGPL=illumina RGPU=unit1 RGSM=Ol

# 1.4) index 
# samtools index Ol_marked_duplicates_addrg.bam
# samtools index Ae_marked_duplicates_addrg.bam

# 2.1) create fastq sequence dictionary file 
# picard CreateSequenceDictionary R=Brassica_napus_v4.1.chromosomes.fa O=Brassica_napus_v4.1.chromosomes.dict

# 2.2) split 'N' trim (visualize bam alignment before & after alignment in IGV)
# GATK -T SplitNCigarReads -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -I Ae_marked_duplicates_addrg.bam -o Ae_marked_duplicates_addrg_split.bam -rf ReassignOneMappingQuality -RMQF 255 -RMQT 60 -U ALLOW_N_CIGAR_READS
# GATK -T SplitNCigarReads -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -I Ol_marked_duplicates_addrg.bam -o Ol_marked_duplicates_addrg_split.bam -rf ReassignOneMappingQuality -RMQF 255 -RMQT 60 -U ALLOW_N_CIGAR_READS 

# 3.1) determine suspicious intervals which are likely in need of realignment 
# GATK -T RealignerTargetCreator -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -I Ae_marked_duplicates_addrg_split.bam -o Ae_forIndelRealigner.intervals
# GATK -T RealignerTargetCreator -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -I Ol_marked_duplicates_addrg_split.bam -o Ol_forIndelRealigner.intervals

# 3.2) running realigner over those intervals 
# GATK -T IndelRealigner -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -I Ae_marked_duplicates_addrg_split.bam -targetIntervals Ae_forIndelRealigner.intervals -o Ae_realignedBam.bam 
# GATK -T IndelRealigner -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -I Ol_marked_duplicates_addrg_split.bam -targetIntervals Ol_forIndelRealigner.intervals -o Ol_realignedBam.bam

# 3-4) Base Recalibration (command line: BaseRecalibrator & PrintReads) 
# GATK -T BaseRecalibrator -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -I Ae_realignedBam.bam -o Ae_recal_data.table
# GATK -T BaseRecalibrator -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -I Ol_realignedBam.bam -o Ol_recal_data.table  

# This step is recommanded, but I don't have a SNP set which I have confidence to use as a reference for calibration now, so decide to run SNP calling w/o this step, and after I get SNP set from freebayes and GATK, check the overlaps between those two, that overlapped set can be used as a reference. (Zhongxu suggested that this step is not necessary when there is no refernece set of SNP... however I should understand and talk to different poeple...) 

# 4) Variant calling 
# ran with chromosome splitted (use script GATK.sh), parallel computing, and get SNP for Ae and Ol at the same time.     
# GATK -T HaplotypeCaller -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -I in_Ae.bam -I in_Ol.bam -dontUseSoftClippedBases -stand_call_conf 20.0 -stand_emit_conf 20.0 -ploidy 4 -o out.vcf 

# 5) variant filtering (use script GATK_filter.sh) 
# GATK -T VariantFiltration -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -V in.vcf -window 35 -cluster 3 -o filtered.vcf

##### GATK -T VariantFiltration -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -V Ae.vcf -filterName FS -filter "FS > 30.0" -filterName QD -filter "QD < 2.0" -o Ae_filtered.vcf (doesn't work) 

# 6) merge SNPs from two genotypes seperately (use GATK_combine_Ae.sh and GATK_combine_Ol.sh)
# done 

# VCF format check/validate 
# GATK -T ValidateVariants -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -V Ae.vcf --validationTypeToExclude ALL
# GATK -T ValidateVariants -R ~/Reference/B.napus/Brassica_napus_v4.1.chromosomes.fa -V Ol.vcf --validationTypeToExclude ALL  

```

# SNP filtering from GATK 
```{r}
### import data and reformat 

vcf.GATK.A01 <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/data/GATK/as_diploid/xaa.intervals.vcf",as.is=T,na.strings = ".") 
head(vcf.GATK.A01) 

vcf.header.GATK <- system("grep '#C' ~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/data/GATK/as_diploid/xaa.intervals.vcf",intern = TRUE) 
head(vcf.header.GATK)
vcf.header.GATK <- sub("#","",vcf.header.GATK) #get rid of the pound sign
vcf.header.GATK

vcf.header.GATK <- unlist(strsplit(vcf.header.GATK,split="\t"))
colnames(vcf.GATK.A01) <- vcf.header.GATK
head(vcf.GATK.A01)
tail(vcf.GATK.A01)
dim(vcf.GATK.A01) # 68068   11

system("grep '##INFO' ~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/data/GATK/as_diploid/xaa.intervals.vcf")
system("grep '##FORMAT' ~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/data/GATK/as_diploid/xaa.intervals.vcf") 

########## 
# correct the file 
which(vcf.GATK.A01$FORMAT=="GT:GQ:PL")
class(vcf.GATK.A01)

vcf.GATK.corrected <- vcf.GATK.A01[-c(2425, 10905, 17842, 17843, 18085, 41722, 43080, 59817),]
sum(vcf.GATK.corrected$FORMAT=="GT:GQ:PL")
dim(vcf.GATK.corrected)

# Before splitting add NAs to blank cells
vcf.GATK.corrected$Ae
sum(vcf.GATK.corrected$Ae=="./.")

vcf.GATK.corrected$Ae[vcf.GATK.corrected$Ae=="./."] <- "NA:NA:NA:NA:NA"
sum(vcf.GATK.corrected$Ae=="NA:NA:NA:NA:NA") 

Ae.tmp.unique.GATK <- matrix(
  unlist(strsplit(vcf.GATK.corrected$Ae,split = ":")),
  nrow=nrow(vcf.GATK.corrected),  
  byrow=TRUE
  ) 

head(Ae.tmp.unique.GATK)
tail(Ae.tmp.unique.GATK) 

colnames(Ae.tmp.unique.GATK) <- paste("Ae",c("gt","ref.alt.depth","approx.depth","genotype.qual","Phred.score"),sep="_")

vcf.GATK.corrected$Ol[vcf.GATK.corrected$Ol=="./."] <- "NA:NA:NA:NA:NA"

Ol.tmp.unique.GATK <- matrix(
  unlist(strsplit(vcf.GATK.corrected$Ol,split = ":")),
  nrow=nrow(vcf.GATK.corrected),  
  byrow=TRUE
  ) 
head(Ol.tmp.unique.GATK)
tail(Ol.tmp.unique.GATK)
dim(Ol.tmp.unique.GATK) # 68060   5 
dim(Ae.tmp.unique.GATK) 

colnames(Ol.tmp.unique.GATK) <- paste("Ol",c("gt","ref.alt.depth","approx.depth","genotype.qual","Phred.score"),sep="_")

vcf.GATK.reform <- cbind(vcf.GATK.corrected, Ae.tmp.unique.GATK, Ol.tmp.unique.GATK,stringsAsFactors=FALSE) 

summary(vcf.GATK.reform)
head(vcf.GATK.reform)

######## 
vcf.GATK.reform[,c("Ae_approx.depth","Ae_genotype.qual",
            "Ol_approx.depth","Ol_genotype.qual")] <-
  apply(vcf.GATK.reform[,c("Ae_approx.depth","Ae_genotype.qual",
            "Ol_approx.depth","Ol_genotype.qual")],
        2,
        as.numeric
        )

dim(vcf.GATK.reform) # 68060 21       

#### filter (based on QUAL score and snpcluster)
# 1) filter snpcluster 
# vcf.GATK.reform.pass <- vcf.GATK.reform[vcf.GATK.reform$FILTER!="SnpCluster",]
# nrow(vcf.GATK.reform.pass)/nrow(vcf.GATK.reform) # 50% 
unique(vcf.GATK.reform$Ae_gt)
unique(vcf.GATK.reform$Ol_gt)

### 1) filter based on QUAl score 

# make a histogram of QUAL scores 
hist(vcf.GATK.reform$QUAL) 

length(vcf.GATK.reform$QUAL) # a total of 76461 SNPs 
sum(vcf.GATK.reform$QUAL>40) / length(vcf.GATK.reform$QUAL) # 91% are above QUAL score of 40, which means less than 0.0001 probability that it isn't polymorphic 

# subset the data to keep positions where the quality score is 40 or higher  
vcf.GATK.reform.HQ <- vcf.GATK.reform[vcf.GATK.reform$QUAL>40,]
dim(vcf.GATK.reform.HQ) # 62247    21    

# # count the number (for some reason this desn't work...)
table(vcf.GATK.reform.HQ$Ae_gt)
table(vcf.GATK.reform.HQ$Ol_gt)  

# which SNPs would be most useful for a downstream QTL analysis between Ae & Ol? 

# SNPs between Ae & ref
vcf.GATK.reform.pass.HQ$Ae_gt
sum(vcf.GATK.reform.pass.HQ$Ae_gt=="1/1/1/1") # 12502     
Ae_ref.GATK <- vcf.GATK.reform.pass.HQ[vcf.GATK.reform.pass.HQ$Ae_gt=="1/1/1/1",]
# read depth > 3 and read depth < 501  
# as.numeric(Ae_ref.GATK$Ae_approx.depth)
# Ae_ref_filter.GATK <- Ae_ref.GATK[Ae_ref.GATK$Ae_approx.depth > 3 &Ae_ref.GATK$Ae_approx.depth <  501,] 
dim(Ae_ref.GATK) # 12502  21  
head(Ae_ref.GATK) 

# SNPs between Ol & ref 
sum(vcf.GATK.reform.pass.HQ$Ol_gt=="1/1/1/1") # 11352    
Ol_ref.GATK <- vcf.GATK.reform.pass.HQ[vcf.GATK.reform.pass.HQ$Ol_gt=="1/1/1/1",]
# read depth > 3 and read depth < 501  
# as.numeric(Ae_ref.GATK$Ae_approx.depth)
# Ae_ref_filter.GATK <- Ae_ref.GATK[Ae_ref.GATK$Ae_approx.depth > 3 &Ae_ref.GATK$Ae_approx.depth <  501,] 
dim(Ol_ref.GATK) # 11352  21  
head(Ol_ref.GATK) 

# SNPs between Ae & Ol 
sum(vcf.GATK.reform.pass.HQ$Ae_gt=="1/1/1/1" & vcf.GATK.reform.pass.HQ$Ol_gt=="0/0/0/0") +  sum(vcf.GATK.reform.pass.HQ$Ae_gt=="0/0/0/0" & vcf.GATK.reform.pass.HQ$Ol_gt=="1/1/1/1") # 8491 SNPs between Ae & Ol 

# subset SNPs between Ae & Ol 
vcf.Ae.Ol.GATK <- vcf.GATK.reform.pass.HQ[((vcf.GATK.reform.pass.HQ$Ae_gt=="1/1/1/1" & vcf.GATK.reform.pass.HQ$Ol_gt=="0/0/0/0") | (vcf.GATK.reform.pass.HQ$Ae_gt=="0/0/0/0" & vcf.GATK.reform.pass.HQ$Ol_gt=="1/1/1/1")),] 

dim(vcf.Ae.Ol.GATK) # 8491 21   
# read depth > 3 and read depth < 501  
   
```  

# compare SNPs from freebayes & GATK 
```{r}
vcf.Ae.Ol.freebayes <- Ae_Ol_after_filter.unique[Ae_Ol_after_filter.unique$CHROM=="chrA01",]
nrow(vcf.Ae.Ol.freebayes)  # 11379  
dim(vcf.Ae.Ol.GATK) # 8491   

vcf.Ae.Ol.freebayes.SNP <- paste(vcf.Ae.Ol.freebayes$CHROM, vcf.Ae.Ol.freebayes$POS, sep = "_")
vcf.Ae.Ol.GATK.SNP <- paste(vcf.Ae.Ol.GATK$CHROM, vcf.Ae.Ol.GATK$POS, sep = "_") 

intersect(vcf.Ae.Ol.GATK.SNP, vcf.Ae.Ol.freebayes.SNP) 

#### plot of venn diagram 
library(gplots)
library(VennDiagram)

grid.newpage() 
venn.freebayes_GATK <- 
draw.pairwise.venn(nrow(vcf.Ae.Ol.freebayes), nrow(vcf.Ae.Ol.GATK), length(intersect(vcf.Ae.Ol.GATK.SNP, vcf.Ae.Ol.freebayes.SNP)), 
                 # category = c("freebayes (11379)", "GATK (8491)"), 
                 lty = "blank", 
                 fill = c("skyblue", "pink1"))
save_plot("~/Desktop/Brassica_project/KIAT_RNA_seq/SNP_calling/output/venn_freebayes_GATK.png", venn.freebayes_GATK,ncol = 1, nrow = 1,base_aspect_ratio = 1.5, base_height = 2) 

######### Ae_ref 
vcf.Ae.ref.freebayes <- Ae_ref_after_filter.unique[Ae_ref_after_filter.unique$CHROM=="chrA01",]
nrow(vcf.Ae.ref.freebayes)  # 18059  
dim(Ae_ref.GATK) # 12502 

vcf.Ae.ref.freebayes.SNP <- paste(vcf.Ae.ref.freebayes$CHROM, vcf.Ae.ref.freebayes$POS, sep = "_")
vcf.Ae.ref.GATK.SNP <- paste(Ae_ref.GATK$CHROM, Ae_ref.GATK$POS, sep = "_") 

intersect(vcf.Ae.ref.GATK.SNP, vcf.Ae.ref.freebayes.SNP) 

#### plot of venn diagram 

grid.newpage() 
venn.freebayes_GATK.Ae_ref <- 
draw.pairwise.venn(nrow(vcf.Ae.ref.freebayes), nrow(Ae_ref.GATK), length(intersect(vcf.Ae.ref.GATK.SNP, vcf.Ae.ref.freebayes.SNP)), 
                 category = c("freebayes (11379)", "GATK (8491)"), 
                 lty = "blank", 
                 fill = c("skyblue", "pink1"))

######### Ol_ref 
vcf.Ol.ref.freebayes <- Ol_ref_after_filter.unique[Ol_ref_after_filter.unique$CHROM=="chrA01",]
nrow(vcf.Ol.ref.freebayes)  # 16225   
dim(Ol_ref.GATK) # 11352  

vcf.Ol.ref.freebayes.SNP <- paste(vcf.Ol.ref.freebayes$CHROM, vcf.Ol.ref.freebayes$POS, sep = "_")
vcf.Ol.ref.GATK.SNP <- paste(Ol_ref.GATK$CHROM, Ol_ref.GATK$POS, sep = "_") 

intersect(vcf.Ol.ref.GATK.SNP, vcf.Ol.ref.freebayes.SNP) 

#### plot of venn diagram 

grid.newpage() 
venn.freebayes_GATK.Ol_ref <- 
draw.pairwise.venn(nrow(vcf.Ol.ref.freebayes), nrow(Ol_ref.GATK), length(intersect(vcf.Ol.ref.GATK.SNP, vcf.Ol.ref.freebayes.SNP)), 
                 category = c("freebayes (11379)", "GATK (8491)"), 
                 lty = "blank", 
                 fill = c("skyblue", "pink1"))

``` 

# compare mapping result to napus, rapa, and juncea for Da-Ae & Da-Ol-1 
```{r}
# 1) get material, flowering stage sample for Da-Ae & Da-Ol-1 
# done

# 2) index reference rapa & juncea genome 
# rapa genome 
# juncea genome 

# 3) map 

# 4) check & compare result  
#### total mapping summary  
total_mapping <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/mapping_napus_rapa_juncea/mapping_summary.csv", header = F)
head(total_mapping)
total_mapping$value <- as.numeric(gsub("%", "", total_mapping$V3, fixed=T))/100
total_mapping$ref <- gsub("([[:print:]]+)(rapa|napus|juncea)", "\\2", total_mapping$V1)
total_mapping$lib <- gsub("([[:print:]]+)(_)(paired.star.dir.rapa|paired.star.dir.napus|paired.star.dir.juncea)", "\\1", total_mapping$V1)
total_mapping$gt <- c(rep("Da-Ae",24), rep("Da-Ol-1", 24))
total_mapping$lib

p.total_mapping <- ggplot(data = total_mapping) + theme_grey()
p.total_mapping <- p.total_mapping + geom_bar(aes(x=factor(V2), y = value, fill=V2), stat="identity")
p.total_mapping <- p.total_mapping + facet_grid(ref~lib)
p.total_mapping <- p.total_mapping + geom_text(data = total_mapping, aes(x=factor(V2), y=value, label=factor(value)))
# geom_text(data=DEGs_number_between_gt.melt.parent,aes(x=DE,y=number*1.05,label=factor(number)))
p.total_mapping 
ggsave(p.total_mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/mapping_napus_rapa_juncea/p.total_mapping.png", width = 12, height=8)

# conlussions: no difference between Da-Ae and Da-Ol-1's mapping ratio to different ref genomes. napus has the highest mapping ratio among the 3 ref genomes, followed by juncea and rapa.  

#### subgenome mapping summary (only unique mapped reads)
subgenome_mapping <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/mapping_napus_rapa_juncea/subgenome_mapping_summary.csv", header = F)
head(subgenome_mapping, 20)
subgenome_mapping <- subgenome_mapping[grep("^A$|^B$|^C$", subgenome_mapping$V2),]
unique_mapping_sum <- aggregate(subgenome_mapping$V1, list(subgenome_mapping$V3), sum)
names(unique_mapping_sum) <- c("V3", "total")
head(unique_mapping_sum)
subgenome_mapping.2 <- merge(subgenome_mapping, unique_mapping_sum, by="V3")
subgenome_mapping.2$ratio <- subgenome_mapping.2$V1/subgenome_mapping.2$total

subgenome_mapping.2$ref <- gsub("([[:print:]]+)(_paired.star.dir.)(napus|rapa|juncea)([[:punct:]]+)(subgenome_mapping_summary_main)", "\\3", subgenome_mapping.2$V3)
subgenome_mapping.2$lib <- gsub("([[:print:]]+)(_paired.star.dir.)(napus|rapa|juncea)([[:punct:]]+)(subgenome_mapping_summary_main)", "\\1", subgenome_mapping.2$V3)
subgenome_mapping.2$gt <- c(rep("Da-Ae", 10), rep("Da-Ol-1", 10))

subgenome_mapping.2$lib
subgenome_mapping.2$ratio
subgenome_mapping.2$V2 

p.subgenome_mapping <- ggplot(data = subgenome_mapping.2)
p.subgenome_mapping <- p.subgenome_mapping + geom_bar(aes(x=V2, y = ratio, fill=V2), stat="identity")
p.subgenome_mapping <- p.subgenome_mapping + facet_grid(ref~lib)
p.subgenome_mapping <- p.subgenome_mapping + geom_text(data = subgenome_mapping.2, aes(x=factor(V2), y=ratio, label=factor(round(ratio, digits = 2))))
# geom_text(data=DEGs_number_between_gt.melt.parent,aes(x=DE,y=number*1.05,label=factor(number)))
p.subgenome_mapping 
ggsave(p.subgenome_mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/mapping_napus_rapa_juncea/p.subgenome_mapping.png", width = 12, height=8)
# conlussion: the lowest but good partion of reads mapped to B subgenome, and the percentage is about the same between Da-Ae & Da_Ol-1. This suggest that that the genetic makeup between Da-Ae & Da-Ol-1 is not that different... 

# Ae_mapped <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/SNP_calling/data/Ae_mapping_chr_2", header = F)
# Ae_mapped$value <- gsub("([[:digit:]]+)(|)([[:print:]]+)", "\\1", Ae_mapped$V1)
# Ae_mapped
# 
# 
# Ol_mapped <- read.delim("~/Desktop/Brassica_project/KIAT_RNA_seq/SNP_calling/data/Ol_mapping_chr_2", header = F)
# Ol_mapped$value <- gsub("([[:digit:]]+)(|)([[:print:]]+)", "\\1", Ol_mapped$V1)
# Ol_mapped
# 
# sum(as.numeric(Ae_mapped[c(1,3,5,7,9,11,13,15,17,19), 2])) 
# sum(as.numeric(Ol_mapped[c(1,3,5,7,9,11,13,15,17,19), 2]))
# sum(as.numeric(Ae_mapped[c(1,3,5,7,9,11,13,15,17,19), 2])) + sum(as.numeric(Ol_mapped[c(1,3,5,7,9,11,13,15,17,19), 2]))
# # 383980934 A subgenome mapped reads 
# sum(as.numeric(Ae_mapped[c(22,24,26,28,30,32,34,36,38), 2])) + sum(as.numeric(Ol_mapped[c(22,24,26,28,30,32,34,36,38), 2])) 
# sum(as.numeric(Ae_mapped[c(22,24,26,28,30,32,34,36,38), 2]))
# sum(as.numeric(Ol_mapped[c(22,24,26,28,30,32,34,36,38), 2])) 
# # 786750509 C subgenome mapped reads 
# 
# Ae_mapped[c(1,3,5,7,9,11,13,15,17,19), 1]
# Ol_mapped[c(1,3,5,7,9,11,13,15,17,19), 1]
# Ae_mapped[c(22,24,26,28,30,32,34,36,38), 1]
# Ol_mapped[c(22,24,26,28,30,32,34,36,38), 1]

#### reads level mapping detail check ... 
reads_mapping <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/mapping_napus_rapa_juncea/reads_mapping_summary.csv", header = F)
head(reads_mapping)
reads_mapping <- reads_mapping[grep("^A$|^C$|^non_unique$|^unmapped$", reads_mapping$V2),]
reads_mapping_sum <- aggregate(reads_mapping$V1, list(reads_mapping$V3), sum)
names(reads_mapping_sum) <- c("V3", "total")
head(reads_mapping_sum)
reads_mapping.2 <- merge(reads_mapping, reads_mapping_sum, by="V3")
reads_mapping.2$ratio <- reads_mapping.2$V1/reads_mapping.2$total

reads_mapping.2

reads_mapping.2$ref <- gsub("([[:print:]]+)(juncea|rapa)(_)(6_Da-Ae_Asub.summary|6_Da-Ae_Csub.summary|All1_Gae_Da-Ol_Asub.summary|All1_Gae_Da-Ol_Csub.summary|6_Ae_summary|All1_Gae_Ol_summary)", "\\2", reads_mapping.2$V3)
reads_mapping.2$sub <- c(rep("A", 4), rep("B", 4), rep("A", 3), rep("B", 3), rep("A", 7))
reads_mapping.2$gt <- c(rep("Da-Ae", 8), rep("Da-Ol-1", 6), rep("Da-Ae", 4), rep("Da-Ol-1", 3))
reads_mapping.2$group <- paste(reads_mapping.2$ref, reads_mapping.2$sub, sep = "_")

p.reads_mapping <- ggplot(data = reads_mapping.2)
p.reads_mapping <- p.reads_mapping + geom_bar(aes(x=V2, y = ratio, fill=V2), stat="identity")
p.reads_mapping <- p.reads_mapping + facet_grid(group~gt)
p.reads_mapping <- p.reads_mapping + geom_text(data = reads_mapping.2, aes(x=factor(V2), y=ratio, label=factor(round(ratio, digits = 2))))
# geom_text(data=DEGs_number_between_gt.melt.parent,aes(x=DE,y=number*1.05,label=factor(number)))
p.reads_mapping 
ggsave(p.reads_mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/mapping_napus_rapa_juncea/p.reads_mapping.png", width = 12, height=8)

# conclussion: 

# p.reads_mapping.2 <- ggplot(data = reads_mapping.2) 
# p.reads_mapping.2 <- p.reads_mapping.2 + geom_bar(aes(x="", y= ratio, fill=V2), stat="identity")
# p.reads_mapping.2 <- p.reads_mapping.2 + coord_polar("y") 
# p.reads_mapping.2 <- p.reads_mapping.2 + facet_grid(gt~group) + theme(axis.text.x=element_blank())
# p.reads_mapping.2 <- p.reads_mapping.2 + geom_text(data = reads_mapping.2, aes(x="", y=ratio, label=factor(round(ratio, digits = 2)), size=3), check_overlap = TRUE) 
# p.reads_mapping.2 
# ggsave(p.reads_mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/mapping_napus_rapa_juncea/p.reads_mapping.png", width = 12, height=8)

```

# SNP filter inside linux or R 
```{r}
######## SNP filtering can be done inside R, but need bgzip & tabix to index the vcf file.... 
#### spent quite a long time to figure this out... 

### visulize unique & non-unique SNPs in IGV, done... 

### SNP filtering inside linux using script (see /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_fall/SNP_calling/alignment_freebayes/vcf_file/log file in whitney)

### alignment score can also be taken into consideration for filtering before SNP calling, because there are multiple possible causes of false-positive SNPs: 1) sequencing error [Phred score] 2) alignment error [mapping score inside bam file] 3) incompleness of ref genome (ref: Wang et al. 2015; )

### filter based on read depth (DP)? alternate allele observation count (AO)? QA (sum of quality of the alternate observation count)  

Ae_Ol_after_filter.main.unique.test <- Ae_Ol_after_filter.main.unique[Ae_Ol_after_filter.main.unique$CHROM=="chrA02",]

colnames(Ae_Ol_after_filter.main.unique.test)
total_depth_check <- Ae_Ol_after_filter.main.unique.test[,c("CHROM","POS", "Ae_tot.depth", "Ol_tot.depth")]
head(total_depth_check)
total_depth_check_long <- melt(total_depth_check, id.vars = c("CHROM", "POS"))
head(total_depth_check_long)

pl.total_depth <- ggplot(data = total_depth_check_long)
pl.total_depth <- pl.total_depth + geom_line(aes(x=POS, y = value, color=variable))
# pl.total_depth <- pl.total_depth + facet_facet(~variable)
# pl.SNP.main.2 <- pl.SNP.main.2 + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
# pl.SNP.main.2 <- pl.SNP.main.2 + theme(legend.position = "none")
pl.total_depth

# based on the total depth visualization result, some regions are having high depth for both Ae and Ol, whereas some regions are only exibiting high depth for one of the genotype 

hist(Ae_Ol_after_filter.main.unique.test[Ae_Ol_after_filter.main.unique.test$Ae_tot.depth>10000,]$Ae_tot.depth)
# average depth 
# Ae 
sum(Ae_Ol_after_filter.main.unique$Ae_tot.depth)/nrow(Ae_Ol_after_filter.main.unique) # 185.4755 
# Ol 
sum(Ae_Ol_after_filter.main.unique$Ol_tot.depth)/nrow(Ae_Ol_after_filter.main.unique) # 207.3724 
# 

# people use different cutoff values for max read depth, eg: 
# 1) average read depth + 3*sqrt(average read depth) 2) average read depth + 3 * stdv of read depth ... 
# based on all these different information, I decided to use less than 500 for both Ae and Ol as the cutoff value. 
# In summary, alternate allele count > 4, min read depth of 4, and max read depth of 500 for both genotypes  

```

# SNP annotation 
```{r}
# snpEff # this version of annotated vcf file is not correct... because "ERROR_CHROMOSOME_NOT_FOUND" 
########### 1) 
# build my own database ############### 
# cd data 
# mkdir Bnv5
# cd Bnv5 
# wget ftp://bradata:zhl410ivf@brassicadb.org/Brassica_napus/Brassica_napus.annotation_v5.gff3.gz
# mv Brassica_napus.annotation_v5.gff3.gz genes.gff.gz
# wget ftp://bradata:zhl410ivf@brassicadb.org/Brassica_napus/Brassica_napus.annotation_v5.pep.fa.gz
# wget ftp://bradata:zhl410ivf@brassicadb.org/Brassica_napus/Brassica_napus.annotation_v5.gff3.cds.fa.gz
# mv Brassica_napus.annotation_v5.pep.fa.gz protein.fa.gz
# mv Brassica_napus.annotation_v5.gff3.cds.fa.gz cds.fa.gz
# gunzip *.gz
# cd ..
# mkdir genomes 
# cd genomes/ 
# wget ftp://bradata:zhl410ivf@brassicadb.org/Brassica_napus/Brassica_napus_v4.1.chromosomes.fa.gz
# mv Brassica_napus_v4.1.chromosomes.fa.gz Bnv5.fa.gz 
# gunzip Bnv5.fa.gz
# cd ..  (to snpEff home directory)
# vim snpEff.config #### 
# add "# Brassica_napus v5 (self-built) Bnv5.genome: Brassica_napus Bnv5.reference: ftp://brassicadb.org/Brassica_napus/" 
# java -jar snpEff.jar build -gff3 -v Bnv5 

### SNP annotation 





``` 

# final output 
```{r}
# 1) summary of identified SNPs (table3, synnomynous & non synnomymous SNPs requires SNP annotation, exonic, intergenic, intronic...)

# 2) SNP distribution on sub A and sub C genome 

# 3) SNP distribution across chromosome (normalized by gene number, mapping rate...)

# 4) Ka/Ks... (looks like only available after annotaiton) 

# 5) if GATK result is available, add result, venn diagram show overlaps... 

```




