---
title: "KIAT_RNA_seq"
author: "Ruijuan Li"
date: "August 1, 2016"
output: html_document
---

# download data, QC, and mapping (using kallisto)
```{r}
# 1) download file from ftp using ftp_file_transfer.sh
# 1.1) unzip files using gunzip *  
# 2) get first 1000 sequences for each file extract_1000_reads.sh, actually worked with all data instead of the 1st 1000 seqs
# 3) QC file using fastQC: fastqc *.fq -o /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/fastqc_result
# 4) download Brassica_napus.annotation_v5.cds.fa to whitney 
# 5) collapse fastqc summary result into one file & get overrepresented sequences using ./make_summary_report.sh 

# 6) reformat summary.all.txt file 
setwd("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/")
fastQC_summary <- read.delim("./fastQC_result/summary_all.txt", header = F)
dim(fastQC_summary)
head(fastQC_summary)
tail(fastQC_summary)

fastQC_summary_wide <- reshape(fastQC_summary, timevar="V2",idvar="V3",direction="wide")
colnames(fastQC_summary_wide) <- gsub("(V1)(.)([[:print:]]+)","\\3",colnames(fastQC_summary_wide))
dim(fastQC_summary_wide)

write.csv(fastQC_summary_wide, file = "fastQC_summary_wide.csv")

# 7) checked the source of overrepresented sequences using web nucleotide blast tool, found that most of them are from 
# cholorplast, mitochondra, rRNA, tansposon, etc. Some are illumina or TruSeq adapters 

# blast sequences against Trimmomatic adapters suggest that TruSeq3-PE-2.fa should be used to trimm the data. (Q: how to deal with the others? will they affect the final result? search... )

# Qs on plstids seqs contamination: 1) why are there these sequences? check the lib making step 2) will they affect downstream analysis if kept? CDS mapping VS genome mapping 3) If yes, how to remove them? (I need to figure these Qs out.)

# 8) using kallisto to build the reference cds index 
# kallisto index -k 19 -i Brassica_napus.annotation_v5.cds.19.kai /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/Brassica_napus.annotation_v5.cds.fa

# 9) map to reference cds index and get read count file using bunchrun_kallisto.sh (using est_counts from Kallisto)
```

# using STAR 
```{r}
# 1) get B.napus reference genome: Brassica_napus_v4.1.chromosomes.fa 
# make STAR reference index: STAR --runMode genomeGenerate --genomeDir star_nome/ --genomeFastaFiles Brassica_napus_v4.1.chromosomes.fa --runThreadN 2 
# mapping, only map the 2nd end using run_star.sh 




```

# SNP calling from all individuals & check sample seperation 
```{r}




```

# Expression analysis 1) formatting data   
```{r}
parent.read.count <- read.table("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/read_count/read.count.tsv", header = T, check.names = F)
rownames(parent.read.count) <- parent.read.count[,1]
parent.read.count <- parent.read.count[,-1]

# format data 
head(parent.read.count)
dim(parent.read.count) # 101040     54 
colnames(parent.read.count)
even_indexes<-seq(2,length(colnames(parent.read.count)),2)
parent.read.count.one <- parent.read.count[,even_indexes]
dim(parent.read.count.one) # 101040     27 
colnames(parent.read.count.one)
colnames(parent.read.count.one) <- sub("_2.fq","",colnames(parent.read.count.one),fixed = TRUE) # remove _2.fq 

# sample description  
sample_des <- read.csv("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/parent_summary_corrected.csv")
sample_des
dim(sample_des)
sorted_sample_des <- sample_des[order(sample_des$SampleID),]

sorted_sample_des[,4:7]
new_sample_ID <- paste(sorted_sample_des$Cultivar, sorted_sample_des$Stage, sorted_sample_des$rep, sep = "_")
new_sample_ID


# replace sample ID 
colnames(parent.read.count.one) <- new_sample_ID
head(parent.read.count.one)
save(parent.read.count.one,file="/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.Rdata")
```

# expression analysis 2) set up sample description 
```{r}
# load necessary libs & functions 
library(edgeR)
library(ggplot2)

# filter based on read count, assign group, normalize, design matrix, calculate dispersion   
# set up group 
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.Rdata")
parent.read.count.one <- parent.read.count.one[,colSums(parent.read.count.one) > 1000000]  
dim(parent.read.count.one) # 101040     27 
parent.read.count.one.sample<-data.frame(file=colnames(parent.read.count.one),
                             batch=factor(gsub("(Da-ae|Da-al-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\5",colnames(parent.read.count.one))),  
                             genotype=factor(gsub("(Da-ae|Da-al-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",colnames(parent.read.count.one))),	
                             stage=factor(gsub("(Da-ae|Da-al-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",colnames(parent.read.count.one))),	
                             group=factor(gsub("(Da-ae|Da-al-1)(_)([[:print:]]+)(_)(1|2|3)","\\1\\3",colnames(parent.read.count.one)))
)

parent.read.count.one.sample
ftable(parent.read.count.one.sample,row.vars="stage",col.vars=c("batch","genotype"))

# filter based on read count 
parent.read.count.one.small <- parent.read.count.one[rowSums(parent.read.count.one > 10) >= 3,]
dim(parent.read.count.one.small) # 60975    27 
save(parent.read.count.one.small, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.small.Rdata")
```

# expression analysis 3) voom transformation 
```{r}
# voom transformation 
# source("https://bioconductor.org/biocLite.R")
# biocLite("DESeq2")
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.small.Rdata")
library("DESeq2")

dds.parent <- DESeqDataSetFromMatrix(countData = round(parent.read.count.one.small), colData = parent.read.count.one.sample, design = ~ batch + genotype*stage)

vsd.parent <- varianceStabilizingTransformation(dds.parent)
vstMat.parent <- assay(vsd.parent)
colnames(vstMat.parent) <- colnames(parent.read.count.one)
save(vstMat.parent, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/vstMat.parent.Rdata")
```

# expression analysis 4) MDS, clustering, and expression analysis 
```{r}
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/vstMat.parent.Rdata")
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.small.Rdata")

# normalize 
dge.data.parent <- DGEList(counts=parent.read.count.one.small, group=parent.read.count.one.sample$group)
dge.data.parent <- calcNormFactors(dge.data.parent, method = "TMM") 
dge.data.parent$sample
hist(dge.data.parent$samples$norm.factors)

# MDS to check sample seperation, also using dengrogram. (too long names to show)  
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/MDS.png")
# plotMDS(dge.data.parent, method = "bcv")
# dev.off() 

# clustering to check sample seperation 
vstMat.parent <- as.data.frame(vstMat.parent)
head(vstMat.parent)
# making dengrogram
hc.parent <- hclust(dist(t(vstMat.parent)))

png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/clustering.png")
plot(hc.parent)
dev.off() 

# # check correlation among samples 
# cor.test(parent.read.count.one.small$`Da-ae_Young_1`, parent.read.count.one.small$`Da-ae_Young_2`)
# cor.test(parent.read.count.one.small$`Da-al-1_Young_2`, parent.read.count.one.small$`Da-al-1_Young_1`)
# cor.test(parent.read.count.one.small$`Da-al-1_Young_1`, parent.read.count.one.small$`Da-ae_early-silique_3`)

# pairwise comparison using GLM model 
# design matrix
# design.parent.batch <- model.matrix(~batch+group, data = parent.read.count.one.sample) 
design.parent <- model.matrix(~0+group, data = parent.read.count.one.sample)

# # estimate dispersion
# # w/ batch effect
# dge.data.parent.batch <- estimateGLMCommonDisp(dge.data.parent.batch, design.parent.batch,verbose = TRUE) # Disp = 0.22238 , BCV = 0.4716
# dge.data.parent.batch <- estimateGLMTrendedDisp(dge.data.parent.batch,design.parent.batch)
# dge.data.parent.batch <- estimateGLMTagwiseDisp(dge.data.parent.batch,design.parent.batch)
# plotBCV(dge.data.parent.batch)

# w/o batch effect
dge.data.parent <- estimateGLMCommonDisp(dge.data.parent, design.parent,verbose = TRUE) # Disp = 0.48381 , BCV = 0.6956
dge.data.parent <- estimateGLMTrendedDisp(dge.data.parent,design.parent)
dge.data.parent <- estimateGLMTagwiseDisp(dge.data.parent,design.parent)
plotBCV(dge.data.parent)

# fit GLM model & test for differential expression
# w/ batch effect
# fit.parent.batch <- glmFit(dge.data.parent.batch, design.parent.batch)
# lrt.young.batch <- glmLRT(fit.parent.batch, contrast = c(0,0,0,0,0,0,1,0,0,0,0,-1))
# topTags(lrt.young.batch)
# DEgene.young.batch <- topTags(lrt.young.batch,n = Inf)$table[topTags(lrt.young.batch,n = Inf)$table$FDR<0.05,]
# summary(de.young.batch <- decideTestsDGE(lrt.young.batch, p=0.05))
# # summary(de.young.batch <- decideTestsDGE(lrt.young.batch, p=0.05, lfc = 2))
# # -1  5072
# # 0  51131
# # 1   4772
# 
# detags.young <- rownames(lrt.young.batch)[as.logical(de.young.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.young.png")
# plotSmear(lrt.young.batch, de.tags=detags.young, main="young")
# abline(h = c(-2, 2), col = "blue")
# dev.off()
# 
# lrt.flowering.batch <- glmLRT(fit.parent.batch, contrast = c(0,0,0,0,1,0,0,0,0,-1,0,0))
# topTags(lrt.flowering.batch)
# summary(de.flowering.batch <- decideTestsDGE(lrt.flowering.batch, p=0.05))
# DEgene.flowering.batch <- topTags(lrt.flowering.batch,n = Inf)$table[topTags(lrt.flowering.batch,n = Inf)$table$FDR<0.05,]
# # -1  3150
# # 0  54342
# # 1   3483
# detags.flowering <- rownames(lrt.flowering.batch)[as.logical(de.flowering.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.flowering.png")
# plotSmear(lrt.flowering.batch, de.tags=detags.flowering, main="flowering")
# abline(h = c(-2, 2), col = "blue")
# dev.off()
# 
# lrt.early.silique.batch <- glmLRT(fit.parent.batch, contrast = c(0,0,0,1,0,0,0,0,-1,0,0,0))
# topTags(lrt.early.silique.batch)
# summary(de.early.silique.batch <- decideTestsDGE(lrt.early.silique.batch, p=0.05))
# DEgene.early.silique.batch <- topTags(lrt.early.silique.batch,n = Inf)$table[topTags(lrt.early.silique.batch,n = Inf)$table$FDR<0.05,]
# # -1  3542
# # 0  53647
# # 1   3786
# detags.early.silique <- rownames(lrt.early.silique.batch)[as.logical(de.early.silique.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.early.silique.png")
# plotSmear(lrt.early.silique.batch, de.tags=detags.early.silique, main="early silique")
# abline(h = c(-2, 2), col = "blue")
# dev.off()
# 
# 
# lrt.late.silique.batch <- glmLRT(fit.parent.batch, contrast = c(0,0,0,0,0,1,0,0,0,0,-1,0))
# topTags(lrt.late.silique.batch)
# summary(de.late.silique.batch <- decideTestsDGE(lrt.late.silique.batch, p=0.05))
# # -1  4064
# # 0  52880
# # 1   4031
# detags.late.silique <- rownames(lrt.late.silique.batch)[as.logical(de.late.silique.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.late.silique.png")
# plotSmear(lrt.late.silique.batch, de.tags=detags.late.silique, main="late silique")
# abline(h = c(-2, 2), col = "blue")
# dev.off()
# 
# lrt.bolting.batch <- glmLRT(fit.parent.batch, coef = "groupDa-al-1bolting")
# topTags(lrt.bolting.batch)
# summary(de.bolting.batch <- decideTestsDGE(lrt.bolting.batch, p=0.05, lfc = 1))
# # -1  1949
# # 0  55971
# # 1   3055
# detags.bolting <- rownames(lrt.bolting.batch)[as.logical(de.bolting.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.bolting.png")
# plotSmear(lrt.bolting.batch, de.tags=detags.bolting, main="bolting")
# abline(h = c(-2, 2), col = "blue")
# dev.off()

## w/o batch effect
fit.parent <- glmFit(dge.data.parent, design.parent)

lrt.young <- glmLRT(fit.parent, contrast = c(0,0,0,0,1,0,0,0,0,-1))
topTags(lrt.young)
summary(de.young <- decideTestsDGE(lrt.young, p=0.05))
DEgene.young <- topTags(lrt.young,n = Inf)$table[topTags(lrt.young,n = Inf)$table$FDR<0.05,]
# -1  4553
# 0  51999
# 1   4423

lrt.flowering <- glmLRT(fit.parent, contrast = c(0,0,1,0,0,0,0,-1,0,0))
topTags(lrt.flowering)
summary(de.flowering <- decideTestsDGE(lrt.flowering, p=0.05))
DEgene.flowering <- topTags(lrt.flowering,n = Inf)$table[topTags(lrt.flowering,n = Inf)$table$FDR<0.05,]
# -1  2877
# 0  54894
# 1   3204

lrt.early.silique <- glmLRT(fit.parent, contrast = c(0,1,0,0,0,0,-1,0,0,0))
topTags(lrt.early.silique)
summary(de.early.silique <- decideTestsDGE(lrt.early.silique, p=0.05))
DEgene.early.silique <- topTags(lrt.early.silique,n = Inf)$table[topTags(lrt.early.silique,n = Inf)$table$FDR<0.05,]
# -1  3200
# 0  54280
# 1   3495

lrt.late.silique <- glmLRT(fit.parent, contrast = c(0,0,0,1,0,0,0,0,-1,0))
topTags(lrt.late.silique)
summary(de.late.silique <- decideTestsDGE(lrt.late.silique, p=0.05))
DEgene.late.silique <- topTags(lrt.late.silique,n = Inf)$table[topTags(lrt.late.silique,n = Inf)$table$FDR<0.05,]
# -1  3745
# 0  53562
# 1   3668

lrt.bolting <- glmLRT(fit.parent, contrast = c(1,0,0,0,0,-1,0,0,0,0))
topTags(lrt.bolting)
summary(de.bolting <- decideTestsDGE(lrt.bolting, p=0.05))
DEgene.bolting <- topTags(lrt.bolting,n = Inf)$table[topTags(lrt.bolting,n = Inf)$table$FDR<0.05,]
# -1  2794
# 0  56327
# 1   1854

### plot smear 
# young
detags.young.1 <- rownames(lrt.young)[as.logical(de.young)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.young.1.png")
plotSmear(lrt.young, de.tags=detags.young.1, main="young")
abline(h = c(-2, 2), col = "blue")
dev.off()

# bolting 
detags.bolting.1 <- rownames(lrt.bolting)[as.logical(de.bolting)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.bolting.1.png")
plotSmear(lrt.bolting, de.tags=detags.bolting.1, main="bolting")
abline(h = c(-2, 2), col = "blue")
dev.off()

# early silique 
detags.early.silique.1 <- rownames(lrt.early.silique)[as.logical(de.early.silique)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.early.silique.1.png")
plotSmear(lrt.early.silique, de.tags=detags.early.silique.1, main="early silique")
abline(h = c(-2, 2), col = "blue")
dev.off()

# late silique 
detags.late.silique.1 <- rownames(lrt.late.silique)[as.logical(de.late.silique)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.late.silique.1.png")
plotSmear(lrt.late.silique, de.tags=detags.late.silique.1, main="late silique")
abline(h = c(-2, 2), col = "blue")
dev.off()

# flowering 
detags.flowering.1 <- rownames(lrt.flowering)[as.logical(de.flowering)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.flowering.1.png")
plotSmear(lrt.flowering, de.tags=detags.flowering.1, main="flowering")
abline(h = c(-2, 2), col = "blue")
dev.off()

```

# GO enrichment analysis 
```{r}
# load all necessary functions 
source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")

# young 
DEgene.young.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.young))
DEgene.young.GO.ORA
write.csv(DEgene.young.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.young.GO.ORA.csv")

# bolting 
DEgene.bolting.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.bolting))
DEgene.bolting.GO.ORA
write.csv(DEgene.bolting.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.bolting.GO.ORA.csv")

# flowering 
DEgene.flowering.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.flowering))
DEgene.flowering.GO.ORA
write.csv(DEgene.flowering.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.flowering.GO.ORA.csv")

# early silique 
DEgene.early.silique.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.early.silique))
DEgene.early.silique.GO.ORA
write.csv(DEgene.early.silique.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.early.silique.GO.ORA.csv")

DEgene.late.silique.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.late.silique))
DEgene.late.silique.GO.ORA
write.csv(DEgene.late.silique.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.late.silique.GO.ORA.csv")
```


