---
title: "KIAT_RNA_seq"
author: "Ruijuan Li"
date: "August 1, 2016"
output: html_document
---

# download data, QC, and mapping (using kallisto)
```{r}
# 1) download file from ftp using ftp_file_transfer.sh # ftp works only under whitney. 
# 1.1) unzip files using gunzip *  
# 2) get first 1000 sequences for each file extract_1000_reads.sh, actually worked with all data instead of the 1st 1000 seqs
# 3) QC file using fastQC: fastqc *.fq -o /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/fastqc_result
# 4) download Brassica_napus.annotation_v5.cds.fa to whitney 
# 5) collapse fastqc summary result into one file & get overrepresented sequences using ./make_summary_report.sh 

# 6) reformat summary.all.txt file 
setwd("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/")
fastQC_summary <- read.delim("./fastQC_result/summary_all.txt", header = F)
dim(fastQC_summary)
head(fastQC_summary)
tail(fastQC_summary)

fastQC_summary_wide <- reshape(fastQC_summary, timevar="V2",idvar="V3",direction="wide")
colnames(fastQC_summary_wide) <- gsub("(V1)(.)([[:print:]]+)","\\3",colnames(fastQC_summary_wide))
dim(fastQC_summary_wide)

write.csv(fastQC_summary_wide, file = "fastQC_summary_wide.csv")

# 7) checked the source of overrepresented sequences using web nucleotide blast tool, found that most of them are from 
# cholorplast, mitochondra, rRNA, tansposon, etc. Some are illumina or TruSeq adapters 

# blast sequences against Trimmomatic adapters suggest that TruSeq3-PE-2.fa should be used to trimm the data. (Q: how to deal with the others? will they affect the final result? search... )

# Qs on plstids seqs contamination: 1) why are there these sequences? check the lib making step 2) will they affect downstream analysis if kept? CDS mapping VS genome mapping 3) If yes, how to remove them? (I need to figure these Qs out.)

# 8) using kallisto to build the reference cds index 
# kallisto index -k 19 -i Brassica_napus.annotation_v5.cds.19.kai /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/Brassica_napus.annotation_v5.cds.fa

# 9) map to reference cds index and get read count file using bunchrun_kallisto.sh (using est_counts from Kallisto)
```

# using STAR 
```{r}
##### STAR is only installed on coloma!!! 
# 1) get B.napus reference genome: Brassica_napus_v4.1.chromosomes.fa 
# make STAR reference index: STAR --runMode genomeGenerate --genomeDir star_nome/ --genomeFastaFiles Brassica_napus_v4.1.chromosomes.fa --runThreadN 2 
# 2) mapping, convert SAM to BAM, and sort BAM file using run_star.sh 
### because of the naming problems of All1_Yeong3_160621 & All1_Yeong2_160621, two seperate STAR mapping were made for them using the below codes: 
# STAR --genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/star_genome --readFilesIn /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/raw_data/All1_Yeong2_160621_1.fq /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/raw_data/All1_Yeong2_160621_2.fq --runThreadN 6

# STAR --genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/star_genome --readFilesIn /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/raw_data/All1_Yeong3_160621_1.fq /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/raw_data/All1_Yeong3_160621_2.fq --runThreadN 6

### 2.2) unstranded data with output type as sorted BAM run_star_unstra.sh
### because of the naming problems of All1_Yeong3_160621 & All1_Yeong2_160621, two seperate STAR mapping were made for them using the below codes: 
# STAR --genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/star_genome --readFilesIn /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/raw_data/All1_Yeong2_160621_1.fq /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/raw_data/All1_Yeong2_160621_2.fq --runThreadN 6 --outSAMstrandField intronMotif --outSAMtype BAM SortedByCoordinate 

# STAR --genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/star_genome --readFilesIn /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/raw_data/All1_Yeong3_160621_1.fq  /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/raw_data/All1_Yeong3_160621_2.fq --runThreadN 6 --outSAMstrandField intronMotif --outSAMtype BAM SortedByCoordinate 

### after consulting Julin, I get to know that we used strand specific lib making protocol.(Brad-seq method) 
### whitney has outdated cufflink version (1.0.1), I should use coloma (version 2.2.1) for assembly. 

# --> to do: use cufflinks for assembly 
# cufflinks -o output/ --library-type ff-firststrand test.sorted (stopped)
# cufflinks -o output2/ --library-type ff-secondstrand test.sorted (stopped)
# cufflinks -o output3/ --library-type fr-firststrand test.sorted (running screen -r 39044: 08/26/2016 started)
# cufflinks -o output4/ --library-type fr-secondstrand test.sorted (running screen -r 39106: 08/26/2016 started, stopped)
# since only with strand information, my code can run, I decide that the lib is with strand information, but which specific one I should use among the 4? 
#  --library-type fr-firststrand should be used for the standard dUTP protocol, including Illumina's stranded Tru-Seq 

# try the same assembly on coloma using updated version of cufflinks 
# cufflinks -o output5/ --library-type fr-firststrand test.sorted (running screen -r 34995.ttys018.coloma: 09/01/2016 started) (keep)
# cufflinks -o output6/ --library-type fr-secondstrand test.sorted (running screen -r 35026.ttys020.coloma: 09/01/2016 started) (stopped)   

### After reading Brad-seq method paper, I realized that only the strand generated during first strand synthesis is sequenced. So the library type should be fr-firststand 

### talked with Julin, he said: use IGV to visualize the sequence alignment to decide library type 
# then after visualize genome, bam file, and gff file, we decide our library type is fr-secondstrand 
#### now running cufflinks using run_star.sh (screen -r 39106.ttys010.coloma)
### also running seperate cufflinks for the 2 special libs 
# screen -r 39584.ttys010.coloma (done)
# screen -r 39131.ttys013.coloma (done)


```

# SNP calling  
```{r}
# index reference genome fasta file 
# samtools faidx Brapa_sequence_v1.5.fa  

# 1) merge libarary for Da-Ae and Da-Ol-1 
# 2) do STAR mapping for Da-Ae & Da-Ol-1 paired-end seq 

# cat Ae*1.fq 5_1.fq 6_1.fq 05-11-Final-8ul_1.fq 8_1.fq > Ae_1.fq
# cat Ae*2.fq 5_2.fq 6_2.fq 05-11-Final-8ul_2.fq 8_2.fq > Ae_2.fq

# cat All*1.fq 1_1.fq 2_1.fq 3_1.fq 4_1.fq > Ol_1.fq
# cat All*2.fq 1_2.fq 2_2.fq 3_2.fq 4_2.fq > Ol_2.fq

### STAR mapping of merged data 
# STAR --genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/star_genome --readFilesIn /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/raw_data/Ae_1.fq /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/raw_data/Ae_2.fq --runThreadN 6 --outSAMtype BAM SortedByCoordinate (screen -r 39044.ttys001.coloma, accidently stopped; screen -r 39044.ttys001.coloma in seperate folder... ) 

# STAR --genomeDir /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/star_genome --readFilesIn /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/raw_data/Ol_1.fq /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/2016_summer/raw_data/Ol_2.fq --runThreadN 6 --outSAMtype BAM SortedByCoordinate (screen -r 39131.ttys013.coloma; screen -r 39584.ttys010.coloma in seperate folder... )  



```

# Expression analysis 1) formatting data   
```{r}
parent.read.count <- read.table("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/read_count/read.count.tsv", header = T, check.names = F)
rownames(parent.read.count) <- parent.read.count[,1]
parent.read.count <- parent.read.count[,-1]

# format data 
head(parent.read.count)
dim(parent.read.count) # 101040     54 
colnames(parent.read.count)
even_indexes<-seq(2,length(colnames(parent.read.count)),2)
parent.read.count.one <- parent.read.count[,even_indexes]
dim(parent.read.count.one) # 101040     27 
colnames(parent.read.count.one)
colnames(parent.read.count.one) <- sub("_2.fq","",colnames(parent.read.count.one),fixed = TRUE) # remove _2.fq 

# sample description  
sample_des <- read.csv("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/parent_summary_corrected.csv")
dim(sample_des) # 27 22
sorted_sample_des <- sample_des[order(sample_des$SampleID),]

sorted_sample_des[,4:7]
new_sample_ID <- paste(sorted_sample_des$Cultivar, sorted_sample_des$Stage, sorted_sample_des$rep, sep = "_")
new_sample_ID

# replace sample ID 
colnames(parent.read.count.one) <- new_sample_ID
head(parent.read.count.one)
# save(parent.read.count.one,file="/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.Rdata")
```

# data summary statisitics read count number & expressed genes 
```{r}
parent.read.count.normalized <- read.table("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/read.count.normalized.tsv", header = T, check.names = F)
rownames(parent.read.count.normalized) <- parent.read.count.normalized[,1]
parent.read.count.normalized <- parent.read.count.normalized[,-1]

# format data 
head(parent.read.count.normalized)
dim(parent.read.count.normalized) # 101040     54 
colnames(parent.read.count.normalized)
even_indexes<-seq(2,length(colnames(parent.read.count.normalized)),2)
parent.read.count.one.normalized <- parent.read.count.normalized[,even_indexes]
dim(parent.read.count.one.normalized) # 101040     27 
colnames(parent.read.count.one.normalized)
colnames(parent.read.count.one.normalized) <- sub("_2.fq","",colnames(parent.read.count.one.normalized),fixed = TRUE) # remove _2.fq 

# sample description  
new_sample_ID

# replace sample ID 
colnames(parent.read.count.one.normalized) <- new_sample_ID
head(parent.read.count.one.normalized)
save(parent.read.count.one.normalized,file="/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.normalized.Rdata")

# check number of genes with reads number > 10 in each lib # make summary table 
expressed_gene_number <- 
sapply(colnames(parent.read.count.one.normalized), function(RNAlib){
  sum(parent.read.count.one.normalized[,RNAlib]>3)
}
)
expressed_gene_number <- as.data.frame(expressed_gene_number)
expressed_gene_number

# top 10 highly expressed genes in each lib 
most_highly_expressed_genes <- 
sapply(colnames((parent.read.count.one.normalized)), function(RNAlib){
  rownames(head(parent.read.count.one.normalized[order(parent.read.count.one.normalized[,RNAlib],decreasing = TRUE),], 10))
}
)
# write.csv(t(most_highly_expressed_genes), file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/most_highly_expressed_genes.csv")

temp <- sample_des[,c(4,5,6,8)]
rownames(temp) <- paste(sample_des[,4], sample_des[,5], sample_des[,6], sep = "_")
sample_statistics <- merge(temp, expressed_gene_number, by=0) 
colnames(sample_statistics) <- c("sample_ID", "Cultivar", "Tissue", "rep", "total_reads", "expressed_gene_number")
head(sample_statistics)
colnames(sample_statistics)
# write.csv(sample_statistics, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/sample_statisitcs.csv")

```


# expression analysis 2) set up sample description 
```{r}
# load necessary libs & functions 
library(edgeR)
library(ggplot2)

# filter based on read count, assign group, normalize, design matrix, calculate dispersion   
# set up group 
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.Rdata")
parent.read.count.one <- parent.read.count.one[,colSums(parent.read.count.one) > 1000000]  
dim(parent.read.count.one) # 101040     27 
parent.read.count.one.sample<-data.frame(file=colnames(parent.read.count.one),
                             batch=factor(gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\5",colnames(parent.read.count.one))),  
                             genotype=factor(gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",colnames(parent.read.count.one))),	
                             stage=factor(gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",colnames(parent.read.count.one))),	
                             group=factor(gsub("(Da-Ae|Da-Ol-1)(_)([[:print:]]+)(_)(1|2|3)","\\1\\3",colnames(parent.read.count.one)))
)

parent.read.count.one.sample
ftable(parent.read.count.one.sample,row.vars="stage",col.vars=c("batch","genotype"))

# filter based on read count 
parent.read.count.one.small <- parent.read.count.one[rowSums(parent.read.count.one > 10) >= 3,]
dim(parent.read.count.one.small) # 60975    27 
# save(parent.read.count.one.small, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.small.Rdata")
```

# expression analysis 3) voom transformation 
```{r}
# voom transformation 
# source("https://bioconductor.org/biocLite.R")
# biocLite("DESeq2")
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.small.Rdata")
library("DESeq2")

dds.parent <- DESeqDataSetFromMatrix(countData = round(parent.read.count.one.small), colData = parent.read.count.one.sample, design = ~ batch + genotype*stage)

vsd.parent <- varianceStabilizingTransformation(dds.parent)
vstMat.parent <- assay(vsd.parent)
colnames(vstMat.parent) <- colnames(parent.read.count.one)
# save(vstMat.parent, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/vstMat.parent.Rdata")
```

# expression analysis 4) MDS, clustering, and expression analysis 
```{r}
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/vstMat.parent.Rdata")
load("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/parent.read.count.one.small.Rdata")

# normalize 
dge.data.parent <- DGEList(counts=parent.read.count.one.small, group=parent.read.count.one.sample$group)
dge.data.parent <- calcNormFactors(dge.data.parent, method = "TMM") 
dge.data.parent$sample
hist(dge.data.parent$samples$norm.factors)

# MDS to check sample seperation, also using dengrogram.  
mds <- plotMDS(dge.data.parent, method = "bcv",labels = dge.data.parent$samples$group)

x <- as.data.frame(mds$x)
y <- as.data.frame(mds$y)
distance_matrix <- merge(x, y, by="row.names")
distance_matrix$group <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1\\3",distance_matrix$Row.names)
distance_matrix$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",distance_matrix$Row.names)
distance_matrix$tissue <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",distance_matrix$Row.names)

colnames(distance_matrix) <- c("lib","x","y","group","gt","tissue")
head(distance_matrix)

# making color MDS figure 
p.mds <- ggplot(data = distance_matrix)
p.mds <- p.mds + geom_point(aes(x, y, color=factor(tissue)), size=3) 
p.mds <- p.mds + labs(y = "BCV distance 2", x="BCV distance 1")
# p.mds <- p.mds + facet_grid(~gt)
p.mds

# ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/MDS.png", width = 11, height = 8)

# clustering to check sample seperation 
library(ggdendro)

hc.parent <- hclust(dist(t(vstMat.parent)))

ggdata.parent <- dendro_data(as.dendrogram(hc.parent))
ggdata.parent$labels$gt <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\1",ggdata.parent$labels$label) 
ggdata.parent$labels$tissue <- gsub("(Da-Ae|Da-Ol-1)(_)(Young|flowering|early-silique|late-silique|bolting)(_)(1|2|3)","\\3",ggdata.parent$labels$label)
ggdata.parent$labels$group <- paste(ggdata.parent$labels$gt, ggdata.parent$labels$tissue, sep = "_")
ggdata.parent$labels

# start ggplot here 
p.dengro <- ggplot(data = segment(ggdata.parent))
p.dendro <- p.dengro + geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) + theme_dendro()
p.dendro <- p.dendro + geom_text(data = label(ggdata.parent), aes(x = x, y = y, label = label, hjust=0, color=group)) + coord_flip() + scale_y_reverse(expand=c(0.2, 0))
p.dendro
# ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/clustering_new.png", width = 15, height = 8) 

# pairwise comparison using GLM model 
# design matrix
# design.parent.batch <- model.matrix(~batch+group, data = parent.read.count.one.sample) 
design.parent <- model.matrix(~0+group, data = parent.read.count.one.sample)

# # estimate dispersion
# # w/ batch effect
# dge.data.parent.batch <- estimateGLMCommonDisp(dge.data.parent.batch, design.parent.batch,verbose = TRUE) # Disp = 0.22238 , BCV = 0.4716
# dge.data.parent.batch <- estimateGLMTrendedDisp(dge.data.parent.batch,design.parent.batch)
# dge.data.parent.batch <- estimateGLMTagwiseDisp(dge.data.parent.batch,design.parent.batch)
# plotBCV(dge.data.parent.batch)

# w/o batch effect
dge.data.parent <- estimateGLMCommonDisp(dge.data.parent, design.parent,verbose = TRUE) # Disp = 0.25572 , BCV = 0.5057
dge.data.parent <- estimateGLMTrendedDisp(dge.data.parent,design.parent)
dge.data.parent <- estimateGLMTagwiseDisp(dge.data.parent,design.parent)
plotBCV(dge.data.parent)

# fit GLM model & test for differential expression
# w/ batch effect
# fit.parent.batch <- glmFit(dge.data.parent.batch, design.parent.batch)
# lrt.young.batch <- glmLRT(fit.parent.batch, contrast = c(0,0,0,0,0,0,1,0,0,0,0,-1))
# topTags(lrt.young.batch)
# DEgene.young.batch <- topTags(lrt.young.batch,n = Inf)$table[topTags(lrt.young.batch,n = Inf)$table$FDR<0.05,]
# summary(de.young.batch <- decideTestsDGE(lrt.young.batch, p=0.05))
# # summary(de.young.batch <- decideTestsDGE(lrt.young.batch, p=0.05, lfc = 2))
# # -1  5072
# # 0  51131
# # 1   4772
# 
# detags.young <- rownames(lrt.young.batch)[as.logical(de.young.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.young.png")
# plotSmear(lrt.young.batch, de.tags=detags.young, main="young")
# abline(h = c(-2, 2), col = "blue")
# dev.off()
# 
# lrt.flowering.batch <- glmLRT(fit.parent.batch, contrast = c(0,0,0,0,1,0,0,0,0,-1,0,0))
# topTags(lrt.flowering.batch)
# summary(de.flowering.batch <- decideTestsDGE(lrt.flowering.batch, p=0.05))
# DEgene.flowering.batch <- topTags(lrt.flowering.batch,n = Inf)$table[topTags(lrt.flowering.batch,n = Inf)$table$FDR<0.05,]
# # -1  3150
# # 0  54342
# # 1   3483
# detags.flowering <- rownames(lrt.flowering.batch)[as.logical(de.flowering.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.flowering.png")
# plotSmear(lrt.flowering.batch, de.tags=detags.flowering, main="flowering")
# abline(h = c(-2, 2), col = "blue")
# dev.off()
# 
# lrt.early.silique.batch <- glmLRT(fit.parent.batch, contrast = c(0,0,0,1,0,0,0,0,-1,0,0,0))
# topTags(lrt.early.silique.batch)
# summary(de.early.silique.batch <- decideTestsDGE(lrt.early.silique.batch, p=0.05))
# DEgene.early.silique.batch <- topTags(lrt.early.silique.batch,n = Inf)$table[topTags(lrt.early.silique.batch,n = Inf)$table$FDR<0.05,]
# # -1  3542
# # 0  53647
# # 1   3786
# detags.early.silique <- rownames(lrt.early.silique.batch)[as.logical(de.early.silique.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.early.silique.png")
# plotSmear(lrt.early.silique.batch, de.tags=detags.early.silique, main="early silique")
# abline(h = c(-2, 2), col = "blue")
# dev.off()
# 
# 
# lrt.late.silique.batch <- glmLRT(fit.parent.batch, contrast = c(0,0,0,0,0,1,0,0,0,0,-1,0))
# topTags(lrt.late.silique.batch)
# summary(de.late.silique.batch <- decideTestsDGE(lrt.late.silique.batch, p=0.05))
# # -1  4064
# # 0  52880
# # 1   4031
# detags.late.silique <- rownames(lrt.late.silique.batch)[as.logical(de.late.silique.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.late.silique.png")
# plotSmear(lrt.late.silique.batch, de.tags=detags.late.silique, main="late silique")
# abline(h = c(-2, 2), col = "blue")
# dev.off()
# 
# lrt.bolting.batch <- glmLRT(fit.parent.batch, coef = "groupDa-al-1bolting")
# topTags(lrt.bolting.batch)
# summary(de.bolting.batch <- decideTestsDGE(lrt.bolting.batch, p=0.05, lfc = 1))
# # -1  1949
# # 0  55971
# # 1   3055
# detags.bolting <- rownames(lrt.bolting.batch)[as.logical(de.bolting.batch)]
# png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.bolting.png")
# plotSmear(lrt.bolting.batch, de.tags=detags.bolting, main="bolting")
# abline(h = c(-2, 2), col = "blue")
# dev.off()

## w/o batch effect
fit.parent <- glmFit(dge.data.parent, design.parent)

lrt.young <- glmLRT(fit.parent, contrast = c(0,0,0,0,1,0,0,0,0,-1))
topTags(lrt.young)
summary(de.young <- decideTestsDGE(lrt.young, p=0.05))
DEgene.young <- topTags(lrt.young,n = Inf)$table[topTags(lrt.young,n = Inf)$table$FDR<0.05,]
# -1  4553
# 0  51999
# 1   4423

lrt.flowering <- glmLRT(fit.parent, contrast = c(0,0,1,0,0,0,0,-1,0,0))
topTags(lrt.flowering)
summary(de.flowering <- decideTestsDGE(lrt.flowering, p=0.05))
DEgene.flowering <- topTags(lrt.flowering,n = Inf)$table[topTags(lrt.flowering,n = Inf)$table$FDR<0.05,]
# -1  2877
# 0  54894
# 1   3204

lrt.early.silique <- glmLRT(fit.parent, contrast = c(0,1,0,0,0,0,-1,0,0,0))
topTags(lrt.early.silique)
summary(de.early.silique <- decideTestsDGE(lrt.early.silique, p=0.05))
DEgene.early.silique <- topTags(lrt.early.silique,n = Inf)$table[topTags(lrt.early.silique,n = Inf)$table$FDR<0.05,]
# -1  3200
# 0  54280
# 1   3495

lrt.late.silique <- glmLRT(fit.parent, contrast = c(0,0,0,1,0,0,0,0,-1,0))
topTags(lrt.late.silique)
summary(de.late.silique <- decideTestsDGE(lrt.late.silique, p=0.05))
DEgene.late.silique <- topTags(lrt.late.silique,n = Inf)$table[topTags(lrt.late.silique,n = Inf)$table$FDR<0.05,]
# -1  3745
# 0  53562
# 1   3668

lrt.bolting <- glmLRT(fit.parent, contrast = c(1,0,0,0,0,-1,0,0,0,0))
topTags(lrt.bolting)
summary(de.bolting <- decideTestsDGE(lrt.bolting, p=0.05))
DEgene.bolting <- topTags(lrt.bolting,n = Inf)$table[topTags(lrt.bolting,n = Inf)$table$FDR<0.05,]
# -1  2794
# 0  56327
# 1   1854

### barplot for DEGs number 
### reformat for ggplot barplot 
DEGs_number_between_gt.parent <- data.frame(young = as.data.frame(summary(de.young <- decideTestsDGE(lrt.young, p=0.05)))$Freq,
                                            bolting = as.data.frame(summary(de.bolting <- decideTestsDGE(lrt.bolting, p=0.05)))$Freq,
                                     flowering = as.data.frame(summary(de.flowering <- decideTestsDGE(lrt.flowering, p=0.05)))$Freq, 
                                     early.silique = as.data.frame(summary(de.early.silique <- decideTestsDGE(lrt.early.silique, p=0.05)))$Freq,
                                     late.silique = as.data.frame(summary(de.late.silique <- decideTestsDGE(lrt.late.silique, p=0.05)))$Freq)

rownames(DEGs_number_between_gt.parent) <- c("down", "no", "up")
DEGs_number_between_gt.parent <- DEGs_number_between_gt.parent[c("down", "up"),]
DEGs_number_between_gt.parent

library(reshape2)
DEGs_number_between_gt.melt.parent <- melt(DEGs_number_between_gt.parent)
DEGs_number_between_gt.melt.parent$DE <- rep(c("down", "up"), 5)
colnames(DEGs_number_between_gt.melt.parent) <- c("tissue", "number", "DE")
DEGs_number_between_gt.melt.parent

# reorder: up 1st down 2nd 
DEGs_number_between_gt.melt.parent$DE <- factor(DEGs_number_between_gt.melt.parent$DE, levels = c("up", "down"))

DEGs_number_between_gt.melt.parent <- DEGs_number_between_gt.melt.parent[order(DEGs_number_between_gt.melt.parent$DE),]
DEGs_number_between_gt.melt.parent

### making ggplot for DEGs 
p.DEGs_number_between_gt.parent <- ggplot(data = DEGs_number_between_gt.melt.parent)
p.DEGs_number_between_gt.parent <- p.DEGs_number_between_gt.parent + geom_bar(mapping = aes(fill=DE, x = factor(DE), y = number), stat = "identity")
p.DEGs_number_between_gt.parent <- p.DEGs_number_between_gt.parent + facet_grid(~tissue)
p.DEGs_number_between_gt.parent <- p.DEGs_number_between_gt.parent + labs(y = "number of differentially expressed genes", x = "")

p.DEGs_number_between_gt.parent
# ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/DEGs_number_between_gt.parent.png", width = 11, height = 8)

## between developmental stages 
# Da-Ae 
lrt.young.vs.bolting.ae <- glmLRT(fit.parent, contrast = c(1,0,0,0,-1,0,0,0,0,0))
topTags(lrt.young.vs.bolting.ae)
summary(de.young.vs.bolting.ae <- decideTestsDGE(lrt.young.vs.bolting.ae, p=0.05))
DEgene.young.vs.bolting.ae <- topTags(lrt.young.vs.bolting.ae,n = Inf)$table[topTags(lrt.young.vs.bolting.ae,n = Inf)$table$FDR<0.05,]
# -1   442
# 0  59872
# 1    661

lrt.bolting.vs.flowering.ae <- glmLRT(fit.parent, contrast = c(-1,0,1,0,0,0,0,0,0,0))
topTags(lrt.bolting.vs.flowering.ae)
summary(de.bolting.vs.flowering.ae <- decideTestsDGE(lrt.bolting.vs.flowering.ae, p=0.05))
DEgene.bolting.vs.flowering.ae <- topTags(lrt.bolting.vs.flowering.ae,n = Inf)$table[topTags(lrt.bolting.vs.flowering.ae,n = Inf)$table$FDR<0.05,]
# -1   736
# 0  57389
# 1   2850

lrt.flowering.vs.early.silique.ae <- glmLRT(fit.parent, contrast = c(0,1,-1,0,0,0,0,0,0,0))
topTags(lrt.flowering.vs.early.silique.ae)
summary(de.flowering.vs.early.silique.ae <- decideTestsDGE(lrt.flowering.vs.early.silique.ae, p=0.05))
DEgene.flowering.vs.early.silique.ae <- topTags(lrt.flowering.vs.early.silique.ae,n = Inf)$table[topTags(lrt.flowering.vs.early.silique.ae,n = Inf)$table$FDR<0.05,]
# -1  3811
# 0  56054
# 1   1110

lrt.early.vs.late.silique.ae <- glmLRT(fit.parent, contrast = c(0,-1,0,1,0,0,0,0,0,0))
topTags(lrt.early.vs.late.silique.ae)
summary(de.early.vs.late.silique.ae <- decideTestsDGE(lrt.early.vs.late.silique.ae, p=0.05))
DEgene.early.vs.late.silique.ae <- topTags(lrt.early.vs.late.silique.ae,n = Inf)$table[topTags(lrt.early.vs.late.silique.ae,n = Inf)$table$FDR<0.05,]
# -1   285
# 0  60430
# 1    260

### ggplot 
### reformat for ggplot barplot 
DEGs_number_between_tissue.ae <- data.frame("bolting-vs-young" = as.data.frame(summary(de.young.vs.bolting.ae <- decideTestsDGE(lrt.young.vs.bolting.ae, p=0.05)))$Freq, 
                                            "flowering-vs-bolting" = as.data.frame(summary(de.bolting.vs.flowering.ae <- decideTestsDGE(lrt.bolting.vs.flowering.ae, p=0.05)))$Freq,
                                     "early.slique-vs-flowering"= as.data.frame(summary(de.flowering.vs.early.silique.ae <- decideTestsDGE(lrt.flowering.vs.early.silique.ae, p=0.05)))$Freq, 
                                     "late.silique-vs-early.silique" = as.data.frame(summary(de.early.vs.late.silique.ae <- decideTestsDGE(lrt.early.vs.late.silique.ae, p=0.05)))$Freq)

rownames(DEGs_number_between_tissue.ae) <- c("down", "no", "up")
DEGs_number_between_tissue.ae <- DEGs_number_between_tissue.ae[c("down", "up"),]
DEGs_number_between_tissue.ae

DEGs_number_between_tissue.ae.melt.parent <- melt(DEGs_number_between_tissue.ae)
DEGs_number_between_tissue.ae.melt.parent$DE <- rep(c("down", "up"), 4)
colnames(DEGs_number_between_tissue.ae.melt.parent) <- c("tissue", "number", "DE")
DEGs_number_between_tissue.ae.melt.parent

# reorder: up 1st down 2nd 
DEGs_number_between_tissue.ae.melt.parent$DE <- factor(DEGs_number_between_tissue.ae.melt.parent$DE, levels = c("up", "down"))

DEGs_number_between_tissue.ae.melt.parent <- DEGs_number_between_tissue.ae.melt.parent[order(DEGs_number_between_tissue.ae.melt.parent$DE),]
DEGs_number_between_tissue.ae.melt.parent

### making ggplot for DEGs 
p.DEGs_number_between_tissue.ae.parent <- ggplot(data = DEGs_number_between_tissue.ae.melt.parent)
p.DEGs_number_between_tissue.ae.parent <- p.DEGs_number_between_tissue.ae.parent + geom_bar(mapping = aes(fill=DE, x = factor(DE), y = number), stat = "identity")
p.DEGs_number_between_tissue.ae.parent <- p.DEGs_number_between_tissue.ae.parent + facet_grid(~tissue)
p.DEGs_number_between_tissue.ae.parent <- p.DEGs_number_between_tissue.ae.parent + labs(y = "number of differentially expressed genes", x = "")

p.DEGs_number_between_tissue.ae.parent
# ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/DEGs_number_between_tissue.ae.parent.png", width = 11, height = 8)

### Da-Ol-1 
lrt.young.vs.bolting.ol <- glmLRT(fit.parent, contrast = c(0,0,0,0,0,1,0,0,0,-1))
topTags(lrt.young.vs.bolting.ol)
summary(de.young.vs.bolting.ol <- decideTestsDGE(lrt.young.vs.bolting.ol, p=0.05))
DEgene.young.vs.bolting.ol <- topTags(lrt.young.vs.bolting.ol,n = Inf)$table[topTags(lrt.young.vs.bolting.ol,n = Inf)$table$FDR<0.05,]
# -1  2683
# 0  54969
# 1   3323

lrt.bolting.vs.flowering.ol <- glmLRT(fit.parent, contrast = c(0,0,0,0,0,-1,0,1,0,0))
topTags(lrt.bolting.vs.flowering.ol)
summary(de.bolting.vs.flowering.ol <- decideTestsDGE(lrt.bolting.vs.flowering.ol, p=0.05))
DEgene.bolting.vs.flowering.ol <- topTags(lrt.bolting.vs.flowering.ol,n = Inf)$table[topTags(lrt.bolting.vs.flowering.ol,n = Inf)$table$FDR<0.05,]
# -1   914
# 0  58313
# 1   1748

lrt.flowering.vs.early.silique.ol <- glmLRT(fit.parent, contrast = c(0,0,0,0,0,0,1,-1,0,0))
topTags(lrt.flowering.vs.early.silique.ol)
summary(de.flowering.vs.early.silique.ol <- decideTestsDGE(lrt.flowering.vs.early.silique.ol, p=0.05))
DEgene.flowering.vs.early.silique.ol <- topTags(lrt.flowering.vs.early.silique.ol,n = Inf)$table[topTags(lrt.flowering.vs.early.silique.ol,n = Inf)$table$FDR<0.05,]
# -1  4845
# 0  54087
# 1   2043

lrt.early.vs.late.silique.ol <- glmLRT(fit.parent, contrast = c(0,0,0,0,0,0,-1,0,1,0))
topTags(lrt.early.vs.late.silique.ol)
summary(de.early.vs.late.silique.ol <- decideTestsDGE(lrt.early.vs.late.silique.ol, p=0.05))
DEgene.early.vs.late.silique.ol <- topTags(lrt.early.vs.late.silique.ol,n = Inf)$table[topTags(lrt.early.vs.late.silique.ol,n = Inf)$table$FDR<0.05,]
# -1  1042
# 0  58199
# 1   1734

### ggplot 
### reformat for ggplot barplot 
DEGs_number_between_tissue.ol <- data.frame("bolting-vs-young" = as.data.frame(summary(de.young.vs.bolting.ol <- decideTestsDGE(lrt.young.vs.bolting.ol, p=0.05)))$Freq, 
                                            "flowering-vs-bolting" = as.data.frame(summary(de.bolting.vs.flowering.ol <- decideTestsDGE(lrt.bolting.vs.flowering.ol, p=0.05)))$Freq,
                                     "early.slique-vs-flowering"= as.data.frame(summary(de.flowering.vs.early.silique.ol <- decideTestsDGE(lrt.flowering.vs.early.silique.ol, p=0.05)))$Freq, 
                                     "late.silique-vs-early.silique" = as.data.frame(summary(de.early.vs.late.silique.ol <- decideTestsDGE(lrt.early.vs.late.silique.ol, p=0.05)))$Freq)

rownames(DEGs_number_between_tissue.ol) <- c("down", "no", "up")
DEGs_number_between_tissue.ol <- DEGs_number_between_tissue.ol[c("down", "up"),]
DEGs_number_between_tissue.ol

DEGs_number_between_tissue.ol.melt.parent <- melt(DEGs_number_between_tissue.ol)
DEGs_number_between_tissue.ol.melt.parent$DE <- rep(c("down", "up"), 4)
colnames(DEGs_number_between_tissue.ol.melt.parent) <- c("tissue", "number", "DE")
DEGs_number_between_tissue.ol.melt.parent

# reorder: up 1st down 2nd 
DEGs_number_between_tissue.ol.melt.parent$DE <- factor(DEGs_number_between_tissue.ol.melt.parent$DE, levels = c("up", "down"))

DEGs_number_between_tissue.ol.melt.parent <- DEGs_number_between_tissue.ol.melt.parent[order(DEGs_number_between_tissue.ol.melt.parent$DE),]
DEGs_number_between_tissue.ol.melt.parent

### making ggplot for DEGs 
p.DEGs_number_between_tissue.ol.parent <- ggplot(data = DEGs_number_between_tissue.ol.melt.parent)
p.DEGs_number_between_tissue.ol.parent <- p.DEGs_number_between_tissue.ol.parent + geom_bar(mapping = aes(fill=DE, x = factor(DE), y = number), stat = "identity")
p.DEGs_number_between_tissue.ol.parent <- p.DEGs_number_between_tissue.ol.parent + facet_grid(~tissue)
p.DEGs_number_between_tissue.ol.parent <- p.DEGs_number_between_tissue.ol.parent + labs(y = "number of differentially expressed genes", x = "")

p.DEGs_number_between_tissue.ol.parent
# ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/DEGs_number_between_tissue.ol.parent.png", width = 11, height = 8) 


### plot smear 
# young
detags.young.1 <- rownames(lrt.young)[as.logical(de.young)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.young.1.png")
plotSmear(lrt.young, de.tags=detags.young.1, main="young")
abline(h = c(-2, 2), col = "blue")
dev.off()

# bolting 
detags.bolting.1 <- rownames(lrt.bolting)[as.logical(de.bolting)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.bolting.1.png")
plotSmear(lrt.bolting, de.tags=detags.bolting.1, main="bolting")
abline(h = c(-2, 2), col = "blue")
dev.off()

# early silique 
detags.early.silique.1 <- rownames(lrt.early.silique)[as.logical(de.early.silique)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.early.silique.1.png")
plotSmear(lrt.early.silique, de.tags=detags.early.silique.1, main="early silique")
abline(h = c(-2, 2), col = "blue")
dev.off()

# late silique 
detags.late.silique.1 <- rownames(lrt.late.silique)[as.logical(de.late.silique)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.late.silique.1.png")
plotSmear(lrt.late.silique, de.tags=detags.late.silique.1, main="late silique")
abline(h = c(-2, 2), col = "blue")
dev.off()

# flowering 
detags.flowering.1 <- rownames(lrt.flowering)[as.logical(de.flowering)]
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/detags.flowering.1.png")
plotSmear(lrt.flowering, de.tags=detags.flowering.1, main="flowering")
abline(h = c(-2, 2), col = "blue")
dev.off()

# plot histogram 
### 

```

# explore the silique problem, much fewer genes were differentially expressed between early & late silique in Da-Ae than in Da-Ol-1, why? (actually De-Ae has fewer DEGs between all stages, need to find out why later) 
```{r}
# calcuate overlaps & unique genes 
dim(DEgene.young.vs.bolting.ol) # 6006    5 
dim(DEgene.young.vs.bolting.ae) # 1103    5
summary(rownames(DEgene.young.vs.bolting.ol) %in% rownames(DEgene.young.vs.bolting.ae)) 
# Mode   FALSE    TRUE    NA's 
# logical    5556     450       0 
dim(DEgene.bolting.vs.flowering.ol) # 2662    5 
dim(DEgene.bolting.vs.flowering.ae) # 3586    5 
summary(rownames(DEgene.bolting.vs.flowering.ol) %in% rownames(DEgene.bolting.vs.flowering.ae)) 
#    Mode   FALSE    TRUE    NA's 
# logical    1528    1134       0
dim(DEgene.flowering.vs.early.silique.ol) # 6888    5 
dim(DEgene.flowering.vs.early.silique.ae) # 4921    5 
summary(rownames(DEgene.flowering.vs.early.silique.ol) %in% rownames(DEgene.flowering.vs.early.silique.ae)) 
#   Mode   FALSE    TRUE    NA's 
# logical    3200    3688       0 
dim(DEgene.early.vs.late.silique.ol) # 2776 5 
dim(DEgene.early.vs.late.silique.ae) # 545 5 
summary(rownames(DEgene.early.vs.late.silique.ol) %in% rownames(DEgene.early.vs.late.silique.ae)) 
#    Mode   FALSE    TRUE    NA's 
# logical    2436     340       0 

# draw venn diagram 
# install.packages("VennDiagram")
require(VennDiagram)

# young vs bolting
grid.newpage()
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/young.vs.bolting.png")
venn.plot <- draw.pairwise.venn(6006, 1103, 450, # 1st number for group A, 2nd for group B, 3rd for overlaps 
                                c("Da-Ol-1", "Da-Ae"),
                                fill = c("lightblue","pink"),
                                lty          = "blank",
                                cex = 2,
                                cat.cex = 1.5,
                                cat.pos = 9)
grid.draw(venn.plot)
dev.off()

# bolting vs flowering 
grid.newpage()
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/bolting.vs.flowering.png")
venn.plot <- draw.pairwise.venn(2662, 3586, 1134, # 1st number for group A, 2nd for group B, 3rd for overlaps 
                                c("Da-Ol-1", "Da-Ae"),
                                fill = c("lightblue","pink"),
                                lty          = "blank",
                                cex = 2,
                                cat.cex = 1.5,
                                cat.pos = 9)
grid.draw(venn.plot)
dev.off()

# flowering vs early silique 
grid.newpage()
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/flowering.vs.early.silique.png")
venn.plot <- draw.pairwise.venn(6888, 4921, 3688, # 1st number for group A, 2nd for group B, 3rd for overlaps 
                                c("Da-Ol-1", "Da-Ae"),
                                fill = c("lightblue","pink"),
                                lty          = "blank",
                                cex = 2,
                                cat.cex = 1.5,
                                cat.pos = 9)
grid.draw(venn.plot)
dev.off()

# early vs late silique 
grid.newpage()
png(filename = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/early.vs.late.silique.png")
venn.plot <- draw.pairwise.venn(2776, 545, 340, # 1st number for group A, 2nd for group B, 3rd for overlaps 
                                c("Da-Ol-1", "Da-Ae"),
                                fill = c("lightblue","pink"),
                                lty          = "blank",
                                cex = 2,
                                cat.cex = 1.5,
                                cat.pos = 9)
grid.draw(venn.plot)
dev.off()

# explore the uniquely differentially expressed genes in Da-Ol-1, what are their expression patterns in Da-Ae? 
# 1) find genes that are only differentially expressed in Da-Ol-1 
index <- which(!(rownames(DEgene.early.vs.late.silique.ol) %in% rownames(DEgene.early.vs.late.silique.ae)))

# 2) extract the expression values of these genes from early, late silique of Da-Ae & Da-Ol-1 
uniq.gene.ol <- parent.read.count.one.small[index,]
uniq.gene.ol <- uniq.gene.ol[,c("Da-Ae_early-silique_3", "Da-Ol-1_early-silique_3", "Da-Ol-1_late-silique_3", "Da-Ae_late-silique_3", "Da-Ae_early-silique_1", "Da-Ae_early-silique_2", "Da-Ae_late-silique_1", "Da-Ae_late-silique_2", "Da-Ol-1_early-silique_1", "Da-Ol-1_early-silique_2", "Da-Ol-1_late-silique_1", "Da-Ol-1_late-silique_2")]

dim(uniq.gene.ol) # 2436   12 
```

# fatty acid biosynthesis pathways genes (GO:0006633)
```{r}
fatty_acid_gene_index <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/fatty_acid_biosynthesis_genes_ID.txt", header = F)
head(fatty_acid_gene_index)
nrow(fatty_acid_gene_index)
rownames(fatty_acid_gene_index) <- fatty_acid_gene_index$V1

# extract only fatty acid genes from dataset 
fatty.acid.genes <- (merge(parent.read.count.one.small, fatty_acid_gene_index, by="row.names"))[,1:28]
head(fatty.acid.genes)
rownames(fatty.acid.genes) <- fatty.acid.genes$Row.names
fatty.acid.genes <- fatty.acid.genes[,-1]
dim(fatty.acid.genes) # 207 27
fatty.acid.genes$total <- rowSums(fatty.acid.genes)
dim(fatty.acid.genes)
```

# new design to see how many fatty acid genes' expression are differentially expressed between genotypes and during developmental stages  
```{r}
source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")
# assign group 
ftable(parent.read.count.one.sample,row.vars=c("stage"),col.vars=c("batch","genotype"))

# normalize 
dge.new <- DGEList(counts=parent.read.count.one.small, group=parent.read.count.one.sample$group)
dim(dge.new) # [1] 30876    28 
dge.new <- calcNormFactors(dge.new, method = "TMM") 
dge.new$sample
# design matrix 
parent.read.count.one.sample$genotype <- as.factor(parent.read.count.one.sample$genotype)
parent.read.count.one.sample$tissue <- as.factor(parent.read.count.one.sample$stage)
parent.read.count.one.sample$genotype <- relevel(parent.read.count.one.sample$genotype,ref="Da-Ol-1")
parent.read.count.one.sample$tissue <- relevel(parent.read.count.one.sample$tissue,ref="Young")

design.new <- model.matrix(~tissue*genotype,data = parent.read.count.one.sample) 
colnames(design.new)

# calculate dispersion
dge.new <- estimateGLMCommonDisp(dge.new, design.new,verbose = TRUE) # Disp = 0.25646 , BCV = 0.5064 
dge.new <- estimateGLMTrendedDisp(dge.new,design.new)
dge.new <- estimateGLMTagwiseDisp(dge.new,design.new)
# plotBCV(dge.new)

# use interaction as coefficient and see how many fatty acid genes are important for interaction term... 
fit.new <- glmFit(dge.new, design.new)
lrt.new.interaction <- glmLRT(fit.new,coef = c("tissuebolting:genotypeDa-Ae", "tissueearly-silique:genotypeDa-Ae", "tissueflowering:genotypeDa-Ae", "tissuelate-silique:genotypeDa-Ae"))
topTags(lrt.new.interaction)
DEgene.new.interaction <- topTags(lrt.new.interaction,n = Inf)$table[topTags(lrt.new.interaction,n = Inf)$table$FDR<0.05,]
nrow(DEgene.new.interaction) # number of genes with interaction effect in any tissue types 

dim(fatty.acid.genes) # 207 fatty acid genes expressed in my data 

index_2 <- which(rownames(fatty.acid.genes) %in% rownames(DEgene.new.interaction)) 
rownames(fatty.acid.genes[index_2,])
ID.15.fatty.acid.genes <- rownames(fatty.acid.genes[which(rownames(fatty.acid.genes) %in% rownames(DEgene.new.interaction)),])
fatty.acid.genes.annotation <- read.csv("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/fatty_acid_gene_name.csv", header = F)

expression.pattern.Bn.parent.with.annot(ID.15.fatty.acid.genes, fatty.acid.genes.annotation)

# ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/fatty.acid.interaction.png", width = 8, height = 20)

# get only 3 genes 
ID <- c("BnaA05g06830D","BnaA10g09300D",  "BnaC09g50630D") 
expression.pattern.Bn.parent.with.annot(ID, fatty.acid.genes.annotation)

# ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/figure/fatty.acid.three.png", width = 11, height = 8)
```

# GO enrichment analysis 
```{r}
# load all necessary functions 
# source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")

# young 
DEgene.young.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.young))
DEgene.young.GO.ORA
write.csv(DEgene.young.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.young.GO.ORA.csv")

# bolting 
DEgene.bolting.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.bolting))
DEgene.bolting.GO.ORA
write.csv(DEgene.bolting.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.bolting.GO.ORA.csv")

# flowering 
DEgene.flowering.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.flowering))
DEgene.flowering.GO.ORA
write.csv(DEgene.flowering.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.flowering.GO.ORA.csv")

# early silique 
DEgene.early.silique.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.early.silique))
DEgene.early.silique.GO.ORA
write.csv(DEgene.early.silique.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.early.silique.GO.ORA.csv")

DEgene.late.silique.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.late.silique))
DEgene.late.silique.GO.ORA
write.csv(DEgene.late.silique.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.late.silique.GO.ORA.csv")

DEgene.silique.ae.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.early.vs.late.silique.ae))
DEgene.silique.ae.GO.ORA
write.csv(DEgene.late.silique.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.silique.ae.GO.ORA.csv")

DEgene.silique.ol.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.early.vs.late.silique.ol))
DEgene.silique.ol.GO.ORA
write.csv(DEgene.silique.ol.GO.ORA, file = "/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/DEgene.silique.ol.GO.ORA.csv")

DEgene.young.vs.bolting.ol.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.young.vs.bolting.ol))
DEgene.young.vs.bolting.ol.GO.ORA

DEgene.young.vs.bolting.ae.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.young.vs.bolting.ae))
DEgene.young.vs.bolting.ae.GO.ORA

DEgene.bolting.vs.flowering.ol.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.bolting.vs.flowering.ol))
DEgene.bolting.vs.flowering.ol.GO.ORA

DEgene.bolting.vs.flowering.ae.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.bolting.vs.flowering.ae))
DEgene.bolting.vs.flowering.ae.GO.ORA

DEgene.flowering.vs.early.silique.ol.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.flowering.vs.early.silique.ol))
DEgene.bolting.vs.flowering.ol.GO.ORA

DEgene.flowering.vs.early.silique.ae.GO.ORA <- GOseq.Bn.ORA(rownames(DEgene.flowering.vs.early.silique.ae))
DEgene.bolting.vs.flowering.ae.GO.ORA
```

### other important genes 
# Acetyl-CoA-ACP-transacetylase: GO:0004313 none 
# B-Ketoacyl-ACP-synthase: IPR014030 --> A 
# Malonyl-CoA-ACP-transacetylase: IPR016036 --> B  
# B-ketoacyl-ACP-reductase: GO:0004316 --> C 
# Enoyl-ACP-hydrase: IPR018376 --> D
# Enoyl-ACP-reductase: GO:0016631 & GO:0004319 none 
# Palmitoyl-thioesterase: IPR002472 --> E 
```{r}
source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")

# B-Ketoacyl-ACP-synthase  
A.ID <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/B-Ketoacyl-ACP-synthase", header = F)
dim(A.ID)
expression.pattern.Bn.parent(A.ID) 
# ggsave("~/Desktop/Brassica_project/KIAT_RNA_seq/figure/A.png", width = 8, height = 20) 

# Malonyl-CoA-ACP-transacetylase  
B.ID <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/Malonyl-CoA-ACP-transacetylase", header = F) 
dim(B.ID)
expression.pattern.Bn.parent(B.ID)
# ggsave("~/Desktop/Brassica_project/KIAT_RNA_seq/figure/B.png", width = 8, height = 10) 

# B-ketoacyl-ACP-reductase: GO:0004316 --> C 
C.ID <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/B-ketoacyl-ACP-reductase", header = F) 
dim(C.ID)
expression.pattern.Bn.parent(C.ID)
# ggsave("~/Desktop/Brassica_project/KIAT_RNA_seq/figure/C.png", width = 8, height = 8) 

# Enoyl-ACP-hydrase: IPR018376 --> D
D.ID <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/Enoyl-ACP-hydrase", header = F)
dim(D.ID)
expression.pattern.Bn.parent(D.ID)
# ggsave("~/Desktop/Brassica_project/KIAT_RNA_seq/figure/D.png", width = 8, height = 20) 

# Palmitoyl-thioesterase: IPR002472 --> E 
E.ID <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/data/Palmitoyl-thioesterase", header = F) 
dim(E.ID)
expression.pattern.Bn.parent(E.ID)  
# ggsave("~/Desktop/Brassica_project/KIAT_RNA_seq/figure/E.png", width = 8, height = 20) 
```

# Q: whether 1cm/early or 5cm/late silique is the better choice for RNAseq for looking at differences in fatty acid synthesis. 
```{r}
# 1) used GO# & IPR# related to different fatty acid pathways as keywords to search for gene IDs the reason that I didn't use Arabidopsis blast results is: the inconsistency between the annotation of gene list from FnP & the annotation in Arabidopsis dataset. 

# data 
# gene list related to fatty acid biosynthesis 
rownames(fatty.acid.genes) # fatty acid biosynthesis pathways genes (GO:0006633) 
A.ID # B-Ketoacyl-ACP-synthase: IPR014030
B.ID # Malonyl-CoA-ACP-transacetylase: IPR016036 
C.ID # B-ketoacyl-ACP-reductase: GO:0004316
D.ID # Enoyl-ACP-hydrase: IPR018376
E.ID # Palmitoyl-thioesterase: IPR002472 

length(unique(c(rownames(fatty.acid.genes), as.character(A.ID$V1), as.character(B.ID$V1), as.character(C.ID$V1), as.character(D.ID$V1), as.character(E.ID$V1)))) # 247 genes in total 

# all the expression result 
DEgene.early.vs.late.silique.ae
DEgene.early.vs.late.silique.ol
DEgene.early.silique
DEgene.late.silique

# extract fold change value & p-value from DEG dataset 
# between early & late silique 
DEgene.early.vs.late.silique.ae[rownames(DEgene.early.vs.late.silique.ae) %in% rownames(fatty.acid.genes),]
DEgene.early.vs.late.silique.ae[rownames(DEgene.early.vs.late.silique.ae) %in% A.ID$V1,]
DEgene.early.vs.late.silique.ae[rownames(DEgene.early.vs.late.silique.ae) %in% B.ID$V1,]
DEgene.early.vs.late.silique.ae[rownames(DEgene.early.vs.late.silique.ae) %in% C.ID$V1,]
DEgene.early.vs.late.silique.ae[rownames(DEgene.early.vs.late.silique.ae) %in% D.ID$V1,]
DEgene.early.vs.late.silique.ae[rownames(DEgene.early.vs.late.silique.ae) %in% E.ID$V1,]

gene.1 <- DEgene.early.vs.late.silique.ae[rownames(DEgene.early.vs.late.silique.ae) %in% rownames(fatty.acid.genes),]
gene.1
# only one fatty acid related gene is differentially expressed between early & late silique for Da-Ae 

DEgene.early.vs.late.silique.ol[rownames(DEgene.early.vs.late.silique.ol) %in% rownames(fatty.acid.genes),] 
DEgene.early.vs.late.silique.ol[rownames(DEgene.early.vs.late.silique.ol) %in% A.ID$V1,]
DEgene.early.vs.late.silique.ol[rownames(DEgene.early.vs.late.silique.ol) %in% B.ID$V1,]
DEgene.early.vs.late.silique.ol[rownames(DEgene.early.vs.late.silique.ol) %in% C.ID$V1,]
DEgene.early.vs.late.silique.ol[rownames(DEgene.early.vs.late.silique.ol) %in% D.ID$V1,]
DEgene.early.vs.late.silique.ol[rownames(DEgene.early.vs.late.silique.ol) %in% E.ID$V1,]

oil.synthesis.related.early.vs.late.ol <- unique(rbind(
  DEgene.early.vs.late.silique.ol[rownames(DEgene.early.vs.late.silique.ol) %in% rownames(fatty.acid.genes),], 
  DEgene.early.vs.late.silique.ol[rownames(DEgene.early.vs.late.silique.ol) %in% A.ID$V1,],
  DEgene.early.vs.late.silique.ol[rownames(DEgene.early.vs.late.silique.ol) %in% B.ID$V1,],
  DEgene.early.vs.late.silique.ol[rownames(DEgene.early.vs.late.silique.ol) %in% C.ID$V1,],
  DEgene.early.vs.late.silique.ol[rownames(DEgene.early.vs.late.silique.ol) %in% D.ID$V1,],
  DEgene.early.vs.late.silique.ol[rownames(DEgene.early.vs.late.silique.ol) %in% E.ID$V1,]
))

gene.9 <- oil.synthesis.related.early.vs.late.ol # 10 genes are differentially expressed between early & late silique for Da-Ol-1 

# between two genotypes at early & late silique stages 
DEgene.early.silique[rownames(DEgene.early.silique) %in% rownames(fatty.acid.genes),]
DEgene.early.silique[(rownames(DEgene.early.silique) %in% A.ID$V1),]
DEgene.early.silique[(rownames(DEgene.early.silique) %in% B.ID$V1),]
DEgene.early.silique[(rownames(DEgene.early.silique) %in% C.ID$V1),]
DEgene.early.silique[(rownames(DEgene.early.silique) %in% D.ID$V1),]
DEgene.early.silique[(rownames(DEgene.early.silique) %in% E.ID$V1),]

oil.synthesis.related.early.silique <- unique(rbind(
  DEgene.early.silique[rownames(DEgene.early.silique) %in% rownames(fatty.acid.genes),],
  DEgene.early.silique[(rownames(DEgene.early.silique) %in% A.ID$V1),],
  DEgene.early.silique[(rownames(DEgene.early.silique) %in% B.ID$V1),],
  DEgene.early.silique[(rownames(DEgene.early.silique) %in% C.ID$V1),],
  DEgene.early.silique[(rownames(DEgene.early.silique) %in% D.ID$V1),],
  DEgene.early.silique[(rownames(DEgene.early.silique) %in% E.ID$V1),]
))

gene.28 <- oil.synthesis.related.early.silique # 28 genes are differentially expressed between Da-Ae & Da-Ol-1 at early silique stage 
gene.28
# write.table(gene.28, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/data/gene.28.txt")

DEgene.late.silique[rownames(DEgene.late.silique) %in% rownames(fatty.acid.genes),]
DEgene.late.silique[(rownames(DEgene.late.silique) %in% A.ID$V1),]
DEgene.late.silique[(rownames(DEgene.late.silique) %in% B.ID$V1),]
DEgene.late.silique[(rownames(DEgene.late.silique) %in% C.ID$V1),]
DEgene.late.silique[(rownames(DEgene.late.silique) %in% D.ID$V1),]
DEgene.late.silique[(rownames(DEgene.late.silique) %in% E.ID$V1),]

oil.synthesis.related.late.silique <- unique(rbind(
  DEgene.late.silique[rownames(DEgene.late.silique) %in% rownames(fatty.acid.genes),],
  DEgene.late.silique[(rownames(DEgene.late.silique) %in% A.ID$V1),],
  DEgene.late.silique[(rownames(DEgene.late.silique) %in% B.ID$V1),],
  DEgene.late.silique[(rownames(DEgene.late.silique) %in% C.ID$V1),],
  DEgene.late.silique[(rownames(DEgene.late.silique) %in% D.ID$V1),],
  DEgene.late.silique[(rownames(DEgene.late.silique) %in% E.ID$V1),]
))

gene.36 <- oil.synthesis.related.late.silique # 36 genes are differentially expressed between Da-Ae & Da-Ol-1 at late silique stage 
gene.36
write.table(gene.36, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/data/gene.36.txt")

gene.annotation <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/data/gene_annotation.csv", header = F)
gene.annotation <- gene.annotation[,c(1:2)]
rownames(gene.annotation) <- gene.annotation$V1

### combined gene list 
gene <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/data/gene.txt")
gene

rownames(gene.annotation) %in% gene$V1

## make tables 
gene.1 <- merge(gene.1, gene.annotation, by="row.names")[,c(1,2,3,4,5,6,8)] 
gene.1 
colnames(gene.1) <- c("geneID", "logFC", "logCPM", "LR", "PValue", "FDR", "Arabidopsis.hit")
write.csv(gene.1, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/data/gene.1.csv")

gene.9 <- merge(gene.9, gene.annotation, by="row.names")[,c(1,2,3,4,5,6,8)]
gene.9
colnames(gene.9) <- c("geneID", "logFC", "logCPM", "LR", "PValue", "FDR", "Arabidopsis.hit")
write.csv(gene.9, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/data/gene.9.csv")

gene.28 <- merge(gene.28, gene.annotation, by="row.names")[,c(1,2,3,4,5,6,8)]
gene.28
colnames(gene.28) <- c("geneID", "logFC", "logCPM", "LR", "PValue", "FDR", "Arabidopsis.hit")
write.csv(gene.28, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/data/gene.28.csv")

gene.36 <- merge(gene.36, gene.annotation, by="row.names")[,c(1,2,3,4,5,6,8)]
gene.36
colnames(gene.36) <- c("geneID", "logFC", "logCPM", "LR", "PValue", "FDR", "Arabidopsis.hit")
write.csv(gene.36, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/data/gene.36.csv")

rownames(fatty.acid.genes) 
vstMat.parent

### are fatty acid genes expressed at higher level in early or late silique? 
summary(de.early.vs.late.silique.ae.1 <- decideTestsDGE(lrt.early.vs.late.silique.ae, p=1))
DEgene.early.vs.late.silique.ae.1 <- topTags(lrt.early.vs.late.silique.ae,n = Inf)$table[topTags(lrt.early.vs.late.silique.ae,n = Inf)$table$FDR<0.075,]

summary(de.early.vs.late.silique.ol.1 <- decideTestsDGE(lrt.early.vs.late.silique.ol, p=1))
DEgene.early.vs.late.silique.ol.1 <- topTags(lrt.early.vs.late.silique.ol,n = Inf)$table[topTags(lrt.early.vs.late.silique.ol,n = Inf)$table$FDR<0.075,]

all.fatty.acid.genes <- unique(c(rownames(fatty.acid.genes), as.character(A.ID$V1), as.character(B.ID$V1), as.character(C.ID$V1), as.character(D.ID$V1), as.character(E.ID$V1)))

# check fatty acid genes expression 
all.fatty.acid.genes.ol.1.early.vs.late <- DEgene.early.vs.late.silique.ol.1[rownames(DEgene.early.vs.late.silique.ol.1) %in% all.fatty.acid.genes,]
dim(all.fatty.acid.genes.ol.1.early.vs.late)

sum(all.fatty.acid.genes.ol.1.early.vs.late$logFC >0) # 135 # 4
sum(all.fatty.acid.genes.ol.1.early.vs.late$logFC < 0) # 108 # 5

all.fatty.acid.genes.ae.1.early.vs.late <- DEgene.early.vs.late.silique.ae.1[rownames(DEgene.early.vs.late.silique.ae.1) %in% all.fatty.acid.genes,]
dim(all.fatty.acid.genes.ae.1.early.vs.late)

sum(all.fatty.acid.genes.ae.1.early.vs.late$logFC > 0) # 96 # 0 
sum(all.fatty.acid.genes.ae.1.early.vs.late$logFC < 0) # 147 # 2
```





